\documentclass[letterpaper,12pt]{report}
\usepackage{float}
\usepackage[top=0.7in, bottom=0.7in, left=0.75in, right=0.75in,headsep=0.2in,footskip=0.3in,portrait]{geometry}
\usepackage[ascii]{inputenc}
%\usepackage[LGR,T1]{fontenc}
\usepackage[english]{babel}
\usepackage{amsmath}
\usepackage{color}
\usepackage{array}
\usepackage{hhline}
\usepackage{hyphenat}
\usepackage{makeidx}
\makeindex
\newcommand{\indexcmd}[1]{\index{R2R commands!\texttt{#1}}}
\newcommand{\indexconcept}[1]{\index{R2R commands!#1}}
\usepackage[bookmarksopen,bookmarksopenlevel=2,bookmarksnumbered]{hyperref}
\usepackage{fancyvrb}
\hypersetup{pdftex, colorlinks=true, linkcolor=blue, citecolor=blue, filecolor=blue, urlcolor=blue, pdftitle=R2R user manual}
\usepackage[pdftex]{graphicx}
% Footnote rule
\setlength{\skip\footins}{0.0469in}
\renewcommand\footnoterule{\vspace*{-0.0071in}\setlength\leftskip{0pt}\setlength\rightskip{0pt plus 1fil}\noindent\textcolor{black}{\rule{0.25\columnwidth}{0.0071in}}\vspace*{0.0398in}}
\newcommand{\comment}[1]{}
\setcounter{secnumdepth}{3}
\setcounter{tocdepth}{3}
\newcommand{\fig}[4]{
\begin{figure}
\includegraphics{#2}
\caption{#3}

\begin{small}
#4
\end{small}
\label{#1}
\end{figure}
}
\newcommand{\example}[1]{
\begin{quote}
{\raggedright
#1
}
\end{quote}
}
\newcommand{\examplett}[1]{
\example{{\tt #1}}
}
\begin{document}
\newcommand{\myversion}[0]{1.0.6.1-49-g7bb81fb}
%\lhead{}\chead{}\rhead{}\lfoot{R2R version \myversion{} user manual}\cfoot{}\rfoot{\thepage}\renewcommand{\headrulewidth}{0pt}\pagestyle{fancy}
\floatstyle{ruled}
\restylefloat{figure}
\pagestyle{plain}
\begin{titlepage}
%\vskip 3em
\begin{center}
\begin{huge}
\noindent{}R2R version \myversion{} user manual\\
\end{huge}
\vskip 0.5em
\begin{Large}
Software to speed the depiction of\\ aesthetic consensus RNA secondary structures
\end{Large}
\vskip 1em
\begin{large}
Zasha Weinberg
\end{large}
\vskip 1em
\today
\vskip 4em
\includegraphics{figures/manA-cover.pdf}
\end{center}

\vfill
Copyright \copyright{} 2009-\the\year{} by Zasha Weinberg.  All rights reserved.

This software package, including source code and this manual, is distributed freely but without any warranty
under the terms of the GNU General Public License as published by the Free Software
Foundation.
This program is distributed in the hope that it will be useful, but WITHOUT ANY WARRANTY; without
even the implied warranty of MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
For a copy of the full text of the GNU General Public License, see \url{www.gnu.org/licenses}.
\vskip 2em

This package includes the following material from other sources:
\begin{itemize}
%\item The DejaVu font distributed in the {\tt dejavu} subdirectory and encoded within R2R is based on the Bitstream Vera fonts, which are \copyright{}Bitstream
\item The ``Squid'' library from Infernal version 0.7 by Sean Eddy.
\item Free code that is part of the CFSQP package \cite{CFSQP}, distributed by AEM Design.
\item Previously published layouts of RNAs and R2R markup to generate those layouts \cite{Weinberg22Motifs,Weinberg104Motifs,CoolRNAs,SAMIV,SAH,RegulskiMoco,SudarsanGEMM,GEMMII}.
\end{itemize}
\end{titlepage}

\clearpage\setcounter{page}{1}


\setcounter{tocdepth}{3}
\renewcommand\contentsname{}
\tableofcontents

\chapter{Introduction}

\section{What does this software do?}

\newcommand{\researchpapertitle}[0]{R2R---software to speed the depiction of aesthetic consensus RNA secondary structures}

For a full description of this software, please read the paper titled ``\researchpapertitle{}'' \cite{RtoR}.

Briefly, this software is designed to speed the drawing of RNA secondary structure consensus diagrams, which show the conserved
features within a set of related RNAs. The software also supports drawing of single RNA molecules, although this is not the emphasis.
To make RNA drawings, many biologists use general-purpose software such as Adobe Illustrator, at great cost in time
and risk of errors.
However, this strategy produces the highest-quality of drawings.
R2R is designed to allow the user to achieve this highest-quality drawing, while taking much less time than the manual solution.
Because of this goal, R2R imposes more work on the user than highly automated solutions, and 
the current version of R2R is aimed at bioinformaticians, or biologists with some familiarity with UNIX-like command line tools.

I have used R2R to draw over 100 RNA consensus diagrams.  These drawings are made available as ``demo'' files (see Chapter~\ref{GettingStarted}).

\section{What does this software {\em not} do?}

R2R has some important limitations:
\begin{itemize}
\item R2R does not fully automate determination of a layout.  Although its default layouts will get you closer to an ideal layout and it has functions to assist in determining layouts of multistem junctions, R2R does not solve the problem of determining an overall layout.  No currently available computer algorithm can determine an ideal layout that is comparable in quality to the best layouts found by a person.  Therefore, R2R assumes that a user will optimize the layout and tell R2R what to do.
\item R2R does not have a graphical user interface.  As noted above, R2R is aimed at bioinformaticians.  Successful use of R2R will likely require some general familiarity with the UNIX command line, and comfort with command-driven programs.
Because of the lack of a graphical user interface, R2R might have a significant learning curve.
So, if you just want to draw one RNA, it might be most efficient to go straight to Adobe Illustrator, Inkscape or CorelDRAW.
\item R2R is not designed to produce drawings that illustrate many elements of tertiary structure, or whose layouts are based on atomic-resolution 3-D structures.
R2R is only intended to create more abstract layouts of secondary structure that reflect Watson-Crick base pairing.
\item R2R is not a fully general drawing program, or even a fully general RNA-drawing program.  R2R should be used in combination with general-purpose drawing programs like Adobe Illustrator or Inkscape.
  \item R2R does not quite respect the Stockholm file format.  See section~\ref{sec:notstockholm} for details.
\end{itemize}

\section{Credit}

If you use R2R, please cite the paper \cite{RtoR}.

If you distribute R2R or derivatives of it, please continue to credit Infernal, as on the first page of this manual.

\section{Licensing}

The files in the main directory (documentation), the {\tt src} and the {\tt demo}
subdirectories are copyright 2009-\the\year{} by 
Zasha Weinberg, except as noted.  The entire package is distributed freely but without any warranty whatsoever
under the GNU General Public License.
For details, see \url{http://www.gnu.org/licenses}.

\section{Changes from previous versions}

\begin{itemize}
\item Version 1.0.7 (anticipated)
  \begin{itemize}
  \item Added the ability to change the names of the individual drawings that R2R outputs into PDF of SVG files.  This is for use by R-scape.  See Section~\ref{sec:displayName}.
  \item Added ability to change the name of the font that R2R puts in SVG files, so that Inkscape is happier.  This has no effect on generated PDF files.  See Section~\ref{sec:runRtoR} and Section~\ref{sec:Makefile:perloptions}.
  \item When ignoring secondary structures line (usually for pseudoknots), you can now get R2R to draw an outline around the paired nucleotides, without drawing colored rectangles (for covariation) that relate to that secondary structure line.  See the {\tt outline-no-bpannot} option in section~\ref{sec:ignoressexceptforpairs}.
  \item Added ability to use different sequence-weighting algorithm with the Makefile functionality.  See Section~\ref{makefileseqweighting}.
  \item With the the Makefile functionality, added ability to set a global set of drawing parameter values (See Section~\ref{SetDrawingParam} for information on drawing parameters) that apply to all drawings, effectively setting new default values for parameters, in case you don't like the default values I selected.  (See Section~\ref{sec:Makefile:perloptions}.)
  \item R2R will now attempt to draw the text marking the $5^\prime$ end, as well as a line connecting to the first nucleotide, even if the first nucleotide runs at a weird angle (i.e., not a multiple of 90).  It seems to mostly look reasonable.  This is influenced by the new {\tt backboneConnectorCircleRadius} drawing param (See Section~\ref{SetDrawingParam}.), but the default value seems to work pretty reasonably.
    \item Added an ability to control what R2R calls the various pseudoknots in callout-style drawings of pseudoknots.  See the {\tt prefixSsWithPkInDrawings} parameter in Section~\ref{SetDrawingParam}.  This feature is for R-scape.
  \item Added various drawing parameters (See Section~\ref{SetDrawingParam}.):
    \begin{itemize}
    \item {\tt nucTickLabel\_tickColor}, which determines the color of all lines that are part of tick labels.  There is no way to set the color of individual tick lines.
    \item {\tt varBackboneFontSize} : font size for text with {\tt var\_backbone} and related commands.
    \item {\tt varTermLoopFontSize} : font size for text with {\tt var\_term\_loop} command.
    \item {\tt backboneAnnotTextOffset} : is now, by default, set based on {\tt varBackboneFontSize}.  (And it's now documented.)
    \item {\tt backboneConnectorCircleRadius} :the radius of the arc that R2R uses to bend the backbone, esp. when drawing the $5^\prime$ end of the molecule.
    \end{itemize}
  \item The script {\tt SelectSubFamilyFromStockholm.pl} is now installed into the {\tt bin} directory in {\tt make install}.
  \item added {\tt -{}-cutEmptyLines} flag for {\tt r2r}, to supress blank lines added in some circumstances by the {\tt -{}-GSC-weighted-consensus} flag. These blank lines are intended to preserve line numbers, so that error messages make sense with the original file, but they violate Stockholm format, which is a hassle in some cases. (See Section~\ref{sec:gscgeneral}.)
  \item Fixed a bug where an input file without a {\tt \#=GC SS\_cons} line caused a segmentation fault.
  \item Documented incompatibilities between R2R and the Stockholm format.  See section~\ref{sec:notstockholm}.
  \item Documented other flags for {\tt r2r} with the {\tt -{}-GSC-weighted-consensus} flag.  See Section~\ref{sec:gscgeneral}.
  \item Added ability to explicitly set the covariation shading of base pairs.  See Section~\ref{sec:setcovaryshade}.  
  \end{itemize} 
\item Version 1.0.6
  \begin{itemize}
  \item Integrated NLOPT optimization library as an alternative to CFSQP.  NLOPT does not seem to work quite as well as CFSQP for some problems posed by R2R.  But, it mostly works comparably and is free software, whereas CFSQP is annoying to obtain.  Note: NLOPT has a lot of solvers and parameters I didn't try out, so it's possible that it could work better if set up differently.
  \item Added ability to have secondary structure lines for drawing individual sequences (i.e., in {\tt oneseq} mode).  Any line beginning {\tt \#=GR {\it hitId} SS\_cons\_{\it name}} will be addressable as a secondary structure under the name ``{\tt \_}{\it name}''.
  \item R2R now has an option to change the meaning of red shading of base pairs.  Red shading means there is no variation in the base pair.  In this case, covariation could arise simply due to an absence of variation, and there is no evidence against base pairing.  However, due to a detail in implementation, R2R previously allowed for rare occurrences (by default up to 10\%) of non-canonical base pairs.  This remains the default behavior, but there is an option to avoid the red shading in cases where there are non-canonical pairs.
  \item Other minor changes / bug fixes.
  \end{itemize}
\item Version 1.0.5
  \begin{itemize}
    \item R2R now has a {\tt ./configure} script.  The installation instructions have changed accordingly.
    \item The manual has been corrected as follows.  R2R does not work with the interleaved format of Stockholm format alignments.  R2R only works with the Pfam format, which is equivalent to Stockholm format, except that the full sequence or column annotation must be given in one line (for hit sequences, or data in \#=GC or \#=GR lines), and these lines cannot be split.
    \item R2R now prints a warning about inappropriate use of R2R, specifically against using R2R to evaluate structural conservation.  R2R is not intended to evaluate evidence for covariation or RNA structure where this is in question. It is not appropriate to use R2R's covariation markings to declare that there is evidence of structural conservation within an alignment.  R2R is a drawing program.  As the R2R paper wrote\cite{RtoR}:
\begin{quote}This automated R2R annotation [of covariation] does not reflect the extent or confidence of covariation. While such information can be useful, we believe that thorough evaluation of covariation evidence ultimately requires analysis of the full sequence alignment. For example, misleading covariation can result from an incorrect alignment of sequences, or from alignments of sequences that do not function as structured RNAs. Unfortunately, there is no accepted method to assign confidence that entirely eliminates the need to analyze the full alignment.
\end{quote}
You can disable this warning by running R2R with the flag {\tt -{}-disable-usage-warning}.
    \item When drawing a single sequence (i.e., in {\tt oneseq} mode), R2R can highlight nucleotides that are highly conserved in the consensus.  This feature makes these highly conserved nucleotides apparent when, for example, depicting an RNA molecule on which experiments were done, and where mutations might have been made. To enable this feature, see {\tt makeRedNucsRedInOneseq} and {\tt makeNondegenRedNucsRedInOneseq} in Section~\ref{SetDrawingParam}.
    \item Added the command {\tt tick\_label\_disable\_default\_numbering} and the ability to start numbering nucleotides with an arbitrary number, by adding {\tt firstNucNum} to the {\tt tick\_label\_regular\_numbering} command. See Section~\ref{sec:ticklabel}
    \item Improved miscellaneous error messages.
    \end{itemize}
\item Version 1.0.4
	\begin{itemize}
          \item Fixed problem where tab characters between sequence name and nucleotides within the input alignment can cause errors.  Tab characters are now treated like spaces.
          \item When you use the {\tt ignore\_ss\_except\_for\_pairs} command with the {\tt outline} option, R2R will annotate the pseudoknot outlines with the name of the corresponding {\tt \#=GC SS\_cons} line.  This helps to match up pseudoknots for RNAs that contain multiple pseudoknots.
            \item Changes to {\tt SelectSubFamilyFromStockholm.pl}, allowing a predicate to easily use markup from a name other than the predicate's name.  See {\tt SUBFAM\_STRING} field in {\tt SUBFAM\_REGEX\_PRED} and {\tt SUBFAM\_PERL\_PRED} (Section \ref{subfampredicates}).
          \item Fixed typo in tutorial.
	\end{itemize}
\item Version 1.0.3 (August, 2012)
        \begin{itemize}
	\item R2R 1.0.3 cannot exactly reproduce older drawings of RNA structures.  If you want to reproduce the old drawings, you should use R2R version 1.0.1.

The way that R2R handles degenerate nucleotides when calculating the consensus sequence has changed.  Degenerate nucleotides are now ignored in the calculations, where previously they were treated as a uniform distribution of the possible nucleotides (e.g., degenerate nucleotide `R' was treated as 50\% A and 50\% G).  As a result, the current R2R does not exactly reproduce old drawings.  In some cases, the layouts change when the differences lead to nucleotides falling above or below the 50\%-present level, thus adding or removing nucleotide positions from the drawing.
        \item Added ability to use alignment positions as labels.  See Section~\ref{alignPosLabel}
	\item Added {\tt dumpInfoFile} drawing parameter, which dumps information on layout to a tab-delimited file.
        \item Added section on interacting with R2R using scripts.  See Section~\ref{scriptedInteraction}.
        \end{itemize}
\item Version 1.0.2
	\begin{itemize}
	\item Added citation of R2R paper in {\it BMC Bioinformatics} to this manual.
	\item Edited some error messages to make them more clear.
	\item Check for conflicting use of {\tt bulge} and {\tt place\_explicit}.
	\item Added {\tt make\_pair} command.
	\item Added {\tt DNA}, {\tt shadeAlongBackboneWidth} and {\tt alongBackboneMidpointGapLength} drawing parameters.
	\item Added ability for user to weight sequences using an arbitrary method (instead of the GSC algorithm), while calculating frequencies otherwise as before.  See Section~\ref{manualweighting} and Section~\ref{makefilemanualweighting}.
	\end{itemize}
\item Version 1.0.1 (distributed with revised version of paper, November 2010)
	\begin{itemize}
	\item Added demo: {\tt demo/c-di-GMP-II/c-di-GMP-II-update.sto} (re-drawing based on new sequences, for Wikipedia article), also some files in the {\tt demo/pedagogical} directory.
	\item Added Additional files 3 and 4 from the paper to software distribution.
	\item Added feature: {\tt indicateOneseqWobblesAndNonCanonicals} added as a drawing parameter.
	\item Fixed bugs: 
		\begin{itemize}
		\item Fixed a problem when internal loops are converted to bulges (explicitly using the {\tt internal\_loop\_to\_bulges} command, or implicitly by applying a {\tt bulge} or {\tt place\_explicit} command to a nucleotide within the internal loop), and when the user did not specify what to do with both sides of the internal loop.  
%If the user did not specify an action with both the $5^\prime$ and $3^\prime$ sides of the internal loop, an internal error was triggered.  R2R now deals with the issue more gracefully, by drawing the un-specified side as a straight line.
		\item Fixed various issues with applying the {\tt multistem\_junction\_circular} command to $3^\prime$ bulges.
                \item Formatting issues with the user manual. The {\tt -{}-GSC-weighted-consensus} flag was explained in more detail.
		\end{itemize}
	\end{itemize}
\item Version 1.0 (distributed with initial submission of paper, July 2010)
\end{itemize}

\chapter{Installation}
\label{sec:install}

Installation essentially means making the executable file {\tt r2r}.%  You might need to install the DejaVu Sans Condensed font.

\section{Platforms on which R2R is known to work}

Building of R2R has only been tested with the gcc compiler suite (\url{http://gcc.gnu.org}).  To ensure compatibility, you will need the GNU C compiler (which provides the {\tt gcc} command) and the GNU C++ compiler (which provides the {\tt g++} command).

R2R has been tested on the following platforms, however only the original version:
\begin{itemize}
\item gcc version 3.4.4 under Cygwin 1.5.25 (32-bit) on Windows XP.  NOTE: you'll need to install the C/C++ compiler gcc and the make program.  These are not installed by default in Cygwin.
\item gcc version 4.3.4 under Cygwin 1.7.1 (32-bit) on Windows 7.
\item gcc version 5.3.0 under Cygwin 64-bit on Windows 7.  R2R version 1.0.5 works on this platform.
\item gcc version 4.1.2 under Ret Hat Linux running Linux kernel 2.6.18 (64-bit).  R2R version 1.0.5 works on this platform.
\item gcc version 3.3 under MacOS's Darwin version 8.8.0 (32-bit).  NOTE: you'll need to install the Xcode package to get the C/C++ compiler gcc and the make program.
\end{itemize}
I presume that R2R will work on gcc version 3 or higher on Cygwin, Linux or MacOS Darwin.

R2R is known to produce output that can be used with Adobe Illustrator CS on Windows XP, Inkscape 0.46 on Windows XP and CorelDRAW Graphics Suite X4 on Windows Vista.
I presume, however, that the output will work on any version of these programs, at least
with appropriate setting of fonts (see below).

\section{Decide what NLP solver to use (if any)}
\index{CFSQP}

R2R implements multiple methods to find a good layout for multistem junctions.  Some automated solutions require software that can solve non-linear optimization problems.  You have three options: don't install any NLP solver, install the NLOPT solver or install the CFSQP solver.

These options affect whether you can use the following commands that using a non-linear optimization to solve the layout of multistem junctions: {\tt multistem\_junction\_circular\_solver}, {\tt multistem\_junction\_bulgecircley\_solver} and {\tt multistem\_junction\_bulgecircleynormals\_solver}.  If you do not use these commands, there's no advantage to installing an NLP solver.

\begin{itemize}
\item Option 1: Do not install a solver.  (This is the default.)

\begin{itemize}
\item Effort required of you: none.  It's the default.
\item Ability to solve multistem junctions: none.  The functions requiring NLP optimization will not work if there's no solver.  If you try to use one of the related commands, R2R will report an error.
\end{itemize}

\item Option 2: Install NLOPT.

\begin{itemize}
\item 
Effort required of you: a bit.  You must download \& install the NLOPT library, but that's not especially hard.
\item Ability to solve multistem junctions: pretty good.  In some cases, NLOPT does not work as well as CFSQP (see next option, below).  NLOPT is more often than CFSQP unable to solve the optimization problem well.  However, most of the time, NLOPT and CFSQP yield the same result.  Note: NLOPT has a lot of solvers and parameters I didn't try out, so it's possible that it could work better if set up differently.
\end{itemize}
\item Option 3: Install CFSQP.

\begin{itemize}
\item 
Effort required of you: more.  Due to its licensing requirements, CFSQP must be requested from AEM Design, which requires additional time, and has been a hassle in the past.  (I am not legally allowed to distribute CFSQP with R2R.) Once you get the relevant file, however, it's pretty straightforward.
\item Ability to solve multistem junctions: good.  CFSQP is not always able to do it, but it works well most of the time.
\end{itemize}
\end{itemize}

\comment{
\section{Install the DejaVu font}

R2R will generate PDF files with the DejaVu Sans Condensed font.
You should ensure that this font is installed, and available to your
drawing program (Adobe Illustrator, Inkscape or Corel Draw).
Otherwise, your drawing program might map the font to a different one, which could
change the layout.

{\bf Note}: if you want to get started quickly, you can skip this step for now.
The PDF files generated by R2R can be viewed in Adobe Reader regardless of whether
the font is installed.

The font and installation instructions are available in the {\tt dejavu} subdirectory of the R2R distribution.  Only the DejaVu Sans Condensed font is actually used by R2R-generated PDF files.  However, you might also want to install the bold and oblique versions of this font, for use in your drawing program.
The source of the font files are \url{http://dejavu.sourceforge.net}
}

\section{Build R2R}

\subsection{Extract files from the R2R .tgz archive file}

The R2R files are contained within the .tgz file.  You can extract these files
using WinZip on Windows or by double-clicking on Mac OS X.
In a UNIX system (including Darwin and Cygwin), you can run a command like
{\tt tar xzf /path/to/where/you/downloaded/R2R.tgz}.

\subsection{Optionally choose an SVG font (very unusual)}

If you intend to use Adobe Illustrator or CorelDRAW, you can use the PDF (Adobe Acrobat) output of R2R.
In this case, you do not need to set a font.

If you intend to use Inkscape, you should use the SVG output of R2R, since Inkscape is not reliable when importing
PDFs generated by R2R.
By default, R2R is set up to produce SVG output using the ``Bitstream Vera Sans'' font, which is freely available
(\url{http://www.gnome.org/fonts/}) and might already be installed on your system.
If, however, Inkscape is not mapping your font correctly, you can get R2R to generate SVG files that reference a different font.

(Note: it might be easiest to skip this step initially, and change the font later if there are problems in Inkscape.  To change the font later, you'll need to edit {\tt src/Makefile.in}, and run {\tt ./configure} again, as described below.)

To change this font, open the file {\tt src/Makefile.in} in a text editor and change the definition of the {\tt INKSCAPE\_HELVETICA\_FONTNAME} define within the double quotes.  (Search for ``FONTNAME''.)  I'm not sure that anyone really has to do this, so I haven't made this part of the configure script.

Note: R2R is internally hardcoded with font geometry that is appropriate to Helvetica, Arial, Bitstream Vera or DejaVu fonts.  These fonts have similar sizes to each other.  If you use fonts whose symbols have significantly different sizes, nucleotides will not be positioned correctly.

\subsection{Optionally install NLOPT}

If you want to use NLOPT, you'll have to install the library.  The home page for the NLOPT library is here:
\url{https://nlopt.readthedocs.io/en/latest/}

When you build R2R, pass the {\tt -{}-enable-nlopt} flag to {\tt ./configure} (see below).

\subsection{Optionally get CFSQP}

If you want to use the CFSQP solver, you must request the 
proprietary file {\tt cfsqp.c}.  Go to \url{http://www.umventures.org/technologies/cfsqp\%E2\%84\%A2-version-23} or \url{https://www.isr.umd.edu/news/news_story.php?id=4087} or Google for
CFSQP.  (The contact has changed recently.)

Once you have obtained it, put the file {\tt cfsqp.c} into the R2R directory {\tt NotByZasha/cfsqp}.

When you build R2R, pass the {\tt -{}-enable-cfsqp} flag to {\tt ./configure} (see below).

\subsection{configure and make}

Run {\tt ./configure} to configure R2R for your system.
\begin{quote}
{\tt ./configure}
\end{quote}

The {\tt ./configure} script can take the following flags, in addition to the standard {\tt ./configure} script flags (run {\tt ./configure -{}-help} for a complete list):
\begin{itemize}
\item {\tt -{}-enable-nlopt} : compile with NLOPT.  Note: this requires that the compiler can find the NLOPT include and library files.  You might need to adjust the CFLAGS and LDFLAGS environment variables appropriately before running {\tt ./configure} so that it can find the NLOPT files.
\item {\tt -{}-enable-cfsqp} : compile with CFSQP.  Note: this requires that you have the file {\tt cfsqp.c} (see previous section)  If you enable both NLOPT and CFSQP, then R2R will use CFSQP.
\end{itemize}

Then run
\begin{quote}
{\tt make}
\end{quote}
Note: if you include CFSQP, you might get warnings about redefinitions of the symbol {\tt \_\_STDC\_\_}.  These warnings are benign, and you can ignore them.

If you want to copy the executable into your binary directory, optionally run
\begin{quote}
{\tt make install}
\end{quote}
Otherwise, you can use the {\tt r2r} executable file from within the R2R package directory.

\section{Adobe Reader glitch}
\label{sec:SillyReader}

This section describes a potential problem when viewing PDF output with Adobe Reader.  With recent versions of Adobe Reader and certain computer configurations, circles and arcs appear jagged---more like polygons.  To prevent this, while running Adobe Reader, select ``Preferences'' under the ``Edit'' menu.  Within the category ``Page Display'', uncheck ``Use 2D graphics acceleration'' and make sure that ``Smooth line art'' is checked.

\section{Alignment editor software}

R2R's input is based upon multiple-sequence alignments that are
stored in Pfam format, which is a special case of Stockholm format in which interleaved lines are not permitted.
Although these files can be edited in any text editor, some programs
are customized for them.
One example is RALEE \cite{RALEE}, which is a set of macros for
the Emacs text editor.
Please see section~\ref{sec:notstockholm} for a full discussion of how R2R does not fully implement the Stockholm file format.
You can convert Stockholm-format alignments into the Pfam format using the {\tt esl-reformat} command that comes with the HMMER and Infernal software packages in the {\tt easel/miniapps} directory.

\chapter{Tutorial: a guide to R2R with examples}
\label{GettingStarted}

This chapter gives a practical introduction with examples on how to create RNA drawings with R2R.
Example input files containing R2R commands to draw various RNA structures are available in the {\tt demo} subdirectory.

Credit: the Stockholm-format input files in the {\tt demo} directory and its subdirectories are similar or identical to supplementary data published in previous reports \cite{Weinberg22Motifs,SAMIV,RegulskiMoco,SAH,Weinberg104Motifs,CoolRNAs,GEMMII}.

\section{About the demo files}

\subsection{Demo output is already provided}

I have created the raw output of R2R when run on the demo files, and this output is provided within the {\tt output-pdf} (PDF format) and {\tt output-svg} (SVG format) directories.
If you just want to see the output, I recommend opening the PDF files using Adobe Reader (\url{http://get.adobe.com/reader}), as this should work on any platform.
(Note: if circles look very unsmooth when viewed in Adobe Reader, please see Section~\ref{sec:SillyReader}.)
If you want to try editing the output files, please see Section~\ref{sec:pdforsvg} (next) to find out which files to use.

If you want to create the output files yourself, see Section~\ref{sec:MakeDemos}.

\subsection{Drawing programs: Adobe Illustrator or Inkscape}
\label{sec:pdforsvg}
\index{Adobe Illustrator}
\index{Inkscape}
\index{PDF output}
\index{SVG output}

Simple conclusion:
\begin{itemize}
\item If you're using Adobe Illustrator or CorelDRAW, use PDF output.  The example commands below use PDF.
\item If you're using Inkscape, use SVG output.  In the commands below, wherever it says {\tt pdf},
instead write {\tt svg}.
%  The {\tt Makefile} will make both PDF and SVG outputs, so you can see the SVG commands, and they are basically the same as with PDF.  
(R2R decides the output format based on the file extension
of its output file, either {\tt .pdf} or {\tt .svg}.)

If you have problem with the fonts in Inkscape (or if the letters don't show up), you might need to re-create the SVG output with a different font name.  Please see the next section (``more technical information'') to learn more about this, and Chapter~\ref{sec:install} for information on how to change the font.
\end{itemize}

\subsubsection{More technical information}

R2R will generate valid PDF and SVG files as output. In principle, any program should be able to read them.
However, I am exploiting the fact that the Helvetica font is built in to PDF.  Some versions of Inkscape
will not render any text using PDFs built by R2R because they don't have the Helvetica font.

Therefore, R2R can create SVG output, which is Inkscape's native file format.
The SVG output uses the Bitstream Vera Sans font, which is similar to Helvetica and to Inkscape's default font.  (Meanwhile, at least
on Windows, Illustrator will substitute Arial, which is similar.)

Note: in principle you can substitute other fonts, even by creating scripts to edit SVG output.
However, R2R knows the heights and width of font symbols, which is uses for its layout.  Thus, for example, if you use a narrower font, the spacing might not be appropriate.

\subsection{Where the demo files come from}

Files within the {\tt demo} directory were made to support this user manual.  Most are based on previously drawn RNAs.
RNA drawings derived from previous studies are organized into subdirectories.
The subdirectories and relevant citations are as follows:
\begin{itemize}
\item {\tt demo/22} (Weinberg {\it et al.}, 2007) \cite{Weinberg22Motifs} and (Sudarsan {\it et al.}, 2008) \cite{SudarsanGEMM}.
\item {\tt demo/104} (Weinberg {\it et al.}, 2010) \cite{Weinberg104Motifs}.
\item {\tt demo/exceptional} (Weinberg {\it et al.}, 2009) \cite{CoolRNAs}. Note:
	\begin{itemize}
	\item The RNAs depicted in Supplementary Figure 11 of this paper are primarily described in the (Weinberg {\it et al.}, 2010) paper \cite{Weinberg104Motifs}, and are therefore found within the {\tt demo/104} directory.
	\item The drawing of GOLLD RNA is somewhat complicated by different degrees of conservation of the structural elements.  For instructions on reproducing the drawing of GOLLD RNA in Fig. 2a, see Section~\ref{sec:GOLLD}.

Also, a careful look at Fig. 2a in {\it Nature} will reveal that many nucleotides are slightly misaligned.  The reason is that the R2R drawings in this case were done using Adobe's Myriad font, which is narrower than Helvetica.  The journal changed the font, but did not realign the nucleotide letters.  The output of R2R is always aligned correctly.
	\item The HEARO RNA ``skeleton'' drawing in Figure 3b was not drawn with R2R.  However, an R2R drawing of this style is included.  Figure 3b was drawn before I had implemented skeleton drawings in R2R.
	\end{itemize}
\item {\tt demo/c-di-GMP-II} (Lee {\it et al.}, 2010) \cite{GEMMII}
\item {\tt demo/Moco} (Regulski {\it et al.}, 2008) \cite{RegulskiMoco}.  (New drawing, based on original layout.)
\item {\tt demo/SAH} (Wang {\it et al.}, 2008) \cite{SAH}. (New drawing, based on original layout.)
\item {\tt demo/SAM-IV} (Weinberg {\it et al.}, 2008) \cite{SAMIV}.
\item {\tt demo/hammerhead} (Perreault {\it et al.}, 2010) \cite{hammerhead}
\item {\tt demo/additional} Drawings of RNAs that were identified by other groups.  The tRNA drawing is based on the standard layout.
\item {\tt demo/pedagogical} Some drawings I did to explain issues in RNA drawing to non-biologists, or for the R2R paper.
\end{itemize}

Figures within each directory are organized by motif name.
Thus, for example, in (Weinberg {\it et al.}, 2009) \cite{CoolRNAs}, Figure 2a is a consensus diagram of GOLLD RNA.
This corresponds to the files {\tt demo/exceptional/GOLLD.sto} and {\tt demo/exceptional/GOLLD.r2r\_meta}.
If you run {\tt r2r} on these files, the output will be found in {\tt output-pdf/GOLLD.pdf} or {\tt output-svg/GOLLD.svg}.
Similarly, Supplementary Figure 1 of the same previous publication contains a skeleton diagram of GOLLD RNA,
Supplementary Figure 2 has an expanded consensus diagram and Supplementary Figure 3 holds a different skeleton diagram and
a drawing of certain parts of the {\it L. brevis} GOLLD RNA.
All these figures are generated from the same files, and R2R output used for this will also appear in
{\tt output-pdf/GOLLD.pdf} or {\tt output-svg/GOLLD.svg}.
However, updates to the way that consensus nucleotides are calculated with degenerate IUPAC nucleotides have slightly changed the picture.  Versions of R2R up to 1.0.1 will exactly reproduce the old drawing.

\section{Example legend to help you with finished drawings}

To create a finished drawing for a published figure, a legend is necessary to explain the annotations.
We have created generic annotations usable for a legend:
\begin{itemize}
\item Use the file {\tt demo/Additional-file-3.pdf} for PDF format that can be imported into
Adobe Illustrator or CorelDRAW.
\item Use the file {\tt demo/Additional-file-4.svg} for SVG format that can be imported into Inkscape.
\end{itemize}

\section{A hypothetical motif to illustrate the basics}

Now the actual tutorial begins.

\subsection{Default output of R2R}
\fig{fig:demo1}{figures/demo1.pdf}{demo1: A simple alignment and R2R drawing}{
(A)
A minimal Stockholm-format alignment of a hypothetical RNA motif.
This is the file {\tt demo/demo1.sto}.
(B)
The raw output of R2R when run on the alignment in part A.
Note that the text ``demo1'' is part of R2R's raw output.
(C)
Annotations of the alignment file.
(D)
The Stockholm-format alignment file generated by R2R
(with the {\tt -{}-GSC-weighted-consensus} flag)
 that has information on the consensus
of the demo1 motif.
(Some minor changes were made to improve readability.)
Note that it is not necessary to understand the information in this file, because the file is simply used by R2R to produce its final output.
The way in which R2R calculates consensus diagrams, and the meanings of the colors and symbols is explained in the research paper, as well as in Chapter~\ref{consensus}.
(E)
The contents of the file {\tt demo/demo.r2r\_meta}.  In this case, the file simply lists the name of input
file used to draw the consensus.
}
Figure~\ref{fig:demo1} shows a simple alignment, and R2R's drawing of the consensus of the alignment.  
Elements of the Stockholm-format alignment file are illustrated.  Note that R2R cannot process ``interleaved'' Stockholm files, in which sequences can span multiple lines; this restricted form of the Stockholm format is called the Pfam format.
Please see section~\ref{sec:notstockholm} for a complete discussion on how R2R does not fully conform to the Stockholm file format.
You can convert Stockholm-format alignments into the Pfam format using the {\tt esl-reformat} command that comes with the HMMER and Infernal software packages in the {\tt easel/miniapps} directory.
\index{Stockholm file format}

The drawing was made by running the following commands within the {\tt demo} directory.  (Actually,
I really just ran {\tt make}, which reads the {\tt Makefile} in that directory to run these commands.
If you're familiar with the UNIX {\tt make} command, and you're going to make many drawings, you might want
to read Chapter~\ref{sec:metamake} after reading this tutorial.)

First, in order to draw the consensus structure, we must first run a command
to calculate it.
We calculate the alignment's consensus (Figure~\ref{fig:demo1}D) using the following command line (note: the following command should be one line, but was split into two lines for typesetting):
\index{\texttt{r2r} executable!\texttt{-{}-GSC-weighted-consensus}}

\example{
{\tt commandprompt\$ ../src/r2r -{}-GSC-weighted-consensus demo1.sto}\hfill\begin{small}{\it (continued)}\end{small}\\
{\tt\hspace{0.3in} intermediate/demo1.cons.sto 3 0.97 0.9 0.75 4 0.97 0.9 0.75 0.5 0.1}
}
(All commands in this section should be run within the {\tt demo} directory that is part of the
R2R distribution.)

The general form of this command is
\example{
{\tt commandprompt\$ ../src/r2r -{}-GSC-weighted-consensus} {\it input-file}{\tt .sto}\\
\hspace{0.5in} {\it output-file}{\tt .sto} 3 0.97 0.9 0.75 4 0.97 0.9 0.75 0.5 0.1
}
This command is explained in Chapter~\ref{consensus}, and the parameters are fully explained in
Section~\ref{sec:gscgeneral}.
Also, while it is necessary to generate consensus information so that R2R can draw it,
users can write their own routine to generate the consensus data instead of using
the {\tt -{}-GSC-weighted-consensus} command; this alternative is
explained in Section~\ref{sec:consensuswithoutGSC}.

Given the consensus data generated by the previous command, we run R2R to create a PDF file.  The file {\tt demo1.r2r\_meta} simply
gives the path of {\tt intermediate/demo1.cons.sto}, which is what R2R will use to draw.

\index{\texttt{r2r} executable!Invoking for drawing}
\examplett{
commandprompt\$ ../src/r2r demo1.r2r\_meta output-pdf/demo1.pdf
}

Alternately, to create SVG-format output:
\examplett{
commandprompt\$ ../src/r2r demo1.r2r\_meta output-svg/demo1.svg
}

Note: a warning is printed in the output PDF or SVG about misusing R2R to analyze structural conservation, rather than just drawing diagrams.  You can disable this warning by adding the {\tt -{}-disable-usage-warning} right after {\tt ../src/r2r}

Key points:
\begin{itemize}
\item R2R inputs are a subset of Stockholm-format where interleaved lines are not permitted.  The Stockholm format is explained more at \url{http://en.wikipedia.org/wiki/Stockholm_format}.
\item Consensus secondary structure is expressed using brackets (e.g., $<$ and $>$) in the {\tt \#=GC SS\_cons} line (e.g., Figure~\ref{fig:demo1}).
\item R2R will infer the consensus sequence and classify covariation.
\item Creation of the output requires two UNIX commands.
\end{itemize}

\subsection{Some typical customizations for consensus diagrams}
\fig{fig:demo1ii}{figures/demo-ii.pdf}{demo1-ii: Some annotation for the consensus diagram}{
(A)
Stockholm-format alignment that includes a labeling line (beginning {\tt \#=GF R2R\_LABEL}) and associated R2R commands (beginning {\tt \#=GF R2R}).
Annotations in blue show how R2R commands are associated with a symbol in the labeling line. Annotations also mark the special dash symbol in the {\tt \#=GC R2R\_LABEL} line that causes the column to be deleted from R2R's drawing.
Also it is important to note that a single space character separates all elements of R2R commands.
(This text is identical to the file {\tt demo/demo1-ii.sto}.)
(B)
Raw output of R2R, with annotation (in blue) showing what elements of the drawing were the result of the R2R commands in part A.
%\hrule
}

In this section we will use some R2R commands to improve the consensus diagram drawn in the previous section.
This will entail adding an {\tt \#=GC R2R\_LABEL} line, which will allow us to label and refer to specific columns in the alignment.
\indexconcept{Labels}

The demo1 drawing in Figure~\ref{fig:demo1}B has $5^\prime$ and $3^\prime$ flanking regions that are not
conserved according to R2R's definition of conservation.  Although it is often desirable 
to keep such poorly conserved flanking regions in curated alignments, they add little to a diagram.
By adding dashes to the {\tt R2R\_LABEL} line, we cause R2R to remove those columns in the output (see Figure~\ref{fig:demo1ii}).
\indexconcept{Columns, removing}

\indexcmd{var\_hairpin}
\indexcmd{var\_backbone\_range}
The first hairpin of the demo1 motif varies in length, and its terminal loop is poorly conserved.
The R2R {\tt var\_hairpin} command allows us to replace part of the hairpin with an abstract representation of a variable-length hairpin.
Similarly, a part of the junction between the hairpins is poorly conserved, and the R2R {\tt var\_backbone\_range} command replaces this with a line.  The line is annotated with the range in the number of nucleotides contained in those columns, which is automatically calculated.
For both commands, the range is represented by two labels, each of which refer to a column in the {\tt \#=GC R2R\_LABEL} line. For example, the {\tt var\_hairpin} command used the labels {\tt \textless} and {\tt \textgreater}, each of which refers to a specific column, as shown (Figure~\ref{fig:demo1ii}A).

\indexcmd{tick\_label}
Finally, the drawing uses the {\tt tick\_label} command, which labels a specific nucleotide position.
In this case, I imagined that a particular G nucleotide has a special biochemical significance in this fictitious RNA structure.

Key points:
\begin{itemize}
\item R2R commands refer to specific alignment columns using labels, in the {\tt \#=GC R2R\_LABEL} line.
It is also possible to expand the number of labels beyond simply one symbol
(see Section~\ref{sec:labels}).
\item Columns can be deleted by putting dashes in the {\tt \#=GC R2R\_LABEL} line.
\item R2R has commands to represent variable-length regions of different sorts.
For more information on R2R commands related to variable-length regions see Section~\ref{sec:varlen}.
\item Specific positions can be labeled with a tick mark using the {\tt tick\_label} command.
\end{itemize}

\subsection{Other kinds of drawings: single-molecule and skeleton diagrams}
\fig{fig:demo1iii}{figures/demo-iii.pdf}{demo1-iii: Other kinds of drawings}{
(A)
The {\tt .r2r\_meta} file name ({\tt demo/demo1-iii.r2r\_meta}),
which directs R2R to draw a regular consensus diagram (first line),
a single sequence ({\tt oneseq}) from the ``{\tt bigfoot}'' sequence
and a skeleton-style schematic drawing of the consensus.
Note that the components in each line are separated by a single tab character.
(B)
Stockholm-format alignment.
(This text comes from the file {\tt demo/demo1-iii.sto}.  Some irrelevant markup has been removed.)
The two lines beginning {\tt \#=GF R2R\_oneseq} are only processed within {\tt oneseq} mode (i.e., only for the {\tt bigfoot} sequence in this example).
They refer to column labels with alternate names ({\tt bf} and {\tt bfl}) that
are introduced using {\tt \#=GC R2R\_XLABEL\_bf} and {\tt \#=GC R2R\_XLABEL\_bfl}.
In the {\tt oneseq} case, dashes in the {\tt R2R\_LABEL} line are ignored,
but dashes in the {\tt \#=GR bigfoot DEL\_COLS} lines cause nucleotides to be
eliminated from the drawing.
The last line, which contains {\tt if\_skeleton}, is only interpreted
in skeleton mode, and it overrides a default drawing parameter to set the
basepair bond width to 0.5 points.
(C)
Raw output of R2R (rearranged to fit the figure space efficiently).
From top to bottom is a standard consensus diagram,
a single RNA molecule and a skeleton schematic.
}
\fig{fig:SAMIValternate}{figures/demo-SAM-IV-alternate.pdf}{Other kinds of drawings illustrated with SAM-IV riboswitches}{
Examples of {\tt oneseq} and skeleton drawings using SAM-IV riboswitches \cite{SAMIV}.
These examples are built from the files {\tt demo/SAM-IV/SAM-IV.sto} and {\tt demo/SAM-IV/SAM-IV.r2r\_meta}.
(A) {\tt oneseq} drawing of a SAM-IV riboswitch from
{\it Streptomyces coelicolor} annotated with cleavage data from
in-line probing experiments \cite{SAMIV}.
The annotation of cleavage data used a {\tt \#=GR ... CLEAVAGE} line, as
explained in Section~\ref{cleavage}.
Note that this is the raw output of R2R, and some adjustment
(in Adobe Illustrator or Inkscape) would be needed to position some of the nucleotide
number labels in appropriate places.
(B) R2R skeleton drawing of SAM-IV riboswitch consensus.
This layout is the same as in Figure~\ref{fig:multistemsolver}D, but at a reduced scale.
}

\indexconcept{Single RNA molecule drawing}
Figure~\ref{fig:demo1iii} demonstrates two additional types of R2R drawings.
The {\tt .r2r\_meta} file is needed to specify both kinds of alternate drawing.
The first is drawing of a single molecule whose sequence is present in the alignment.
Every tenth nucleotide is numbered by default.
These kind drawings are called ``{\tt oneseq}'' drawings, because they depict one sequence.
A real-world example of this kind of drawing is given in Figure~\ref{fig:SAMIValternate}A.

\indexconcept{Skeleton drawing}
The second kind of drawing is a ``skeleton'' drawing, which is an outline of the shape.
Skeleton drawings can be applied to consensus or to individual sequences,
and are useful to give a compact summary of the whole structure.
Larger skeleton drawings are demonstrated in the files {\tt demo/exceptional/GOLLD.sto} (and {\tt .r2r\_meta}) and {\tt demo/exceptional/HEARO.sto} (and {\tt .r2r\_meta}).  Those files also demonstrate making scaled-down ``thumbnail'' skeleton drawings.
A more realistic example of a skeleton drawing is given in Figure~\ref{fig:SAMIValternate}B.

\indexconcept{\texttt{.r2r\_meta} file}
\indexcmd{SetDrawingParam}
\indexcmd{if\_skeleton}
\indexcmd{R2R\_oneseq}
Key points:
\begin{itemize}
\item Fields in the {\tt .r2r\_meta} file are separated by tab characters.
\item A single RNA molecule can be drawn using the {\tt oneseq} directive in the {\tt .r2r\_meta} file.
	\begin{itemize}
	\item For {\tt oneseq} drawings, nucleotides are removed from the drawing using the {\tt DEL\_COLS} sequence-specific line.
	\item R2R commands beginning with {\tt \#=GF R2R\_oneseq bigfoot} are only applied in {\tt oneseq} mode for the sequence {\tt bigfoot}.
	\end{itemize}
\item A skeleton-style schematic can be drawn by adding {\tt skeleton-with-pairbonds} (or {\tt skeleton}) in the {\tt .r2r\_meta} file.  Commands beginning {\tt if\_skeleton} are only applied in {\tt skeleton} mode.
\item Drawing parameters can be modified with the {\tt SetDrawingParam} command.  (This command can also be used in the {\tt .r2r\_meta} file.) A list of parameters that can be modified with {\tt SetDrawingParam} is provided in Section~\ref{SetDrawingParam}.
\item Lines beginning {\tt \#=GC R2R\_XLABEL\_}{\it something} can be used to introduce an extra dimension of labels with the name {\it something}.  See Section~\ref{sec:labels} for details.
\end{itemize}

\section{Common error: ``One or more pairs is getting broken by one side getting deleted''}
\fig{fig:demobreakpair}{figures/demo-breakpair.pdf}{demo-breakpair: common error}{
(A) input Stockholm file ({\tt demo/demo-breakpair.sto.err}).
Column numbers are labeled, with the left-most column having the number zero.  (This is the numbering system used by the text editor program ``Emacs''.)
(B) output of R2R with an error.  The referenced column numbers are circled in blue.
(C) one resolution: since the base pair is not well conserved, remove it from the {\tt SS\_cons} line.
(You should only do this if you truly believe, on reflection, that the pairing is not biological.)
(file {\tt demo/demo-breakpair-fix1.sto})
(D) typical resolution: the base pair is not that well conserved anyway, so don't show its nucleotides in the consensus diagram at all.  However, retain the pairing in the alignment, since the pair is sometimes formed.  The nucleotides in the pairing are removed from the drawing by putting dashes in the {\tt R2R\_LABEL} line.
(file {\tt demo/demo-breakpair-fix2.sto})
(E) another resolution: break the pair in R2R, making it a bulge.
(file {\tt demo/demo-breakpair-fix3.sto})
You can also make R2R apply this solution automatically to all otherwise-broken base pairs by adding {\tt \#=GF R2R SetDrawingParam autoBreakPairs true} to the file.  This is useful when you want to draw an RNA quickly.
(F) another resolution: keep both sides of the pair.  The nucleotide that would normally have been deleted is drawn with a circle that uses a gray line, rather than the normal black line.
(file {\tt demo/demo-breakpair-fix4.sto})
(G) The drawing resulting from part F.
}
\fig{fig:demobreakpairvarlen}{figures/demo-breakpair-varlen.pdf}{demo-breakpair 2: Another cause}{
(A) input Stockholm file ({\tt demo/demo-breakpair-varlen.sto.err}).
(B) output of R2R with an error.  The line number with the problematic command is circled in blue, and corresponds to the {\tt var\_backbone\_range} command.
}

\indexconcept{Error: one or more base pairs is getting broken...}
This sections explains how to handle a common error when using R2R.
R2R defines which nucleotide positions are shown in a consensus based on how frequently they are present in sequences.  Sometimes one nucleotide position involved in a base pair is removed from the consensus while the other is retained.  This can happen with stems that are loosely conserved.
When this arises, R2R will report an error (``one or more base pairs is getting broken...''), and you must choose how to resolve the problem.  Figure~\ref{fig:demobreakpair} gives an example of this problem, and shows 4 ways to resolve it.

\indexcmd{var\_backbone\_range}
Pairs can also be broken when {\tt var\_backbone\_range} commands refer to one side of a base pair.  In this case, R2R will detect that this command was responsible for the problem (Figure~\ref{fig:demobreakpairvarlen}).  If you have variable-length stems, you should use the {\tt var\_stem} or {\tt var\_hairpin} commands.

Key points:
\begin{itemize}
\item Pairs can be broken due to the consensus rules, and Figure~\ref{fig:demobreakpair} shows 4 ways to resolve the problem.
\item Pairs can also be broken due to variable-length backbones.
\item R2R error messages often refer to lines or columns in the input Stockholm-format file.  The left-most column is numbered zero.  The first line is numbered one.
\end{itemize}

\section[The place\_explicit command]{The {\tt place\_explicit} command}
\fig{fig:placeexplicit}{figures/place_explicit.pdf}{The {\tt place\_explicit command}}{
(A)
Schematic of the positioning system in the {\tt place\_explicit} command.
The diagram illustrates the effect of the command {\tt place\_explicit G A -45 5 -1 0 0 -90},
which defines how the nucleotide labeled G is positioned depending on the position of the nucleotide A.
The direction of the backbone at the A (gray array pointing right) is rotated -45 degrees to form the black arrow.
The position proceeds 5 drawing units along the black arrow, and -1 unit orthogonally to the black arrow.
This is where the G is positioned.
The direction of the backbone of the G (gray arrow pointing up) is taken by adding -90 to the direction of the backbone at the A.
The two zeroes mean that no absolute positioning is used.
One drawing unit is equal to the distance between consecutive nucleotides along the backbone.
(B)
Hypothetical Stockholm alignment to demonstrate {\tt place\_explicit} command (contents of the file {\tt demo/demo-pe.sto}).
Note that the {\tt f} at the end of the first and second {\tt place\_explicit} commands directs R2R to ``flip'' the backbone such that the $3^\prime$ 
side of a base pair is on the left of the $5^\prime$ side, instead of the usual right side.
The second flipping command flips the orienation back to the default for the $3^\prime$ tail of the hypothetical motif.
Synthesizing an appropriate joke based on the organisms from which the hypothetical RNA sequences were taken is left as an exercise for the reader.
(C)
Annotated raw output of R2R when run on the alignment in part B.
Nucleotide positions that were the referent of {\tt place\_explicit}
command (i.e., {\tt a-{}-}, {\tt b-{}-} and {\tt c-{}-}) have a blue arrow indicating the direction of the backbone at that position.
(D)
Raw R2R output of the mini-{\it ykkC} motif ({\tt demo/22/mini-ykkC.sto}) \cite{Weinberg22Motifs}, which uses a {\tt place\_explicit} command
to change the direction of the {\tt var\_backbone\_range} just before the SD sequence.
}

\indexcmd{place\_explicit}
R2R encodes defaults for the layout of RNAs that work in many cases, but often you will want to change this layout.
The {\tt place\_explicit} command allows you to position nucleotides relative to one another using
explicit relative or absolute coordinates (Figure~\ref{fig:placeexplicit}).
The syntax for the command is:
\example{
{\tt place\_explicit} {\it label relative-label place-angle x-relative y-relative x-absolute y-absolute next-angle}
}
(illustrated with an example in Figure~\ref{fig:placeexplicit}).
The nucleotide at position {\it label} is positioned relative to the nucleotide at position {\it relative-label}.
The coordinate system of ({\it x-relative},{\it y-relative}) is based on the direction of the nucleotide at {\it relative-label}
rotated by {\it place-angle} degrees.  
The x,y coordinates ({\it x-absolute},{\it y-absolute}) are added to this position.
Both relative and absolute distances in the {\tt place\_explicit} command are specified in drawing units, where 1 drawing unit is equal to the standard internucleotide distance between consecutive nucleotides (by default 0.105 inches).
The direction of the nucleotide at {\it label} is the angle at {\it relative-label} rotated
by {\it next-angle} degrees.

An optional {\tt f} at the end of the command directs R2R to ``flip'' the stem (Figure~\ref{fig:placeexplicit}B,C).  Flipping is analogous to applying a twist to the RNA.  Where R2R normally positions the $3^\prime$ nucleotide of a base pair to the right of its partner, flipped regions will cause the $3^\prime$ nucleotide to be on the left.
Flipping is used only sometimes, but Figure~\ref{fig:placeexplicit}C shows a typical example where it is useful to avoid crossing stems.  A real-world example of this usage is the IMES-4 motif \cite{CoolRNAs} (see {\tt demo/exceptional/IMES-4.sto}).

Section~\ref{sec:placeexplicit} explains the command in more detail.

More advanced usages of {\tt place\_explicit} are necessary for the ``in-line'' layout of complex pseudoknots
(Section~\ref{sec:inlinepseudoknot}).  I explain more about {\tt place\_explicit} commands and their semantics in that section.
You can use set the {\tt showPlaceExplicit} variable to {\tt true}, in order to see the {\tt place\_explicit} commands
in the context of the structure.
This is demonstrated in the same section (Section~\ref{sec:inlinepseudoknot}).

\section{Multistem junctions and turning internal loops}
\label{sec:demomultistem}

Multistem junctions are a structural element within a RNA secondary structure in which a loop
includes more than two base pairs (e.g., see Figure~\ref{fig:multistemdefault}).
This section can be skipped if you are drawing RNAs that don't have multistem junctions.
However, the material in this section is also sometimes useful for turning the direction of the stem at internal loops (see Figure~\ref{fig:internalloop} for an example of an internal loop).

There are three ways to draw multi-stem junctions with R2R:
\begin{itemize}
\item The junction is laid out perfectly on a circle, and stems are allowed to go in arbitrary directions.
This is the default.  It usually looks acceptable (except when nucleotides clash), 
although I think it's typically not the most aesthetic solution.
\item Manual layout using {\tt place\_explicit} commands.
\item Automated solutions that attempt to position the junction on a circle as well as possible, but subject to
constraints on the directions of the stems.  This allows for orthogonal stems, a design feature that I think tends to result in more attractive layouts.
\end{itemize}

\subsection{Simple circular layout}
\label{sec:circlemultistem}
\fig{fig:multistemdefault}{figures/multistem-default.pdf}{Simple circular layout of multistem junctions}{
(A)
Default layout of a hypothetical three-stem junction that is modeled on a multistem junction in HEARO RNA \cite{CoolRNAs}.
This output is based on the files {\tt demo/demo-multistem.sto} and {\tt demo/demo-multistem.r2r\_meta}.
(B)
In the default layout, nucleotides along the multistem junction are positioned on a circle. The R2R command shown here directs R2R to draw this circle.
The command refers to position ``{\tt M}''. Position {\tt M} is labeled, and is the left nucleotide
of the base pair that encloses the whole multistem junction.
(Note: the command should be one line, but is displayed on multiple lines in this figure so it fits the page.)
(C)
Stems can be flipped so that they are directed into the multistem junction.
This can save space when the multistem loop is larger than one of the stems.
The stems are numbered starting at zero.  The enclosing stem does not have a number, and cannot be flipped.
(D)
The SAM-IV riboswitch \cite{SAMIV} is drawn using the default layout of its multistem junction.
(Note: the gray lines correspond to pseudoknots.  Drawing of pseudoknots are explained in a later section of this tutorial.)
}
The default circular layout for multistem junctions is demonstrated in Figure~\ref{fig:multistemdefault}.
This is the layout that R2R uses if you do not give R2R any instructions on how to draw the multistem junction.

\subsection{Manual layout of multistem junctions}
\fig{fig:multistemmanual}{figures/multistem-manual.pdf}{Manual layout of multistem junctions}{
(A)
The published layout of SAM-IV riboswitches \cite{SAMIV}.  Layout of the multistem junction uses the
{\tt multistem\_junction\_bulgey} command.
This drawing is based on the file {\tt demo/SAM-IV/SAM-IV.sto}, in {\tt multistem=original} mode.  This mode is entered in the {\tt .r2r\_meta} file using the specification {\tt define multistem original}.
(Note: at the time at which SAM-IV riboswitches were drawn, manual layout was the only option for drawing multistem junctions.
Also, the gray lines are for pseudoknots, as described in Section~\ref{sec:calloutpseudoknot}.)
(B)
R2R command to draw the multistem junction, and larger view of the multistem junction.
The junctions called ``{\tt J0}'' and ``{\tt J1}'' are labeled.  There are no nucleotides
within junction {\tt J1}.
}
Manual layout of multistem junctions can be accomplished by positioning
the stems using {\tt place\_explicit} commands, and directing the layout of the single-stranded
regions using the {\tt bulge} command.  Since the use of these commands would result in many labels to identify the various positions,
R2R has the {\tt multistem\_junction\_bulgey} command, for which only one label needs to be defined.

In the {\tt multistem\_junction\_bulgey} command, the stems within the multistem junction are positioned
using functionality similar to {\tt place\_explicit} commands.  The junctions are then positioned automatically
between the stems by positioning them on circles (equivalent to the {\tt bulge} command).
Figure~\ref{fig:multistemmanual} shows an example.

Junctions are identified using the symbols {\tt J0}, {\tt J1}, {\tt J2}, etc.
Junction {\tt J0} identifies the junction between the enclosing stem and the next $5^\prime$-most stem.
Junction {\tt J1} identifies the junction between this next $5^\prime$-most stem, and the next stem to that (i.e.,
the immediately $3^\prime$ stem).
Positioning of {\tt J0} is equivalent to a {\tt place\_explicit} command that positions X relative to Y,
where X is the left nucleotide of the stem on the $3^\prime$ side of the junction {\tt J0} and Y is the 
left base-paired nucleotide of the stem on the $5^\prime$ side of the junction {\tt J0}.
For junctions {\tt J1}, {\tt J2}, etc., Y is the {\em right} nucleotide of the stem on the $5^\prime$ side of the junction.
An equivalent viewpoint is that X and Y are the base-paired nucleotides immediately $3^\prime$ (X) and $5^\prime$ (Y)
to the nucleotides within the junction.
This is also illustrated in Figure~\ref{fig:multistemmanual}.

The {\tt multistem\_junction\_bulgey} command also provides alternate layouts for the bulges, such as flipping the direction
of the bulge, and some linear layouts for the junctions.
It is also possible to position the $5^\prime$ base-paired nucleotides in all stems relative to the enclosing stem,
rather than to the previous stem by using {\tt J1/base} in place of {\tt J1}.
These variations are described in Section~\ref{sec:multistemjunctionbulgey}.

\subsection{Automated multistem junction layout with direction constraints}
\label{sec:automultistem}
\fig{fig:multistemsolver}{figures/multistem-solver.pdf}{Automated layout of multistem junctions}{
(A)
Layout of the demo structure used in previous figures, using the solver.
The solver command is given below the drawing.  
The lower case ``{\tt m}'' parameters say that the circle should intersect
in the midpoint ({\tt m}) of stems.
Note that the length of the variable-length region (marked as ``0-3 nt'') is adjusted by the solver
to optimize the circular layout.
(The drawing is based on the file {\tt demo/demo-multistem.sto}, with {\tt solver1=1}.)
(B)
Illustration of stem numbers and directions from the diagram of part A.
The enclosing stem is {\tt s0}, then the stems are number {\tt s1}, {\tt s2} from $5^\prime$ to $3^\prime$ around the multistem junction.
The stem {\tt s0} defines the angle zero, as indicated.  As usual within R2R, positive angles are in the clockwise direction, and negative angles are counter-clockwise.
Thus, {\tt s1} is oriented in the direction {\tt -90} (going to the left), while {\tt s2} is in the direction {\tt 0} (up the page).
If the enclosing stem (stem {\tt s0}) were rotated, the angles of stems {\tt s1} and {\tt s2} would also rotate to be the same,
relative to the enclosing stem.
The angle of the enclosing stem (stem {\tt s0}) can be changed, which has the effect of rotating the entire
picture.  This is purely for convenience, and is illustrated in Figure~\ref{fig:multistemsolvermanA}.
(C)
A variation on the drawing in part A.
The {\tt ai=0} parameters directs R2R to automatically decide on the intersection point with the circle for each
stem.  This leads to a circle that better goes through each nucleotide.
(D)
Drawing of the SAM-IV riboswitch \cite{SAMIV} using the solver.
In this case, the bulged {\tt Y} within stem {\tt s1} is very
close to the {\tt A} in the junction between stems {\tt s0} and {\tt s1}.
The positions of these nucleotides could be adjusted in Adobe Illustrator of Inkscape.
Alternately, the stem {\tt s1} could be rotated to avoid this problem,
as demonstrated in the next sub-figure.
(E)
Drawing of the SAM-IV riboswitch with an alternate angle for stem {\tt s1}.
}
\fig{fig:multistemsolvermanA}{figures/multistem-solver-manA.pdf}{Automated layout of multistem junctions: unusual stem directions}{
This figure illustrates the strategy of using the default circular layout to suggest aesthetic stem directions.
In this case, I was not able to find directions that are multiples of 90 or even 45 that worked well.
This example is based on a multistem junction in the {\it manA} motif \cite{Weinberg104Motifs} (see {\tt demo/104/manA.sto} for the full motif).
(A)
Default circular layout of this motif.
(This drawing is derived from the file
{\tt demo/demo-multistem-manA.sto}.)
(B)
Layout using solver, with stem directions that roughly match the default orientations in part A.  Stem {\tt s0} is rotated 30 degrees clockwise in this figure.
The directions of stems {\tt s0} and {\tt s4} are symmetric about the vertical axis, which results in
an aesthetic design.
(Note: R2R does not draw the $5^\prime$ marking since the $5^\prime$ nucleotide's direction is not orthogonal to the X or Y axis.)
(C)
Alignment constraints are added to the layout of part B.
{\tt align\_angle -30 0 4} dictates that the stems {\tt s0} and {\tt s4} should be vertically
aligned. (The alignment angle {\tt -30} is used because the enclosing stem is
rotated by 30 degrees; see part D.)
The constraint {\tt align\_nuc\_centers\_angle 60 align:X . align:Y .} forces
the midpoint of the nucleotides defined by {\tt align:X} to be horizontally
aligned with the midpoint of the {\tt align:Y} nucleotides.
This alignment further improves the symmetry.
Note that R2R cannot aesthetically satisfy any arbitrary constraint,
and so stems {\tt s1} and {\tt s3} cannot be vertically aligned in this layout.
%Also, the alternate solver {\tt multistem\_junction\_circular\_solver} does not
%achieve this layout, and indeed cannot provide a feasible layout without help
%(in the form of the {\tt try\_harder} directive, which directs the solver to
%try multiple initial points in an attempt to converge on a better solution).
%It is not always clear which solver will work best.
(D)
The drawing of part C is annotated.  The nucleotides that were aligned are drawn in
alternate colors as shown.  The {\tt align:X} nucleotides were vertically aligned
with each other when their stems ({\tt s0} and {\tt s4}) were aligned.
Zero degrees is defined as the direction of stem {\tt s0} (as shown), so the
vertical direction is {\tt -30} degrees.
The solver directives {\tt draw\_circ}
and {\tt draw\_zero\_degrees} were added to draw the multistem junction circle,
and the zero degrees mark, and {\tt nuc\_color} was used to color
the {\tt align:X} and {\tt align:Y} nucleotides.
}
\fig{fig:multistemsolverIMESone}{figures/multistem-solver-IMES-1.pdf}{Automated layout of multistem junctions: additional example}{
This figure illustrates an additional example of a multistem junction drawing, based on a multistem junction
found in IMES-1 RNAs \cite{CoolRNAs}.
(The drawing is based on the file {\tt demo/demo-multistem-IMES-1.sto}.)
(A)
Layout of the given command.
(B)
Layout a different {\tt \_solver} command, as shown.
}
\fig{fig:multistemsolverHEAROtwo}{figures/multistem-solver-HEARO2.pdf}{Automated layout of multistem junctions: starting point is important}{
This drawing is based on a multistem junction in HEARO RNA \cite{CoolRNAs} that is different from the earlier multistem junction shown.
(The drawing is based on the file {\tt demo/demo-multistem-HEARO2.sto}.)
(A)
Layout of the given command, with {\tt ai} parameters using the default midpoint as the initial point for the optimization search.
(B)
Layout of the given command, with {\tt ai=0} and {\tt ai=1} specifying alternate starting points.
The resulting layout is slightly different, and I think a bit better.
}
R2R provides the ability to automatically determine a layout of a
multistem junction, given user-specified directions in which each
stem must go.
This functionality requires that CFSQP is available (see
Chapter~\ref{sec:install}).

This automated layout facility is refered to as the {\tt \_solver} commands
in this manual, and is applied with one of the following three
related commands:
\begin{itemize}
\item {\tt multistem\_junction\_circular\_solver}
\item {\tt multistem\_junction\_bulgecircley\_solver}
\item {\tt multistem\_junction\_bulgecircleynormals\_solver}
\end{itemize}
All {\tt \_solver} variants use a very similar syntax and framework.
They are explained in full detail in Section~\ref{sec:multistemjunctionsolver},
but for now we will see the most essential aspects of the commands.

Examples of the {\tt \_solver} commands are given in Figures~\ref{fig:multistemsolver}, \ref{fig:multistemsolvermanA}, \ref{fig:multistemsolverIMESone} and \ref{fig:multistemsolverHEAROtwo}.
The command begins with a {\tt \_solver} command, then specifies a label.
As with previous multistem junction commands, the label denotes the left nucleotide in the base pair
that encloses the multistem junction.
Then, then properties of each stem in the multistem junction are written.
The enclosing stem is named {\tt s0}, then each stem from $5^\prime$ to $3^\prime$ are labeled
{\tt s1}, {\tt s2}, ... (see Figure~\ref{fig:multistemsolver}).
The names for the stems are different from those of multistem junction commands described earlier in this tutorial.  In particular, the enclosing stem can be explicitly referenced (as {\tt s0}).

For each stem, the desired angle of that stem is given, in a coordinate system defined
by the enclosing stem (Figure~\ref{fig:multistemsolver}B).
The point at which the circle should intersect the inner base pair of each stem is also specified.
Useful values are {\tt m} (intersects the {\tt m}idpoint of the base pair),
{\tt l} (intersects the center of the {\tt l}eft nucleotide in the base pair),
{\tt r} (intersects the {\tt r}ight nucleotide in the base pair)
and {\tt ai} ({\tt a}utomatically solve the {\tt i}ntersection).

To find a good layout of a particular multistem junction, it is often
necessary to play with several formulations,
involving different stem angles and solvers.
One strategy to finding good directions for the stems is to use the default circular positioning of the multistem junction
to find directions that fit the circle perfectly, then consider nearby angles that are multiples of 90 degrees, or have
other pleasing characteristics.
See Figure~\ref{fig:multistemsolvermanA} for an example.
As noted above, the starting point for the circle-intersection point of each base pair can sometimes
have an effect on the solution.

Variable-length backbones (introduced using the {\tt var\_backbone\_range} command) can help to improve aesthetics of drawings, as the solver is permitted
to change the length of the lines to improve the circularity of the layout.  This additional degree of freedom can be helpful.
Of course, some multistem junctions do not have variable-length regions, so this tactic cannot always be used.

Additional examples illustrating aspect of the {\tt \_solver} commands are shown in Figures~\ref{fig:multistemsolverIMESone} and \ref{fig:multistemsolverHEAROtwo}.


Note: the starting point for the intersection can be supplied for automatic inference,
and this sometimes assists the solution, as the problem that the computer has to solve is
highly non-linear, and the computer can easily get stuck in local minima.
See Section~\ref{sec:multistemjunctionsolver} for details on various initial
values that can be set.
For example, {\tt ai=0} means start with the intersection at the left nucleotide and {\tt ai=1} means start at the
right nucleotide of the base pair.  Any number from 0 to 1 can be given, so {\tt ai=0.5} means start at the midpoint.  
The circle intersection parameter is only relevant to {\tt multistem\_junction\_circular\_solver}; the {\tt multistem\_junction\_bulgecircley...} commands try to have both
nucleotides on the circle.  However, other parameters are relevant to the
{\tt bulgecircley} variant commands.
You can also add {\tt try\_harder} to the {\tt multistem\_junction\_circular\_solver} command
to improve it's ability to find a good global optimum (details in 
Section~\ref{sec:multistemjunctionsolver}).

\subsubsection{Solving multistem junctions can be slow}

On even modern machines, the computer can take a significant amount of time to solve the problem.
Often the solver goes through a period of very gradually improving the solution, and many iterations are required to reach an optimal solution---and the difference in the ultimate solution is noticeable.
Therefore, I have conservatively set the solver to use up to 100,000 iterations.
You can decrease this number with an R2R command like {\tt SetDrawingParam solverMaxIters 50000}.

I have also implemented a ``caching'' strategy, in which solutions to problems solved by CFSQP are store in a file that is associated with the {\tt .r2r\_meta} input file.
These files end in the extention {\tt .solver-cache}.
If you re-run R2R, but do not change anything related to the multistem junction, the solution will be re-used.
By deleting the {\tt .solver-cache} file, you can force a re-computation.

Note: R2R is designed to detect subtle changes that can change the problem, and R2R will
re-compute the solution in these cases.  For example, if you change the value of
{\tt internucleotideLen} using the {\tt SetDrawingParam} command, this changes a parameter
in the multistem junction Non-Linear Program, even though it is an indirect effect.
\comment{
Another way of improving speed is to define the {\tt R2R\_DUMP\_FILE} environment variable in your UNIX shell before running {\tt r2r}.
This is done by the default {\tt Makefile} in the {\tt demo} subdirectory.
This environment variable specifies the name of a file, and in the default {\tt Makefile}, the file is {\tt dump-file.txt}.
R2R will output {\tt multistem\_junction\_bulgey} commands into this file that are equivalent to {\tt \_solver} command.
You can substitute these commands to generate the same layout without the CPU time.
}

\subsection{Internal loops}
\fig{fig:internalloop}{figures/turnsteminternal.pdf}{An internal loop with a 90-degree turn}{
This drawing is based on SAM-IV riboswitches \cite{SAMIV,Weinberg22Motifs}, in the file
{\tt demo/22/SAM-IV.sto}.
(A)
An R2R command to turn an internal loop in SAM-IV riboswitches.
The label {\tt K} refers to the $5^\prime$-most nucleotide within the $5^\prime$ part of the internal loop, as shown.
(You can also use the $3^\prime$-most part of the $3^\prime$ part of the internal loop.)
The stem can be turned 90 degrees clockwise or counter-clockwise.  The value {\tt +1} (as shown) is clockwise, while {\tt -1} would be counter-clockwise.
R2R will decide the layout on one side of the internal loop.  The value {\tt r} will make R2R use the {\tt r}ight side (i.e. $3^\prime$ side) for positioning,
while {\tt l} is for the {\tt l}eft side.
(B)
The turned internal loop in the context of the full SAM-IV consensus.
(C-E)
Turning the internal loop using {\tt \_solver} commands.
Since the {\tt \_solver} commands require the label to be the left nucleotide of the enclosing base pair of the internal loop, we use the label {\tt K--}.
}
Internal loops and bulges are unpaired regions that interrupt stems.
It is often desirable to turn the direction of the stem
at these loops for two main reasons.
First, bulges and highly asymmetric internal loops can have a distorted look due to having too many nucleotides on
one side.
A turn can give more space to the longer side, improving the clarity and aesthetic quality of the drawing.
Second, turning of stems can resolve problems with other regions of the structure that might overlap, use space inefficiently,
or otherwise be undesirable.

R2R has a special command for turning the direction of the stem at internal loops or bulges.
This command restricts turning to +90 or -90 degrees, however.
The command, {\tt turn\_stem\_at\_internal}, is demonstrated in Figure~\ref{fig:internalloop},
and documented in detail in Section~\ref{sec:turnsteminternal}.

The {\tt turn\_stem\_at\_internal} command is limited both in the angles it can handle,
and in the layout policy.
You can use the {\tt \_solver} commands that are designed for multistem junctions on internal loops or bulges.
An internal loop is simply a multistem junction with only 2 stems, for the purposes of these commands.
The application of {\tt \_solver} commands to internal loops is demonstrated in Figure~\ref{fig:internalloop}.

Note that R2R does not distinguish between internal loops and bulges, as RNA biologists often do.  For R2R, a bulge is simply an internal loop with zero nucleotides on one side.

\section{Modular structures}
\label{sec:modular}
\fig{fig:modular}{figures/modular-demo.pdf}{Modular structures}{
(A)
The Stockholm alignment and R2R drawing commands for a hypothetical modular structure.
This is the file {\tt demo/demo-modular.sto}.
(B)
Raw output of R2R.
To extract the Stockholm file for the predicate {\tt GNRA}, for example, the 
command {\tt perl ../src/SelectSubFamilyFromStockholm.pl intermediate/demo-modular.cons.sto GNRA > intermediate/demo-modular-GNRA.sto} was used, followed by inference of the consensus using {\tt ../src/r2r -{}-GSC-weighted-consensus intermediate/demo-modular-GNRA.sto intermediate/demo-modular-GNRA.cons.sto 3 0.97 0.9 0.75 4 0.97 0.9 0.75 0.5 0.1}.
The file {\tt intermediate/demo-modular-GNRA.cons.sto} was included in {\tt demo-modular.r2r\_meta}.  Similar commands were run for the 
{\tt UNCG} and {\tt OPT} predicates.
(C)
Integration of the modular structures into a consensus diagram.  This drawing was created in Adobe Illustrator using the output of R2R shown in part B.
}
I define modular structures as sub-structures of an RNA that are sometimes, but not always, present in a conserved RNA structure.
Many RNAs, for example, have stems that are sometimes present.
To draw a consensus, it is often desirable to reflect these modular structures, and their relative frequencies.
As with other frequencies, R2R weights sequences using the GSC algorithm, and calculates the weighted frequency of the set
of sequences that do have the modular structure.
(See Chapter~\ref{consensus} for details.)

Figure~\ref{fig:modular} demonstrates the drawing of two modular substructures in the context of a simple, hypothetical motif.
It covers hairpins that are sometimes, but not always, present. It also covers a loop that has distinct structures that are
possible, namely the stable GNRA or UNCG tetraloops.
The figure legend gives the commands that would be run to process files with modular structures.

The basic idea for processing modular structures in R2R is 
(1) to define a predicate that determines which sequences exhibit a given substructure (e.g., which sequences have the
optional hairpin),
(2) extract those sequences into a new {\tt .sto} file using a Perl script that is part of R2R
and (3) draw the new file.
Predicates can be defined using either regular expressions or using Boolean operations using a subset of the Perl programming language.

Regular expressions are a mathematical formalism for describing sequence patterns \cite{EddyBook}.
Some examples are given in Figure~\ref{fig:modular}.
To implement these regular expressions, I used Perl, and consequently R2R uses the Perl syntax to specify regular expressions.
To learn about all the capabilities of regular expressions in Perl, one good Web site is \url{http://www.perl.com/doc/manual/html/pod/perlre.html}.

Regular expression predicates operate on subsequences in columns defined by lines of the form {\tt \#=GC SUBFAM\_LABEL\_...}.
For example, {\tt \#=GC SUBFAM\_LABEL\_TERM} would define columns that can now be referred to as ``{\tt TERM}''.

Here are some examples of regular expressions that illustrate features relevant to RNA.
For four columns with no gaps, {\tt G[A-Z][AG]A} defines the GNRA tetraloop.
The text in square brackets ({\tt [} and {\tt ]}) define a set of symbols.
{\tt [A-Z]} is the set of upper-case letters, while {\tt [AG]} defines the symbols
{\tt A} or {\tt G}.
A period (``{\tt .}'') is a special symbol that stands for any character, so in order to look for a gap character
you should enter {\tt [-.]}.

A few other special symbols are often relevant.
The caret symbol ``{\tt \verb|^|}'' (shift-6 on a standard US keyboard) stands for the beginning of a subsequence, while the dollar sign ``{\tt \$}'' represents the end
of a subsequence.
The asterisk ``{\tt *}'' represents zero or more occurrences of the previous symbol.
Thus, the expression {\tt \verb|^|[-.]*\$} matches all subsequences that consists only of gap characters.

R2R also provides Boolean operations that can combine the values of other predicates.  These predicates also exploit Perl, and therefore inherit a Perl-like syntax.
In these predicates, the expression {\tt \$predValue\{GNRA\}}, for example, would evaluate to true if the predicate
named {\tt GNRA} is true for a given sequence.
Two ampersands ({\tt \&\&}) represents the Boolean ``and'' operation, which evaluates to true if the two sub-expressions
are true.  So, {\tt \$predValue\{GNRA\} \&\& \$predValue\{FOUR\}} is true if the predicates {\tt GNRA} and {\tt FOUR}
are both true.  Similarly, two pipe symbols ({\tt $||$}) denote the Boolean ``or'' operation, which evaluates to the true
if {\em either} of the sub-expressions are true.
The exclamation mark {\tt !} is the Boolean not operation, so the expression {\tt !\$predValue\{GNRA\}} represents all sequences
that do {\em not} match the {\tt GNRA} predicate.
Warning: if you use Boolean predicates, you must define predicates they use within {\tt \$predValue} earlier in the {\tt .sto} file,
since R2R will evaluate predicates in the order in which they appear in the {\tt .sto} file.

The script {\tt SelectSubFamilyFromStockholm.pl} evaluates predicates and creates a new {\tt .sto} file with the subset
of sequences.  It also modifies which commands are present in the file, in order to use commands that are appropriate
to drawing the modular structure.
All lines beginning with {\#=GF R2R} or {\#=GC R2R} are by default removed.
At the same time, for example with the predicate {\tt GNRA}, lines beginning {\#=GF SUBFAM\_GNRA\_R2R}
are changed to {\#=GF R2R} so that they now become processed.
Similarly, lines beginning {\tt \#=GC SUBFAM\_GNRA\_R2R...} are changed to {\tt \#=GC R2R...}.

Chapter~\ref{chpt:modular} gives further details on the commands.

\comment{
In this section, I will demonstrate the drawing of modular structures using c4 antisense RNA \cite{c4,Weinberg104Motifs}.
A number of terminal loops of hairpins have been demonstrated that are especially stable.
The GNRA tetraloop is the best known of these, but UNCG and CUNG are also established as stable terminal loops.
The terminal loop of the P2 hairpin in the c4 antisense RNA often conforms to one of these stable tetraloops, 
and I want to show how frequent
each of them were.
}

\section{Pseudoknots}

Pseudoknotted RNAs typically lead to confusing drawings when drawn with the default layout of R2R.
Therefore, to help you to plan the final layout better,
the first thing to do with an RNA structure that has a pseudoknot is usually to make R2R ignore the pseudoknot,
which can be accomplished with the {\tt ignore\_ss\_except\_for\_pairs} command.
I give an example of this command in the next section, since it leads naturally to a ``callout'' style of pseudoknot drawing.

\subsection{A note on the representation of pseudoknotted secondary structures within Stockholm files}

R2R uses a non-standard method for writing consensus secondary structures for motifs with pseudoknots.
To describe pseudoknots, in addition to the standard {\tt \#=GC SS\_cons} line, there can be other lines
of the form {\tt \#=GC SS\_cons}{\it X} for any string {\it X} (typically, I call them
{\tt \#=GC SS\_cons\_1}, {\tt \#=GC SS\_cons\_2}, ...).  
In the pseudoknot examples in this section, you will see {\tt \#=GC SS\_cons\_1} used.
GOLLD RNA \cite{CoolRNAs} has 5 pseudoknots (see {\tt demo/exceptional/GOLLD.sto}), and uses up to {\tt \#=GC SS\_cons\_5}.
When you use extra {\tt \#=GC SS\_cons\_...} lines, I recommend that they each only contain only one stem (possibly with bulges/internal loops), and
not more complex structures.  R2R is best tested with this convention.

I find this representation of pseudoknots within Stockholm files to be convenient, because all {\tt \#=GC SS\_cons} lines are interpreted in the same way.

\subsection{Pseudoknots drawn with callouts}
\label{sec:calloutpseudoknot}
\fig{fig:calloutpseudoknot}{figures/demo-pknot-callout.pdf}{A pseudoknotted RNA drawn in  callout style}{
(A)
The Stockholm file {\tt demo/demo-pknot-callout.sto}.
The structure of this hypothetical RNA is inspired by that of SAM-V riboswitches \cite{SAMV}, but is simplified for this example.
(B)
The raw output of R2R when run on {\tt demo/demo-pknot-callout.r2r\_meta} / {\tt demo/demo-pknot-callout.sto}.
(C)
Assembled drawing.  To create this drawing, the components of part B were combined using Adobe Illustrator.
}
Figure~\ref{fig:calloutpseudoknot} gives an example of an RNA with a pseudoknot that is drawn using a callout style.  This style depicts the pseudoknot pairing in a ``callout'' (Figure~\ref{fig:calloutpseudoknot}C) that also allows R2R to shade covarying base pairs.  To achieve this, we consider the pseudoknot as a modular structure, and make R2R draw it separately.
Modular structures were explained in Section~\ref{sec:modular}.  Their application to pseudoknots is somewhat simpler, since all sequences presumably have the pseudoknot, so there is no need to define any regular expressions.

The line {\tt \#=GF SUBFAM\_PERL\_PRED pknot return 1;} (e.g., see Figure~\ref{fig:calloutpseudoknot}A) defines the {\tt pknot} ``modular'' structure as applying to all sequences.
R2R commands and {\tt \#=GC} lines beginning with {\tt SUBFAM\_pknot\_} then define commands that should be run to generate the {\tt pknot} structure.  The actual drawing is initiated by a line in the {\tt .r2r\_meta} file, and running extra commands to create the pseudoknot view of the RNA structure, as in Section~\ref{sec:modular}.

It is also possible to draw callout-style pseudoknotted structures without the use of the modular structure logic, but instead using the {\tt define} and {\tt ifdef} commands to draw the main RNA and separately draw the pseudoknot.  The associated files {\tt demo/demo-pknot-callout-ifdef.sto} and {\tt demo/demo-pknot-callout-ifdef.r2r\_meta} demonstrate how to do this.

Once R2R has drawn the pseudoknot and the main structure, these need to be assembled in a drawing program like Adobe Illustrator or Inkscape.  One style of this assembly is shown in Figure~\ref{fig:calloutpseudoknot}C.

Note: R2R will attempt to automatically join the lines or arcs that comprise an outline corresponding to a pseudoknot.  However, sometimes its heuristics will fail, and it might be missing some lines, i.e., some nucleotides will not be outlined.  In this case, try disabling the automatic joining behavior:
\examplett{\#=GF SetDrawingParam outlineAutoJoin false}
This case arises with the first pseudoknot in SAM-IV riboswitches \cite{SAMIV}, for which I disabled automatic joining (see {\tt demo/SAM-IV/SAM-IV.sto})

\subsection{Pseudoknots drawn with in-line style}
\label{sec:inlinepseudoknot}
\fig{fig:inlinepseudoknot}{figures/demo-pknot-inline.pdf}{A pseudoknotted RNA drawn in inline style}{
The same RNA structure depicted in Figure~\ref{fig:calloutpseudoknot} is now drawn
with an in-line-style pseudoknot.
(A)
R2R's default drawing of the structure.  R2R does not attempt to make a feasible
in-line-style drawing of a pseudoknot structure, since this is quite difficult, and sometimes not possible
within R2R's quality constraints.
(B)
Stockholm input file.
This is the contents of {\tt demo/demo-pknot-inline.sto}.
(Some commands were removed, but these lines are not necessary to generate the drawing in part C.)
(C)
Raw output of R2R when run on the Stockholm file in part B.
(D)
Output of R2R when run using the command {\tt R2R SetDrawingParam showPlaceExplicit true}, which marks where {\tt place\_explicit}
commands were used.  The boxed region is shown at a larger scale in the next subfigure.
(E)
Expansion of the region boxed in part D.  The pink arrow labeled `11' on the top left corner refers to a {\tt place\_explicit} command
on line 11 of the input {\tt .sto} file ({\tt demo/demo-pknot-inline.sto}).  The pink line has two arrowheads.  The thicker one (pointing up and right)
indicates the direction in which layout actually proceeded, although it is not usually necessary to know this.
The gray arrow on the left labeled `def' shows a default rule that R2R used to position
these nucleotides.  In this case, the default rule says that a helix should be positioned
in a linear arrangement.
These arrows are sometimes useful to diagnose problems with {\tt place\_explicit} commands.
The arrows and their labels are intentionally small so that they clearly indicate the relevant nucleotides in
an automated drawing.  Most PDF and SVG viewers allow zooming in, to expand the scale of these features.
}
Sometimes the geometry of a particular RNA motif allows the pseudoknot base pairs to be drawn without the use of a callout.  In these cases, it is often desirable to use an ``in-line'' drawing of the pseudoknots, as shown in Figure~\ref{fig:inlinepseudoknot}.
However, even if you want to draw an in-line style, I recommend starting with a basic callout-style drawing, because it's easier to understand the RNA and decide on a final layout.

Internal loops, bulges, terminal loops and multistem junctions are by default drawn along a circle.  However, adding {\tt place\_explicit} commands that refer to positions within these loop regions will cause them to be drawn in a straight line.  In cases where you want a part of the original loop to be drawn along a circle, you can use the {\tt bulge} or {\tt bulge\_flip} commands.

In order to achieve an in-line pseudoknot layout, it is often necessary to flip stems, using the {\tt f} option to the {\tt place\_explicit} command.
Sometimes it can be confusing to get the proper stems flipped.
It is useful to note that R2R has default rules for layout.
When you use a {\tt place\_explicit} command, your command will override default rules.  (Specifically it will override any {\em conflicting} default rules.)
To diagnose problems with flipped stems, try using the {\tt mark\_flip} R2R command, which shows you in the drawing whether each element is flipped or not.  Another strategy is to simply remove the {\tt f} marking from {\tt place\_explicit} commands that you might think should be flipped, and see if that helps. 
Finally, you can use the command {\tt SetDrawingParam showPlaceExplicit true} to show you how your {\tt place\_explicit} commands were used in the positioning, and where default rules were used (which you might want to override with additional {\tt place\_explicit} commands).  The {\tt showPlaceExplicit} option is demonstrated in Figure~\ref{fig:inlinepseudoknot}.

Examples of R2R-based RNA drawings with in-line pseudoknots are as follows.  Most of these use flipped stems.
\begin{itemize}
\item SAH \cite{SAH,Weinberg22Motifs} (see {\tt demo/SAH/SAH.sto})
\item PreQ$_1$-II \cite{Weinberg22Motifs} (see {\tt demo/22/preQ1-II.sto})
\item The {\it pfl} motif \cite{Weinberg104Motifs} (see {\tt demo/104/pfl.sto})
\item The {\it ykkC}-III motif \cite{Weinberg104Motifs} (see {\tt demo/104/ykkC-III.sto}) 
\item The Downstream-peptide motif \cite{Weinberg104Motifs} (see {\tt demo/104/Downstream-peptide.sto})
\end{itemize}.

\section{Preparing presentations using projectors}

This section concerns the default colors for shading base pairs, which indicate covarying mutations.
These colors were selected with journal papers in mind.  The colors are clear in print or on a computer monitor,
yet are not too distracting.
However, these colors are too light for many projectors.  Therefore, if you're doing a talk and want the audience to be able to
see the shading, I recommend darker colors.

To modify the colors in an R2R drawing, you can add the following R2R command:
\examplett{SetDrawingParam pairShadingColors rgb:156,199,153 rgb:152,199,222 rgb:235,138,126}
(thanks to Kirsten F. Block for these RGB color values).

Or you can change the colors of an existing drawing in a drawing program.  For example, Adobe Illustrator allows you to select all object that
have the same fill color as a selected object.  So, you can select one covarying (pale green) base pair, and then get all the others,
to change this color.

\section{How to draw GOLLD RNA from R2R output}
\label{sec:GOLLD}
As stated previously the published drawing of GOLLD RNA \cite{CoolRNAs}
was based on the most commonly observed structural elements.
Creating the drawing based on the output of R2R therefore requires some assembly.
The drawing is generated from the demo files
{\tt demo/exceptional/GOLLD.sto} and {\tt demo/exceptional/GOLLD.r2r\_meta},
and this description refers to the PDF or SVG output generated when R2R is run on those files.
Most elements of the drawing come from the default drawing, called ``GOLLD''.
However, the second multistem junction (numbered from the $5^\prime$ end)
is taken from the ``GOLLD-skipbadd2'' drawing.
The third and fourth domains are taken from the ``GOLLD-d3classic'' drawing.
All other elements of the structure come from the default drawing.

\section{Making the demo output files}
\label{sec:MakeDemos}

To create the demos yourself, {\tt cd} into the {\tt demo} subdirectory.
To create the PDF output in the {\tt output-pdf} directory, run {\tt make all-pdf}.
Similarly, to create SVG, run {\tt make all-svg}.  I recommend that you {\em not} run plain {\tt make} or {\tt make all}, because of unlikely but possible conflicts with the {\tt .solver-cache} file, as the same file is used both for PDF and for SVG generation.
You might need to delete the files in {\tt output-pdf} or {\tt output-svg} in order to force the {\tt make} command to re-build them.

If you wish to really make the files from scratch, also delete the files in the {\tt intermediate} subdirectory and {\tt .solver-cache} files in various subdirectories, to force CFSQP to run.  (Of course, this requires that you have CFSQP.)
Note that this process will take a considerable amount of time (possibly hours), because several RNAs use the solver, and the solver sometimes requires significant amounts of time to obtain optimal solutions.
(Tip: if you have GNU make, and have a multiprocessor machine, you can use the {\tt j} flag.  For example, if you have 4 CPUs and want to make PDF files, run {\tt make -j 4 all-pdf}.
You can determine if you have GNU make by running {\tt make -{}-version}; GNU make will say ``{\tt GNU Make}'' on the first line.)

\chapter{Reference: automated inference of nucleotide conservation levels}
\label{consensus}

In order to draw a consensus diagram, it is necessary to determine which nucleotides are conserved
(i.e., are part of the consensus sequence).  It is also useful to determine how conserved they are, and
to annotate which base pairs exhibit covariation.
R2R can perform this calculation, and this output is subsequently used by R2R to actually draw this consensus information.
This process of inferring a consensus, then drawing it, was illustrated in Chapter~\ref{GettingStarted}.
To draw a single RNA molecule, it is not necessary to determine a consensus.

The definition of these consensus statistics was defined in a previous report\cite{Weinberg22Motifs}, which read

\begin{quote}
``To establish the extent of conservation reflected in
consensus diagrams..., sequences were
weighted to de-emphasize highly similar homologs.
Weighting used the GSC algorithm\cite{GSC}, as implemented
by Infernal\cite{InfernalUserGuide}, and weighted nucleotide frequencies were
then calculated at each position in the multiple sequence
alignment. To classify base pairs as covarying, the
weighted frequency of Watson-Crick or G-U pairs was
calculated. However, aligned sequences in which both
nucleotides were missing or where the identity of either
nucleotide was uncertain (e.g. was `N', signifying any
of the four bases) were discarded. Classification as a
covarying position was made if two sequences had
Watson-Crick or G-U pairs that differ at both positions
amongst sequences that carry the motif. If only one
position differed, the occurrence was classified as a
compatible mutation. However, if the frequency of non-Watson-Crick or G-U pairs was more than 5\%, we did
not annotate these positions as covarying or as compatible
mutations.''
\end{quote}

Note: it is now possible to weight sequences using methods other than the GSC algorithm (see below).

\section{Recommended command}
\label{sec:gscdefault}

The standard command line is:

{\tt r2r [flags] -{}-GSC-weighted-consensus} {\it input-sto} {\it output-sto} {\tt 3 0.97 0.9 0.75 4 0.97 0.9 0.75 0.5 0.1}

where {\it input-sto} is the name of a Stockholm-format file containing your alignment.
This command infers a consensus of 3 degrees of nucleotide identity, with weighted conservation thresholds
of 97\%, 90\% and 75\%; 4 levels of presence conservation of nucleotides whose identity is not highly conserved,
with thresholds 97\%, 90\%, 75\% and 50\%; and tolerance for up to 10\% of sequences having
non-canonical base pairs in a given paired position.
The command will create a file {\it output-sto} that will have the same content as {\it input-sto}, but will contain additional
markup summarizing what is conserved.

In the {\it output-sto} file, the line {\tt \#=GC cons} will contain a dash ({\tt -}) for columns that are classified as gaps, and a nucleotide for non-gap columns.
Gap columns are, by default, removed from the consensus when R2R draws it.  See Section~\ref{sec:gscmarkup} for details on the added contents of the {\it output-sto} file.

Available flags that alter the behavior of this process are described in Section~\ref{sec:gscmarkup}.

\section{``Fragmentary'' alignments}

The ``fragmentary'' feature described in this section was motivated by GOLLD RNA \cite{CoolRNAs}.
GOLLD RNAs average roughly 800 nucleotides in size, but are predominantly found in metagenome data
from Lake Gat{\'u}n that consists of short sequence scaffolds.
Therefore, most of the detected GOLLD RNAs are fragments of the full RNA structure.
Moreover, since relatively few GOLLD RNAs are currently available, it is not practical to discard fragmentary
GOLLD RNAs, as this would greatly reduce our ability to analyze its structure.

To handle this situation, R2R has a ``fragmentary'' mode.
In this mode, for example, any sequence that consists entirely of gap characters on its $5^\prime$ end is assumed
to be a fragment that is cut off at this $5^\prime$ end.  In this case, nucleotides that are expected to 
be $5^\prime$ to this cut-off point are treated as unavailable data.  As such, they do not participate
in any of the conservation statistics calculations---neither in the degree of conservation of nucleotides, nor in calculation of base pair covariation.
Similar semantics apply to $3^\prime$ ends that consist entirely of gap characters.
The fragmentary mode was also applied to HEARO RNAs \cite{CoolRNAs} in the case of the $3^\prime$-most hairpin
that is often missing, presumably due to truncation events.

Fragmentary mode is enabled by adding the following line to a {\tt .sto} file:
\examplett{
\#=GF FRAGMENTARY 1
}

\section{Weighting sequences using your own algorithms (for calculating nucleotide frequencies)}
\label{manualweighting}

{\bf WARNING:} please be careful about using this functionality.  It is not appropriate in the vast majority of cases.  In almost all scenarios, it is most correct to use R2R's default weighting.

\subsection{Why R2R weights sequences}
\label{whyweights}

Nucleotide frequencies have often been calculated by simply counting the number of sequences with a particular nucleotide in a given alignment column, and dividing that by the total number of sequences.  For example, suppose there are 10 sequences in an alignment, and 9 of them have an A nucleotide in some particular column, while the 10th sequence has a C in that column.  The just-described approach would report that the A nucleotide is then 90\% (9/10) conserved.

However, biological sequences are typically correlated through evolution, and are therefore not indpendent samples.  For example, suppose that the 9 sequences with an A nucleotide are all various types of primates, and the 10th sequence is a kangaroo.  In this case, the 9 primate sequences are essentially redundant.  In other words, the fact that the A nucleotide occurs in 90\% of cases says much more about what organisms people decide to sequence, and not very much about evolutionary constraints on this nucleotide position. A better estimate would be that the A nucleotide is actually only 50\% conserved.

This issue can also occur even within metagenomic sequences from environmental samples (e.g., bacteria in the human gut), even though no-one chooses the organisms in an environment.  First, there is a bias in which environments are chosen to be sequenced (e.g., there is vastly more metagenomic sequence from the human gut than from the guts of any other animal).  More importantly, the frequency of a particular RNA sequence is not likely to be proportional to the fitness or other property of the RNA.  Rather, the frequency of a given RNA sequence is often more related to the fitness of the organism in which it resides---even a pseudogene can be highly abundant in nature if it happens to be encoded within a prolific organism.

Therefore, R2R weights sequences to try to correct for these issues, and does this with the GSC algorithm\cite{GSC} (as implemented
by the Infernal\cite{InfernalUserGuide} software package the R2R uses).

The one case in which I have used my own sequence weighting is for a project where sequence pools from {\it in vitro} selection were sequenced using high-throughput sequencing.  In this case, it's reasonable to assume that the fitness of a sequence is proportional to its frequency, or at least is the dominant factor. (This assumption ignores evolutionary correlation between the sequences, so it's not perfect.)

You might also use this functionality if you know of a better weighting algorithm, i.e., one that solves the redundancy problem better than the GSC algorithm.

\subsection{How to use your own weights}

If you really need to weight sequences yourself, you must create a version of your Stockholm alignment that has a line specifying the weights, as follows:
\example{
{\tt \#=GF USE\_THIS\_WEIGHT\_MAP} {\it map}
}
where {\it map} is a space-separated list, where each element of the list specifies a hitId identifying a sequence in the alignement,
followed by a space, followed by the weight of that sequence.  For example, the following is a valid Stockholm file with custom weights:
\begin{verbatim}
# STOCKHOLM 1.0
A            CCAAAGG
B            ACUAUGU
#=GC SS_cons <<...>>
#=GF USE_THIS_WEIGHT_MAP A 5 B 1
//
\end{verbatim}
and says that RNA molecule {\tt A} should be weighted 5 times more than molecule {\tt B}.

Then run the {-{}-GSC-weighted-consensus} command on this augmented Stockholm file.
The {-{}-GSC-weighted-consensus} command will use the specified weights whenever
the {\tt \#=GF USE\_THIS\_WEIGHT\_MAP} tag is present.

\section{General command}
\label{sec:gscgeneral}

This section gives additional detail on the command that will likely not be relevant to most readers' needs.

{\tt r2r} {\it{} [other flags]} {\tt -{}-GSC-weighted-consensus} {\it input-sto} {\it output-sto} {\it identity-levels} [...] {\it present-levels} [...] {\it max-non-canon}

where
\begin{itemize}
\item optional {\it{} [other flags]} are:
  \begin{itemize}
  \item {\tt -{}-maxNonCanonInNoVariationObserved} {\it fraction} : by default {\it fraction}=1. This means that R2R will tolerate up to 10\% non-canonical base pairs, while depicting base pairs that do not vary (among canonical pairs).  A fraction of {\it fraction}=0 will prevent non-canonical base pairs, and is probably a better idea.
  \item {\tt -{}-cutEmptyLines} : in some circumstances, {\tt r2r} must eliminate lines from the input file.  This occurs when the input file has lines such as {\tt \#=GC cons} that {\tt r2r} wants to output.  Instead of deleting the lines, {\tt r2r} will substitute these lines with empty lines, in order to keep the line numbers consistent, so that any error messages will make sense with the line numbers of the original file.  You can suppress these empty lines with the {\tt -{}-cutEmptyLines} flag.
  \item {\tt -{}-topNMostConserved} {\it N} : indicate which of the columns are among the {\it N} most conserved.  The output is in the form of an extra label line ``{\tt \#=GC R2R\_XLABEL\_mostcommon}''.  The conserved columns are indicated with an `{\tt x}', while the other columns have a dot.
  \item {\tt -{}-position-based-weights} : do not weight sequences with the GSC Algorithm, but instead use a position-weight-based algorithm (which is also implemented by the Squid library of the older version of Infernal that is included as part of the R2R code).  This algorithm is significantly faster with many sequences, but might not work as well as the GSC algorithm.
  \item {\tt -{}-verbose} : print out a bit of extra info to {\tt stderr}.  The extra information is very limited, so ``verbose'' is kind of an exaggeration.
  \end{itemize}
\item {\it identity-levels} is an integer $N$ followed by $N$ real numbers.  $N$ defines the number of different levels for classifying the degree of conservation of a nucleotide identity.  The real numbers define a minimum frequency (from 0 to 1) for each level, and must be in decreasing order.
\item {\it present-levels} is also an integer $N$ followed by $N$ real numbers.  In this case, $N$ defines the number of different levels for classifying how frequently the nucleotide is present.  The real numbers define minimum frequencies (0 to 1) for each level, and must be in decreasing order.
\item {\it max-non-canon} is a real number specifying the maximum frequency (0 to 1) of non-canonical base pairs.  Non-canonical base pairs are nucleotide pairs that are not A-U, C-G, G-C, U-A, G-U or U-G.  If the number of non-canonical base pairs in paired-columns exceeds {\it max-non-canon}, then R2R will never declare covarying, compatible, or non-mutating pairs.  In other words, it will always declare code {\tt ?}, which corresponds to base pairs that are not shaded in consensus diagrams.
\end{itemize}

A concrete example of these parameters and their meaning was given in Section~\ref{sec:gscdefault}.

Note: if {\it identity-levels} is greater than 3 or {\it present-levels} is greater than 4, R2R is not designed to be able to draw the consensus diagram, although it would be relatively straightforward to extend it to interpret additional levels.  I have assumed this is not likely to be an important feature.

\section{Output of R2R for the consensus}
\label{sec:gscmarkup}

You probably won't have to understand R2R's consensus output, except to
understand that R2R removes gap columns when it
draws the consensus (though you can override this with the {\tt keep} command), and that it annotates 
these gap columns in the line beginning {\tt \#=GC cons}.

R2R's consensus output consists of a series of {\tt \#=GC} lines that annotate per-column data.
These lines are:
\begin{itemize}
\item {\tt \#=GC cons} : Consensus sequence.  Gaps are represented as dashes ({\tt -}).  Nucleotide symbols
({\tt A}, {\tt C}, {\tt G}, {\tt U}) represent conserved nucleotides.
Two degenerate IUPAC symbols are used: {\tt R} means the position is conserved as either A or G, 
and {\tt Y} represents either C or U.
A lower case `{\tt n}' is used for columns that are typically or always present, but do not conserve the nucleotide
identity.  These positions are drawn as circles by R2R.
An annotated example is given in Figure~\ref{fig:demo1}D.
\item {\tt \#=GC conss} : The degree of conservation.  ``{\tt 1}'' is the highest level of conservation.
When the consensus letter (in {\tt \#=GC cons}) is a nucleotide
or degenerate nucleotide, by default the levels range from {\tt 1} to {\tt 3}, and represent how often that nucleotide
is found in this alignment position.
When the consensus letter is lower-case {\tt n}, by default the levels range from 
{\tt 1} to {\tt 4}, and represent how often any nucleotide is found in this position (as opposed to a gap).
When the consensus letter is a dash (representing a gap), the number is always {\tt 0} (zero).
An annotated example is given in Figure~\ref{fig:demo1}D.
\item {\tt \#=GC cov\_SS\_cons} : Covariation annotation.
{\tt 2} represents base pairs that show at least one instance of covariation,
{\tt 1} means there is at least one compatible mutation,
{\tt 0} (zero) means there are no mutations, so no data.
{\tt ?} means that there are too many non-canonical base pairs.
An annotated example is given in Figure~\ref{fig:demo1}D.
\item {\tt \#=GC cov\_SS\_cons}{\it x} : covariation data for pseudoknot structure line {\tt \#=GC SS\_cons}{\it x}.
\item {\tt \#=GF NUM\_COV} : Number of base pairs with covariation. (per-file annotation)
\item {\tt \#=GF WEIGHT\_MAP} : Lists GSC-algorithm weights used for each sequence, in a space-spearated list.  The even-numbered fields identify a sequence, and the odd numbered fields are the weights. (per-file annotation)  Note: if you use {\tt \#=GF USE\_THIS\_WEIGHT\_MAP}, the results of {\tt \#=GF WEIGHT\_MAP} will be identical to the weights you supplied
\item {\tt \#=GF USED\_GSC TRUE} : Confirms that R2R used the GSC algorithm.
(per-file annotation) 
\item {\tt \#=GC col\_entropy\_}{\it i} : Experimental feature, the entropy of each column, written vertically.
\item {\tt \#=GF DID\_FRAGMENTARY FALSE} : Use of {\tt FRAGMENTARY} feature is recorded.
(per-file annotation)
\end{itemize}

\section{Generating your own alignment consensus, bypassing R2R}
\label{sec:consensuswithoutGSC}

When drawing a consensus,
R2R depends on data added by the {\tt -{}-GSC-weighted-consensus} command, which is added
to the original information in the Stockholm-format alignment file.
An alternate program to calculate the consensus data need only output these extra data fields.
The required fields are {\tt \#=GC cons}, {\tt \#=GC conss}, {\tt \#=GC cov\_SS\_cons}
and for any extra stem {\tt \#=GC SS\_cons}{\it i}, there must be {\tt \#=GC cov\_SS\_cons}{\it x}.
The form of these fields is described in the previous section.
Examples can be generated from the {\tt demo} files, by looking at the 
output of the  {\tt -{}-GSC-weighted-consensus} command
in the {\tt demo/intermediate} directory (the file names will end with {\tt .cons.sto}).

R2R is, however, committed to up to 9 conservation levels, each corresponding to a different color.
For a larger set of colors, the color of nucleotides can be overridden later using the {\tt nuc\_color} command (see Section~\ref{sec:nuccolor}).  There is no command to specify an arbitrary color 
for shading of base pairs.

\chapter{Reference: drawing}

\section{Running R2R}
\label{sec:runRtoR}

R2R processes Stockholm-format ({\tt .sto}) files with special markup.
Note that interleaved lines are not permitted, so technically R2R can only process Pfam format.  R2R has some other incompatibilities with Stockholm format, as explained in section~\ref{sec:notstockholm} for details.
You can convert Stockholm-format alignments into the Pfam format using the {\tt esl-reformat} command that comes with the HMMER and Infernal software packages in the {\tt easel/miniapps} directory.

The files must have information on the consensus, which is also made by running R2R
(see Chapter~\ref{consensus}).

There are two command lines you can use to process Stockholm-format alignments:

\begin{enumerate}
\item {\tt r2r} {\it input-file}{\tt .sto} {\it output-file}

where {\it output-file} will be created in either Adobe Acrobat PDF format or Scalable Vector Graphics (SVG) format (see below),
and {\it input-file}{\tt .sto} is a file in Stockholm format.

Note: some functionality is not available in this command-line format, so option \#2 below is preferable.
\item {\tt r2r} {\it input-file}{\tt .r2r\_meta} {\it output-file}

where {\it output-file} will be created in PDF or SVG format,
and {\it input-file}{\tt .r2r\_meta} is a file that lists individual Stockholm-format files,
and has additional options.
Its format is described in Section~\ref{metafile}.
The diagrams drawn will be organized into a grid in {\it output-file}.
\end{enumerate}

The {\it output-file} name must end in either {\tt .pdf} or {\tt .svg}, which determines what output format
R2R will write.
Note: PDF output is optimized for Adobe Illustrator, while SVG output is optimized for Inkscape, mainly in terms
of the font.
Of course, with SVG output, you can easily do a search \& replace to change the font.
However, R2R internally knows (approximately) font metrics, which allows it to position text (almost) accurately.
Therefore, you should use a font that is related to Helvetica: Helvetica, Arial, Bitstream Vera Sans or DejaVu Sans.

The following flag can be added after {\tt r2r}:
\begin{itemize}
\item {\tt -{}-disable-usage-warning} : do not output the warning at the top of the PDF or SVG file.  By default, R2R warns about appropriate usage.
\item {\tt -{}-inkscape-helvetica-font-name} {\it fontName} : use this font name in output SVG files.  Inkscape is sometimes happier if this name matches a font that it has.  The default is ``{\tt Bitstream Vera Sans}''.
\end{itemize}

\section{About the Stockholm file format, and how R2R uses it}

R2R uses the generalizable format of Stockholm files to specify markup that defines how a motif is drawn.
Stockholm format is explained on Wikipedia (\url{http://en.wikipedia.org/wiki/Stockholm_format}).

R2R's deviations from the Stockholm format are explained in section~\ref{sec:notstockholm}.

The key ideas is that R2R uses the following types of annotation:
\begin{itemize}
\item {\tt \#=GF} {\it tag} {\it data}

This specifies file-level markup. For example R2R drawing commands use the form
({\it tag} = {\tt R2R}):

{\tt \#=GF R2R} {\it command}
\item {\tt \#=GC} {\it tag} {\it columns}

In this case, each position in the string {\it columns} corresponds to each column in the alignment.
For example, for {\it tag} = {\tt SS\_cons}, the symbols in the {\it columns} string specify the consensus
secondary structure of the alignment.
\item {\tt \#=GR} {\it hitId} {\it tag} {\it columns}

Each position in the string {\it columns} corresponds to each column in the sequence specified by {\it hitId}
(see Section~\ref{hitId}).
This is relevant to identifying columns in an individual RNA when the user wants to draw that RNA sequence, instead
of the consensus.
\end{itemize}

Pedantic point: R2R's input format does not quite respect the Stockholm format, or even the more restrictive Pfam format.  For example, it uses many {\tt \#=GF R2R} lines, each of which represents a distinct command.  However, the Stockholm format says that these lines are logically part of the same line, and the separation of commands is technically not enforced.  However, has not been a problem in practice.

\subsection{A note on R2R's representation of consensus secondary structures}

R2R uses a non-standard method for writing consensus secondary structures for motifs with pseudoknots.
To describe pseudoknots, in addition to the standard {\tt \#=GC SS\_cons} line, there can be other lines
of the form {\tt \#=GC SS\_cons}{\it X} for any string {\it X} (typically, I call them
{\tt \#=GC SS\_cons\_1}, {\tt \#=GC SS\_cons\_2}, ...).  

To mark base pairs, 
a structure line must contain only left
brackets (``('', ``['', ``\textless{}'' or ``\{'') or right brackets (``)'', ``]'', ``\textgreater{}'' or ``\}'').
All other symbols are assumed to be single-stranded.
(Thus, R2R does not use Rfam's method for marking pseudoknots.)

I find that this method reduces special-case code because all pairs are specified in the same way.
It also allows for nucleotides that can be in more than one base pair interaction (not necessarily simultaneously).
This is occassionally relevant to conserved structures (e.g., with alternate pairings), and is convenient
when the exact structure is unknown.

R2R does not work with consensus structures in Rfam format.
Supporting such a format automatically would require some rethinking of structure-related commands such
as {\tt ignore\_ss\_except\_for\_pairs}.

\subsection{R2R does not conform to the Stockholm file format}
\label{sec:notstockholm}

R2R is not fully compatible with the Stockholm file format.  Thanks to Elena Rivas for helping with this information.

\begin{itemize}
\item R2R cannot process interleaved lines, and thus it truly only processes ``Pfam format'', which is a subset of Stockholm format.  With interleaved lines, the sequence for a given ``hit'' within an alignment can span multiple lines of text, and this is used to reduce the width of files.  In Pfam format, sequences and column annotation (for \#=GC and \#=GR lines) must be written in one line.

You can convert Stockholm-format alignments into the Pfam format using the {\tt esl-reformat} command that comes with both the HMMER and Infernal software packages in the {\tt easel/miniapps} directory (in both software packages).
\item R2R and the Stockholm format differ in their usage of whitespace, which is defined as one or more consecutive non-printable characters, like spaces, tabs or newlines.  R2R requires that most whitespace consist of exactly one space or tab character.  Whitespace separates components of the Stockholm format.  For example, the line
  \examplett{\#=GC MAGIC\_COLUMNS  ......*......**}
  has the components {\tt\#=GC}, {\tt MAGIC\_COLUMNS} and {\tt ......*......**}.  Each of these components are separated by whitespace.  For R2R, there must be exactly one space or tab character between {\tt\#=GC} and {\tt MAGIC\_COLUMNS}, but the Stockholm format allows multiple spaces.  Some programs use multiple spaces to line things up to make it easier for humans to read, or for other reasons.  Such alignments would need to have the extra spaces removed in order for R2R to be able to parse them.  Multiple spaces are allowed by R2R between {\tt MAGIC\_COLUMNS} and {\tt ......*......**}.
  \item As noted in the previous section, R2R's definition of conserved secondary structure differs from the mechanism used by the Rfam Database.  However, this definition does not belong to the Stockholm format definition.
\end{itemize}

\section{R2R ``drawing units''}
\label{drawingunits}

R2R divides an RNA structure into units, and positions each unit as a block.
Units are: (1) a single
stranded region, including terminal loops, (2) both sides of an
internal loop or (3) a stem. 
Some commands operating on labels will have no effect if the label corresponds to a position in the middle
of a unit.
For example, to use the {\tt turn\_stem\_at\_internal} command, you must specify the $5^\prime$-most nucleotide
in the internal loop (although in this command, the $3^\prime$-most nucleotide is also permitted).

{\tt place\_explicit} commands that reference the midding of a drawing unit will break it into two units.

\section{Data types in R2R}

\subsection{hitId}
\label{hitId}

I define hitIds as strings in Stockholm-format files that identify each sequence.  In Stockholm files, the hitId begins a line, and is followed by whitespace, which are then followed by the RNA sequence with gap characters.

\subsection{Distances}

Distances within drawing commands are specified in internucleotide units.  One internucleotide unit
is the distance between two consecutive nucleotides along the backbone, and defaults to 0.105 inches.

\subsection{Angles}

Angles are always specified in degrees.

\subsection{Measurements (length/width/size)}
\label{measurements}

Some constant lengths used by R2R can be set to non-default values using the {\tt SetDrawingParam} command (see Section~\ref{SetDrawingParam}). The default units are inches, except for font sizes. Font sizes are by default measured in points.  Suffixes can be used to specify other measurement units.  No whitespace is allowed between the number and the suffix.  The following measurements are equivalent:
\begin{itemize}
\item {\tt 1} (default is inch)
\item {\tt 1in} (just makes the default explicit)
\item {\tt 2.54cm}
\item {\tt 25.4mm}
\item {\tt 72pt}
\end{itemize}

\subsection{Colors}
\label{color}

Colors (e.g. in the {\tt circle\_nuc} command) are specified as follows.  

\begin{itemize}
\item RGB colors are written in the form

{\tt rgb:}\textit{r}{\tt ,}\textit{g}{\tt ,}\textit{b}

where \textit{r,g,b }(written separated by commas) are numbers from 0 to
255.9999.
\item Standard colors used in cleavage diagrams for in-line probing experiments can also be used.  They are specified
as 
{\tt cleavage:?} (unknown/no data, gray), {\tt cleavage:=} (constitutive, yellow), {\tt cleavage:-} (decreasing, red), {\tt cleavage:+} (increasing, green).
\end{itemize}

\section{.r2r\_meta file}
\label{metafile}
You specify what files R2R should process in a file with extension
{\tt .r2r\_meta}.
The file is tab delimited---all fields must be separated by tab characters.  If the first (tab-delimited) field is empty,
the line is ignored (i.e., this is a comment).

If a line has only one field, that field is the path to a Stockholm-format
alignment file.  This file is processed to draw a consensus diagram.

If the line{\textquoteright}s first field is {\tt SetDrawingParam}, then you
can change the default settings of certain drawing parameters (see below).

You can also specify that a single RNA molecule should be drawn (rather than a consensus),
as described below.

\subsection{Display name}
\label{sec:displayName}

R2R will output multiple drawings into its output PDF or SVG file.  There is one drawing per line in the {\tt .r2r\_meta} file.  By default, R2R creates names for these drawings that are based on the file name and any other parameters you specify (such as `defines' or `oneseq' mode, as explained in the following subsections).  You can override the name for any line (in the {\tt .r2r\_meta} file) by adding

\examplett{
  {\tt displayname} {\it name}
}

to the line (as tab-delimited fields, like normal).

This functionality is for R-scape, and any other software that automatically generates R2R input.

\subsection{Defines}

You can define symbols that control which R2R commands are processed, as described in Section~\ref{conditionals}.
You can add {\tt define} {\it name} {\it value} to any line that specifies a {\tt .sto} files in the {\tt .r2r\_meta} file.  These defines affect the drawing of that {\tt .sto} file.

For example, see the file {\tt demo/SAM-IV/SAM-IV.r2r\_meta}.  This file directs the drawing of SAM-IV riboswitches in various ways, and was used for the various examples in the tutorial that used SAM-IV.

Defines are only applicable to the line in which they appear.

\subsection{SetDrawingParam}
\label{SetDrawingParam}

You can include lines like the following in the {\tt .r2r\_meta} file
\example{
{\tt SetDrawingParam} \textit{ name  value}
}
to change internal drawing parameters in R2R.
This affects all subsequent lines of the {\tt .r2r\_meta} file.

Valid \textit{name} codes are:

\begin{itemize}
\item Miscellaneous font sizes (note: as with all font sizes, default units are points for font sizes only)
	\begin{itemize}
	\item {\tt nameFontSize} : for text that shows the name of RNA motifs. (Default: {\tt 12pt}.)
	\item {\tt nucFontSize} : writing nucleotides (Default: {\tt 7.5pt}.)
	% 8pt for Myriad
	\end{itemize}
\item Nucleotide layout distances
	\begin{itemize}
	\item \label{internucleotideLen}{\tt internucleotideLen} : distance between consecutive nucleotides along the backbone.  (Default: {\tt 0.105in}.)
	\item {\tt pairLinkDist} : distance between base-paired nucleotides. (Default: {\tt 0.17in}.)
	\end{itemize}
\item {\tt tick\_label} nucleotide labels
	\begin{itemize}
	\item {\tt nucTickLabel\_distFromNuc} : the distance from the nucleotide at which the tick line starts. (Default: {\tt 3.75pt}.)
	\item {\tt nucTickLabel\_tickLen} : the length of the tick line. (Default: {\tt 4pt}.)
	\item {\tt nucTickLabel\_tickColor} : the color of the tick line. (Default: {\tt rgb:0,0,0} , i.e., black.)
	\item {\tt nucTickLabel\_tickPenWidth} : the width of the tick line. (Default: {\tt 0.5pt}.)
	\item {\tt nucTickLabel\_extraSpaceToText} : extra distance between the end of the tick line and the start of text (might be a bit off because R2R still does not use absolutely exact font metrics). (Default: {\tt 1.5pt}.)
	\item {\tt nucTickLabel\_fontSize} : the font size used to draw the label. (Default: {\tt 6pt}.)
	\end{itemize}
\item {\tt backboneWidth} : width of lines that represent the backbone (used for {\tt skeleton} drawings and for variable-length regions). (Default: {\tt 1.5pt}.)
\item outlining/inlining nucleotides (also used to indicate nucleotides involved in pseudoknot pairings)
	\begin{itemize}
	\item {\tt outlineNucExtraRadius} : extra distance from outside of nucleotide that is added before the outline line. (Default: {\tt 3.75pt}.)
	\item {\tt outlineNucPenWidth} : width of pen used to draw the outline. (Default: {\tt 0.5pt}.)
	\item {\tt outlineNucColor} : color used. (Default: {\tt rgb:92,92,92}.)
	\item {\tt circleRadiusToSmoothDirectionChange} : when two straight lines in the outline meet, the direction change is smoothed using a circular arc.  This parameter is the radius of the arc.  (Default: {\tt 0.025in}.)
	\item {\tt outlineAutoJoin} : if true, R2R will attempt to join lines and arcs that are part of the outline, for both outlines generated implicitly by pseudoknots and by outlines generated explicitly with the commands {\tt outline\_nuc} or {\tt inline\_nuc}.  Sometimes R2R's heuristics for drawing outlines fail, and so it is better to doing the joining yourself.  (Default: {\tt true}.)
        \item {\tt prefixSsWithPkInDrawings} : When you configure an {\tt SS\_cons} line to be a callout-style pseudoknot by using the {\tt ignore-ss-except-for-pairs} command, R2R will indicate the two sides of each stem with the text {\tt pk\_}...  For example, with {\tt SS\_cons\_1}, the added text would be ``{\tt pk\_1}''.  This behavior corresponds to {\tt prefixSsWithPkInDrawings} being {\tt true}.  If {\tt prefixSsWithPkInDrawings} is set to {\tt false}, R2R will not use the prefix {\tt pk\_}, but will just use the label past {\tt SS\_cons\_}, so in our example the label would be ``{\tt 1}''.  (Default: {\tt true}.)
	\end{itemize}
\item Circling nucleotide (also used for cleavage diagrams in {\tt oneseq} mode)
	\begin{itemize}
	\item {\tt cleavageIndicatorRadius} : radius of the circle.  Note: R2R will refuse to allow this to be greater than half the length of {\tt internucleotideLen}, since then circles would overlap. (Default: {\tt 3.75pt}.)
	\item {\tt cleavageIndicatorPenWidth} : width of line surrounding the circle. (Default: {\tt 0.5pt}.)
	\end{itemize}
\item Drawing base pairs bonds
	\begin{itemize}
	\item {\tt pairBondLen} : the length of the line connecting Watson-Crick base pairs. (Default: {\tt 0.054in}.)	\item {\tt pairBondWidth} : the width of the line connecting Watson-Crick base pairs. (Default: {\tt 0.02in}.)
	\item {\tt pairBondGURadius} : radius of the filled circle for representing G-U wobble pairs (only used in {\tt oneseq} mode). (Default: {\tt 0.02in}.)
	\item {\tt pairBondNonCanonRadius} : radius of the un-filled circle for representing non-canonical base pairs (only used in {\tt oneseq} mode). (Default: {\tt 0.01in}.)
	\item {\tt pairBondCircleLineWidth} : width of the line of the circle used for G-U or non-canonical base pairs. (Default: {\tt 0.002in}.)
	\item {\tt minPairShadeGap} : minimum distance between shading rectangles of consecutive base pairs.  When base pairs are drawn at an angle (45 degrees is worst), the shading box's width needs to grow in order to cover the whole nucleotide letter, since nucleotides are rectangles that are always drawn in the same orientation.  Eventually these shading rectangles can get very close or even overlap, which looks bad.  Therefore, R2R will cheat and allow some small parts of nucleotide letters to be unshaded in order to guarantee a minimum distance between the shading rectangles, and this minimum distance is {\tt minPairShadeGap}. (Default: {\tt 2.5pt}.)
	\end{itemize}
\item Circles that represent nucleotides whose identities are not conserved:
	\begin{itemize}
	\item {\tt anyNucCircleWidth} : the width of the circle's line. (Default: {\tt 0.01in}.)
	\end{itemize}
\item Variable-length stems, loops or linkers
	\begin{itemize}
	\item {\tt varHairpinNumFakePairs} : for the {\tt var\_hairpin} command, the number
	of fake base pairs to draw. (Integer. Default: 3.)
      \item {\tt varTerminalLoopRadius} : for the {\tt var\_hairpin} or {\tt var\_term\_loop} commands, the radius of the arc used to represent the terminal loop. (Default: {\tt 0.17in}.)
        \item {\tt backboneConnectorCircleRadius} :the radius of the arc that R2R uses to bend the backbone, esp. when drawing the $5^\prime$ end of the molecule.  This can be more important when the $5^\prime$ end of the molecule is at a weird angle.  However, because R2R sets the position of the text that says ``$5^\prime$'' in an inflexible way, there's a very small range of values for {\tt backboneConnectorCircleRadius} that will work okay.  If the value is too big, R2R will draw the connecting line directly, rather than using an arc.  (Default: {\tt 0.035}.)
        \item {\tt backboneAnnotTextOffset} : the distance between the text annotation and the backbone line for {\tt var\_backbone...} commands.  If this value is negative, then it is set as 4/7.5 the value of {\tt varBackboneFontSize}. The reason for this weird ratio is that originally it was set to 4 points, and the font size was 7.5, so with this default, the default drawings will be like before (Default: {\tt -1} , i.e. set based on the ratio).
      \item {\tt varBackboneFontSize} : the font size used for the annotation in the various forms of the {\tt var\_backbone...} command.  (Default: {\tt 7.5pt})
        \item {\tt varTermLoopFontSize} : the font size used for the annotation in the {\tt var\_term\_loop} command.  (Default: {\tt 7.5pt})
	\end{itemize}
\item Shading along backbone or drawing in {\tt skeleton} mode
	\begin{itemize}
	\item {\tt outlineAlongBackboneWidth} : width of shading for {\tt outline\_along\_backbone} command.  (default: {\tt 0.7pt}.)
	\item {\tt shadeAlongBackboneWidth} : width of shading for {\tt shade\_along\_backbone} command.  If the value is negative, then the computer will automatically set the width based on the size of nucleotide symbols, i.e., the font size.  (default: {\tt -1}, i.e. set automatically.)
	\item {\tt alongBackboneStyle} : permissible styles are integers (default: {\tt 0}) as follows:
		\begin{itemize}
		\item {\tt 0} : shade nucleotides with circles; join consecutive base pairs with line, but keep endpoints rounded.  This tends to look good with {\tt shade\_along\_backbone} where nucleotides are shaded.  It is the default style.
		\item {\tt 1} : shade nucleotides with an appropriately thick line that extends half-way to the previous nucleotide, and half-way to the next.  This looks better for skeleton drawings where you want to change colors (otherwise lone circles look weird when they're small and not shading a nucleotide).
		\end{itemize}
	\item {\tt alongBackboneMidpointGapLength} : If {\tt alongBackboneStyle} is set to {\tt 1}, this parameter allows for gaps between the shading of consecutive nucleotides.  By default differently colored consecutive nucleotides will have the two different colored shading paths joined at the midpoint between the two nucleotides.  However, a gap of length {\tt alongBackboneMidpointGapLength} can be left between these two paths, so that they don't touch.  There are two technical issues that complicate this feature.  First, the length for the gap is the length along the backbone path between the two nucleotides, not the straight-line length.  Thus, if the two consecutive nucleotides lie on a circle (meaning their backbone path is an arc) the straight-line gap between them will be a bit smaller than {\tt alongBackboneMidpointGapLength}.  Second, R2R draws the backbone shading as a line/arc and sets the thickness of the line to simulate the shading.  The length of the gap is thus calculated at the midpoint of the shading line.  But, if the path is on a circle, the two ends of the shading line will not be of equal distances.  Usually the circles are of a large enough radius that these effects have no practical effect.

(By the way, I implemented this feature to experiment with shading nucleotides involved in a pseudoknot base pair to show covariation, as an alternate way of indicating pseudoknots.  These pseudoknotted nucleotides are often on a circular path.  The gaps between shading are meant to simulate the gaps between consecutive nucleotides in a regular stem.  Some other changes would be needed to implement this style, especially since currently all {\tt shade\_along\_backbone} commands use the same drawing style, so covariation shading and other user-directed shading would conflict.)

If {\tt alongBackboneStyle} is not set to {\tt 1}, this parameter has no effect.
(Default: {\tt 0}, i.e. don't leave any gap.)
	\end{itemize}
\item Debugging help
	\begin{itemize}
	\item {\tt showPlaceExplicit} : Boolean ({\tt true} or {\tt false}).
	Shows information on what {\tt place\_explicit} commands were used to position drawing units (shown as thick lines),
	as well as unused commands (thin lines).  
	The lines connect the two nucleotides that were positioned based on a {\tt place\_explicit}.
	User-specified {\tt place\_explicit} commands are shown in magenta,
	while default rules are gray.  Lines are labeled.  Default rules say ``def''.  The positioning of the $5^\prime$-most
	nucleotide says ``1st''.  User-specified {\tt place\_explicit} commands are labeled with their line number in the
	input Stockholm file.
	To simply the layout, I used a very small font size to label the lines.  You will probably need to use display software to zoom into the picture in order to read the labels.
	(Default: {\tt false}.)
	\item {\tt showEachNucleotideDir} : Boolean ({\tt true} or {\tt false}).
	Shows the direction vector at each nucleotide.  Useful in particular for working out {\tt place\_explicit} commands.
	Note that some nucleotides have two directions.  These are nucleotides that are positioned in a straight line on their
	$5^\prime$ direction, but are on a circular layout on their $3^\prime$ end.  Linear directions are shown with a normal arrow.
	Circular directions are shown with a line that has a small circle on the direction end.
	Direction vector lines always originate at the center of the nucleotide.
	Lines are drawn in cyan.
	(Default: {\tt false}.)
        \item {\tt dumpInfoFile} : if non-blank, the name of a file into which R2R should dump information on the layout and other features.  See section~\ref{sec:dumplayout}.  If blank, no information is output.  (Default: blank.)
	\end{itemize}
\item Parameters relevant to {\tt skeleton} mode
	\begin{itemize}
	\item {\tt skeleton\_scaleMeasurementsBy} : alternate value for {\tt scaleMeasuresBy} to be used in {\tt skeleton} mode.  (Default: {\tt 0.25}.)
	\item {\tt skeleton\_pairBondWidth} : alternate value for {\tt pairBondWidth} to be used in {\tt skeleton} mode.  Note that this is {\em before} scaling is applied. (Default: {\tt 0.5pt}.)
	\item {\tt skeleton\_backboneWidth} : alternate value for {\tt backboneWidth} to be used in {\tt skeleton} mode.  Note that this is {\em before} scaling is applied. (Default: {\tt 1.5pt}, which is the same as the default for {\tt backboneWidth}.)
	\item {\tt skeleton\_outlinePseudoknots} : enable drawing an outline for callout-style pseudoknots.  If this variable is set to {\tt true}, then commands {\tt ignore\_ss\_except\_for\_pairs} {\tt outline} will result in an outline even in skeleton mode.  If the variable is set to {\tt false}, then this command will not result in an outline in skeleton mode---I think you usually don't want the outline.  (Default: {\tt false}.)
	\end{itemize}
\item Parameters relevant to {\tt circle\_nuc} or implicit circling from the {\tt \#=GR ... CLEAVAGE} line
in {\tt oneseq} mode.
	\begin{itemize}
	\item {\tt nucShrinkWithCircleNuc} : when the {\tt circle\_nuc} command is used, nucleotides are shrunk to accommodate the circles.  They are scaled by a factor of {\tt nucShrinkWithCircleNuc}.  (Default: 0.8)
	\item {\tt pairBondScaleWithCircleNuc} : scaling applied to base-pair bonds to accommodate circled nucleotides.  (Default: {\tt 1}.)
	\end{itemize}
\item Parameters relevant to {\tt oneseq} mode (drawing a single RNA molecule)
	\begin{itemize}
	\item {\tt drawStandardCleavage} : Boolean ({\tt true} or {\tt false}).  If {\tt true}, R2R will draw the standard cleavage diagram based on the {\tt \#=GR ... CLEAVAGE} line.  (Default: {\tt true}.)
	\item {\tt defaultOneseqLabeling} : If {\tt true}, then in {\tt oneseq} mode (see below), R2R will label the number of every 10th nucleotide. (Default: {\tt true}.)
	\item {\tt indicateOneseqWobblesAndNonCanonicals} : If {\tt true}, then in {\tt oneseq} mode (see below), R2R will indicate whether pairs are Watson-Crick, wobble (i.e., G-U) or non-canonical by using different bond symbols.  If {\tt false}, R2R will use the same symbol for all pairings, specifically all pairings will be drawn as if they are Watson-Crick.  (Default: {\tt true}.)
        \item {\tt makeRedNucsRedInOneseq} : Boolean.  For any nucleotides that belong to the highest conservation level (i.e., level `1', which is usually depicted as red nucleotides that are $geq 97\%$ conserved), color these nucleotides red in the single sequence.  (Default: false.)
        \item {\tt makeNonDegenRedNucsRedInOneseq} : Boolean.  Same as {\tt makeRedNucsRedInOneseq}, except ignore highly conserved {\tt R} or {\tt Y} nucleotides, which in fairness aren't really that conserved.  Note: if {\tt makeRedNucsRedInOneseq} is true, then it doesn't matter what the value of {\tt makeNonDegenRedNucsRedInOneseq} is. (Default: false.)
	\end{itemize}
\item {\tt pairShadingColors}  \textit{covariationColor  compatibleColor 
noMutationsObservedColor }: change the colors used to shade pairs. 
Colors are specified using standard R2R codes.  Good colors for slides
are 

\begin{quote}
\hspace*{-24pt}{\tt SetDrawingParam pairShadingColors rgb:156,199,153 rgb:152,199,222 rgb:235,138,126}
\end{quote}
(thanks to Kirsten F. Block for these values).
\item {\tt solverMaxIters} {\it \#} : sets the maximum number of iterations of the solver (i.e., CFSQP) for 
layout of multistem junctions.  Sometimes a high number of iterations is necessary since the solver gradually
creeps towards the solution.  However, usually fewer iterations are needed, and so the conservatively high default
value is a waste of time.  (Defualt: {\tt 100000}.)

Note: you can compile R2R with a new default by adding {\tt -DMAX\_SOLVER\_ITERS=\#} to the {\tt g++} command
line.
%Also, if you want to speed re-drawing of an individual motif that has already been solved,
%you can extract an equivalent {\tt multistem\_junction\_bulgey} command from R2R's output.
%If you put this {\tt multistem\_junction\_bulgey} command in place of the original {\tt \_solver} command,
%the layout will be identical, but the solver will not be needed.
%\item NOTE: there some additional parameters that have not been
%made settable, since thus far there has been no need.
\item {\tt autoBreakPairs} : if this variable has the value {\tt true}, then R2R will automatically 
separate base pairs where one of them should be deleted and the other should not.
This is suitable for quick drawings where you don't want to think about how to resolve these issues.
If this variable has the value {\tt false}, R2R will report an error. (Default: {\tt false}.)
\item {\tt DNA} : if this variable has the value {\tt true}, R2R will treat the input under the assumption that it is a single-stranded DNA molecule.  In this mode, and {\tt U} nucleotide in the input will be changed into a {\tt T}.  (Default: {\tt false}.)
\item {\tt disableSubfamWeightText} : when you use the modular structure functionality, which is based on the {\tt SelectSubFamilyFromStockholm.pl} script, this script will output a field {\tt \#=GF SUBFAM\_WEIGHT} into the output {\tt .sto} file.  This weight indicates the fraction of sequences that exhibit the modular structure.  {\tt r2r} will then output a message at the top of the relevant drawing that states this fraction (e.g. ``subfam\_weight=1'').  To disable this, set {\tt disableSubfamWeightText} to {\tt true}.  If you set {\tt disableSubfamWeightText} to {\tt false}, R2R will print the subfam\_weight if it is there.  (Default: {\tt false}.)
\end{itemize}

\textit{name }and \textit{value }pairs may be repeated multiple times. 
Thus, for example, the following line is legal, and sets 2 parameters
at once:

{\tt SetDrawingParam\ \ varHairpinNumFakePairs\ \ 2\ \ varTerminalLoopRadius\ \ 0.12}

\subsection{Oneseq mode}

If a line has three fields and the second field is equal to the literal
string {\tt oneseq}, then the first
field is the path to a Stockholm alignment file, and the third field is
a ``hitId'' (Section~\ref{hitId}) specifying a specific sequence within the alignment.  In this
case, R2R will draw that
sequence.
It can optionally draw the sequence like a cleavage diagram for in-line probing experiments
(Section~\ref{cleavage}).

\subsection{Skeleton mode}
\label{skeleton}

If a line has two fields and the second field is equal to either the literal string
{\tt skeleton} or {\tt skeleton-with-pairbonds}, then the RNA is drawn
in skeleton mode.  In this mode, a black line traces the backbone.  Such
drawings are often used to present a small sketch of an overall structure,
especially for larger RNAs.

Drawing of nucleotides, nucleotide conservation, 
length of variable-length regions and covariation annotation
are supressed.
The lines that indicate base-pair bonds are supressed with {\tt skeleton},
but drawn with {\tt skeleton-with-pairbonds}.

Commands can be conditionally performed only in skeleton mode, or never
in skeleton mode (see Section~\ref{conditionals}).

\subsection{Entropy mode}

(experimental feature)

If the line has two fields and the second field is equal to the literal string
{\tt entropy}, R2R will draw each
alignment position{\textquoteright}s entropy.  In this mode, columns
are not removed by the {\tt \#=GC R2R\_LABEL line}, or if the {\tt \#=GC cons} line says
it's a gap.  Similarly, {\tt var\_hairpin} and any
{\tt var\_backbone}... commands  are ignored.  Rather, columns are only removed
if there is a dash in the special {\tt \#=GC ENTROPY\_DEL\_COLS} line.  This
is to allow you to see all the columns with their entropy, and also to
extend to more flanking sequence.

The resulting diagram is drawn twice: once with normal letters/colors
(note: gap columns are drawn with an circle that has a gray line,
rather than the usual black line.), and then with entropy.  The
calculated entropy in the file (by {\tt -{}-GSC-weighted-consensus}) is drawn.  This
entropy counts the gap character as a $5^{\text{th}}$ nucleotide.
 Therefore, the entropy $E$ is guaranteed to be in the range:  $0 \leq E
\leq \log_{2} 5$ ($\approx 2.3$).  An entropy of zero is
mapped to red, and $\log_{2} 5$ is blue.  In between these
extremes, the colors are used linearly (along the log scale).

Entropy diagrams are experimental, and the code is not guaranteed
to work.

\subsection{Cleavage diagrams}
\label{cleavage}

Cleavage diagrams are specific to in-line probing 
experiments \cite{inlines}, although the scheme might be applicable
to other structural probing experiments.
Examples of cleavage diagrams drawn using R2R are
those of the SAM-IV riboswitch \cite[Fig. 2(A)]{SAMIV} and HEARO RNA \cite[Supplementary Fig. 8(b)]{CoolRNAs}.
In {\tt oneseq} mode (see .r2r\_meta file,
above), two special {\tt \#=GR} tags are processed:

\begin{itemize}
\item {\tt DEL\_COLS} : columns with a dash
({\tt -}) will be deleted.  These are used to
specify the RNA subsequence that was actually probed.  (Note that the
{\tt R2R\_LABEL} line does not specify deletion in oneseq mode.)
\item {\tt CLEAVAGE}: nucleotides within a column are colored according to the
following code:

\begin{itemize}
\item {\tt =}  : constitutive cleavage (yellow)
\item {\tt -}  :  decreased cleavage with ligand (red)
\item {\tt +} : increased cleavage with ligand (green)
\item {\tt ?} : no data (gray)
\item Anything else : no cleavage (not shaded)
\item (note: you can also circle nucleotides with your own colors using
the {\tt circle\_nuc} command)
\end{itemize}
\item {\tt R2R\_LABEL{\dots}}, {\tt R2R\_XLABEL} : see labels below.  If present,
these {\tt \#=GR} tags override anything in the global {\tt \#=GC} tags.
Note: I find it easier to just use global {\tt \#=GC R2R\_XLABEL} lines and reference them in sequence-specific drawing commands.

\end{itemize}

\subsubsection{Multiple drawings of the same RNA molecule}

In {\tt oneseq} mode, you might have done multiple experiments with the same sequence, and want to show distinct data with each experiment.
This situation would cause a problem because the hitId that is used with {\tt \#=GR} tags would have to be the same.  

To avoid this problem,
you can add {\tt :}\textit{construct}
(where \textit{construct} is any string that specifies a given
RNA sequence in the MSA, and {\tt :} is a literal colon)
to the keywords {\tt DEL\_COLS}, {\tt CLEAVAGE}, {\tt R2R\_LABEL{\dots}} or
{\tt R2R\_XLABEL{\dots}} .  For example,

{\tt \#=GR NC\_003888.3/5-50 CLEAVAGE:L}

In the rest of the {\tt .sto} file, and in the {\tt .r2r\_meta} file, this RNA would
be called

{\tt NC\_003888.3/5-50:L}

(i.e. concatenate the hitId with the colon stuff).

Presumably the {\tt L} here stands for your
``Large'' construct.

This scenario is illustrated by the {\it yjdF} motif \cite{Weinberg104Motifs}; see {\tt demo/104/yjdF.sto} and {\tt demo/104/yjdF.r2r\_meta}.  In this case, I performed in-line probing experiments for the same {\it yjdF} RNA homolog, but with different $3^\prime$ ends.

You could also solve this problem using {\tt define}s, and draw cleavage markings explicitly
with the {\tt circle\_nuc} command, instead of using {\tt \#=GR CLEAVAGE} lines.

\section{The R2R solver cache}

This section applies to the {\tt \_solver} commands for drawing multistem junctions.  If you do not use these commands in a {\tt .sto} file, this section is not relevant.

Since the {\tt \_solver} commands often take a long time to run, I have implemented a feature to re-use already-calculated solutions.  Thus, it will be slow the first time, but will be quick in subsequent times (unless you change the parameters of the {\tt \_solver} command, in which case R2R must compute the solution for the altered problem).  In other words, I implemented the standard ``cache'' strategy.

Most likely this functionality will just work and be transparent, but it is something I've implemented very recently.
If you change parameters, and the solver does not re-run, try deleting the {\tt .solver-cache} file (see below for where it is).
Note that even parameters not part of the {\tt \_solver} command can force a solver re-run.  For example if you add a variablen-length region within the multistem junction or change certain drawing parameters (like {\tt internucleotideLen}), this indirectly changes the problem, and the solver must solve the new problem.

Stored (cached) solutions are associated with each {\tt .r2r\_meta} file you use as input.  If you process a file called {\tt something.r2r\_meta}, and there is a {\tt \_solver} command used in it, R2R will create a file called {\tt something.r2r\_meta.solver-cache}. This {\tt .solver-cache} file contains cached solutions.  The {\tt .solver-cache} will not be created if the given {\tt .r2r\_meta} file does not use any {\tt \_solver} commands.

I have attempted to make the {\tt .solver-cache} files platform independent, so that you can create a {\tt .solver-cache} file on one system, and use it on a different one.  However, this functionality is not well tested.  In principle, you should be able to build the demo files without CFSQP because of the {\tt .solver-cache} files I've provided.

\section{Labels}
\label{sec:labels}

``Labels'' allow you to identify a column or columns in the alignment by a name.
Defining column(s) by names is somewhat more convenient that using the column number,
and also means that the R2R markup remains valid even if you add or remove columns.
Therefore, R2R only supports referencing columns by name.  (Okay, actually it supports referencing columns by number -- but only do this if you're a computer.)

Tip: to see what labels are available, add an R2R command that refers to an invalid label, e.g.

\example{{\tt \#=GF R2R tick\_label INVALID dummystring}}

R2R will give you an error message that lists the labels that are valid.

\subsection{Main labels}

The main label line is {\tt \#=GC R2R\_LABEL}, which has 1 symbol per
column.  You can create multi-symbol label for each column{\textquoteright}s label
with {\tt \#=GC R2R\_LABEL\_}\textit{i} for integers \textit{i} = 1, 2,
{\dots} If there is a dot ({\tt .}) in a
column, nothing is added (i.e. the dot is the empty string).
For example, the following Stockholm file has three columns with the first two
columns having the labels
``{\tt A}'' and
 ``{\tt mul}''.:

\begin{verbatim}
# STOCKHOLM 1.0
x                AAU
y                CAG
z                GGC
#=GC SS_cons     <.>
#=GC R2R_LABEL   am.
#=GC R2R_LABEL_1 .u.
#=GC R2R_LABEL_2 .l.
//
\end{verbatim}

\subsection{Extra named label lines }

Sometimes having only one set of labels (with the {\it R2R\_LABEL} line) is not enough.
You can add additional labels with names with {\tt \#=GC
R2R\_XLABEL\_}\textit{name} for some name \textit{name}.  The {\it name} must not end with a number. 
You can create multi-symbol names by adding lines like {\tt \#=GC R2R\_LABEL\_}\textit{i}, for integers {\it i}.

The {\it name} of the label is relevant when you want to refer to the label.
Referencing labels is explained below in Section~\ref{sec:usinglabels}.

\subsection{Sequence-specific (optional)}
You can also have sequence-specific labels (in
{\textquoteleft}{\tt oneseq}{\textquoteright} mode).  In this case,
\textbf{all} labels override anything that is specified in the global
{\tt \#=GC R2R\_..LABEL}{\dots} stuff.  To specify sequence-specific labels:

{\tt \#=GR} \textit{hitId} {\tt R2R\_LABEL{\dots}  {\dots}}

Or

{\tt \#=GR }\textit{hitId} {\tt R2R\_XLABEL{\dots}  {\dots}}

(You can also emulate this behavior more conveniently by definining
special {\tt \#=GC R2R\_XLABEL\_...} label lines for each sequence.)

\subsection[SS\_cons ]{SS\_cons}
All {\tt \#=GC} lines beginning with {\tt SS\_cons} are added like named labels (see
{\textquotedblleft}extra named labels{\textquotedblright}, above).  In
this case, for {\tt \#=GC SS\_cons}, the label name is {\tt SS\_cons}. The actual
string is normalized to use only the symbols
{\textquoteleft}{\tt {\textless}}{\textquoteright},
{\textquoteleft}{\tt {\textgreater}}{\textquoteright} or
{\textquoteleft}{\tt .}{\textquoteright}.

\subsection{Using labels and special labels}
\label{sec:usinglabels}
Various commands use a parameter \textit{label}, which refers to the
label of a column(s).  The following are valid labels:

\begin{itemize}
\item {\tt pos0} : the literal string {\tt pos0} refers to the column \#0, the
$5^{\prime}$-most position.
\item {\tt all}  : specifies all positions
\item {\tt allpairs}  : specifies all positions annotated as pairing. 
If only {\tt \#=GC SS\_cons} is used, it{\textquoteright}s equivalent to the union of columns matching label
{\tt SS\_cons:{\textless}} and those matching label {\tt SS\_cons:{\textgreater}} .  If other {\tt SS\_cons} lines
are used, it applies it to them too.
\item \textit{name}{\tt :}\textit{label} : specifies all columns with the
label \textit{label }in the named-label \textit{name}.  (Note: the {\it name} and {\it label} are
separated by a literal colon).
For example, if you had a label ``{\tt e}'' in the line {\tt \#=GC R2R\_XLABEL\_fun},
you could refer to it as {\tt fun:e}.  (Note: that was a self-describing bad joke.)
The {\tt SS\_cons} lines act as named labels.  Thus, {\tt SS\_cons:\textless} refers to all columns that are a left base pair in the {\tt \#=GC SS\_cons} line.
\item \textit{label} : all columns with label \textit{label} in the
default label line ({\tt \#=GC R2R\_LABEL}).  This is a shortform for using
the fully qualified \textit{name}:\textit{label }format.  For example,
the label {\textquotedblleft}{\tt A}{\textquotedblright} is equivalent to the
label {\textquotedblleft}{\tt :A}{\textquotedblright} (note: the part before
the colon is the empty string, which is the name of the {\tt \#=GC
R2R\_LABEL} line).
\item The text {\textquotedblleft}{\tt {}-{}-}{\textquotedblright} (two minus
signs) can be added to the end of any label name, which gets the
position immediately $5^\prime$ to the label position.  Similarly,
{\textquotedblleft}{\tt ++}{\textquotedblright} is the position immediately
$3^\prime$ to the label position.  Obviously this means that labels
cannot end in {\tt -{}-} or {\tt ++}.
\item The special {\tt notinpknot} can only be used in the context of a {\tt SS\_cons} line.
For example, given {\tt SS\_cons\_1:notinpknot}, the label specifies all positions that are
(heuristically) assumed to not be a part of the pseudoknot that is assumed to be
represented by {\tt SS\_cons\_1}.  That is, all positions before the first {\tt \textless},
between the last {\tt \textless} and the first {\tt \textgreater}, and after the last {\tt \textgreater}.
This is useful for drawing pseudoknots using the {\tt define} commands (and not the {\tt SUBFAM} functionality),
in the command {\tt delcol SS\_cons\_1:notinpknot}.
%\item The special {\tt inpknot} is the reverse of {\tt notinpknot} (see previous paragraph).
\item \label{alignPosLabel}The special label form {\tt \#:}{\it col} refers to column number {\it col} in the input alignment.  Column number zero ({\tt \#:0}) is the $5^\prime$-most (i.e. first/left-most) column.

NOTE: In most cases, it is probably a mistake to use this label format, because changes to the alignment that involve adding or removing columns will cause the numeric reference to become wrong.  However, numeric labels are useful for computer scripts that generate R2R input, where the script would simply regenerate the input upon changes to the alignment.

\end{itemize}
\section{R2R commands}
Commands are given with the text

{\tt \#=GF R2R} \textit{commands {\dots}}

All parameters are space-separated with {\em exactly} one space character.  Yes, I have not implemented a robust parser.

\subsection{Conditional commands}
\label{conditionals}

\subsubsection{Define symbols}

R2R implements a simple system for conditional processing that is inspired by (but much simpler than)
the C preprocessor.
Symbols can be defined, and R2R commands can be performed only if a given symbol is defined (or not defined)
or if the symbol is equal (or not equal) to a specific value.
The other conditional processing commands given in later sections are supported for backwards compatibility with
previous structures I have drawn, but are interpreted in the context of defines.

The following commands operate on defined symbols:
\begin{itemize}
\item {\tt define} {\it name} {\it value} : sets the symbol {\it name} to the given {\it value}.
\item {\tt ifdef} {\it name} : true if {\it name} has been {\tt define}d.
{\tt ifdef} commands can apply to a single R2R command, or two multiple lines of R2R commands.
If the {\it name} value is at the end of the line, the {\tt ifdef} command will apply to subsequent commands, until
the next {\tt endif} command.  If there is text beyond the {\it name} value, then the {\tt ifdef} command will apply only to
that line.
\item {\tt ifndef} {\it name} : opposite of {\tt ifdef}.
{\tt ifndef} commands can apply to a single line or to multiple lines exactly like {\tt ifdef} commands.
\item {\tt ifdefeq} {\it name} {\it value} : true if {\it name} has the value of {\it value}.
{\tt ifdefeq} commands can apply to a single line or to multiple lines exactly like {\tt ifdef} commands.
\item {\tt ifdefneq} {\it name} {\it value} : opposite of {\it ifdefeq}.
{\tt ifdefneq} commands can apply to a single line or to multiple lines exactly like {\tt ifdef} commands.
\item {\tt else} : for a multi-line {\tt ifdef} or {\tt ifndef}, reverses the test.
\item {\tt endif} : terminates a multi-line {\tt ifdef} or {\tt ifndef}.
\end{itemize}

The following {\it name}s are predefined based on drawing options selected:
\begin{itemize}
\item {\tt oneseq} : if we're in {\tt oneseq} mode, then the {\tt oneseq} symbol is the {\it hitId} that identifies the sequence.
Otherwise, if we're drawing a consensus (the default), then {\tt oneseq} is not defined.
\item {\tt skeleton}: true if we're in skeleton mode, i.e., if either {\tt skeleton} or {\tt skeleton-with-pairbonds} was specified in the {\tt .r2r\_meta} input file.
\item {\tt skeleton-with-pairbonds} : if {\tt skeleton} was specified in the {\tt .r2r\_meta} command file, then this define symbol has the value {\tt false}.  Otherwise, if {\tt skeleton-with-pairbonds} was specified, it has the value {\tt true}.  Otherwise, it is not defined.
\item {\tt cfsqp} : defined only if the CFSQP solver is available.
\item {\tt entropy} : defined only if entropy mode is used.
\comment{
\item {\tt layout} : This symbol is not defined in advance, but is used by the {\tt set\_layout} command below.  However, this command is deprecated.
}
\end{itemize}

\subsubsection{Commands that apply to only a consensus or single-molecule drawing}
Commands given by

\example{{\tt \#=GF R2R\_allseq }\textit{commands}{\dots}}

or

\example{{\tt \#=GF R2R\_consensus} \textit{commands}{\dots}}

will only be interpreted if we{\textquoteright}re in
drawing-the-consensus-motif mode.

Commands given by

\example{{\tt \#=GF R2R\_oneseq }\textit{hitId commands}{\dots}}

will only be interpreted in {\tt oneseq} mode for the given \textit{hitId}.

Setting {\tt oneseq} or consensus mode is done in the {\tt .r2r\_meta} file
(Section~\ref{metafile}).

Note: these commands are equivalent to certain commands with {\tt define}s, as follows:
\begin{itemize}
\item {\tt R2R\_consensus} is equivalent to {\tt ifndef oneseq}
\item {\tt R2R\_oneseq NC\_003888.3/100-200} is equivalent to {\tt ifdefeq oneseq NC\_003888.3/100-200}
\end{itemize}

\comment{

\subsubsection[The layout variable]{The {\tt layout} variable}

This functionality is deprecated, and I believe it is not used in the demo files.  The {\tt define} functionality gives a more powerful syntax than the older ``layout'' functionality.

Sometimes it{\textquoteright}s not clear what the best way to layout a
motif is. In this case, you might need to write conflicting
R2R commands, and it{\textquoteright}s a hassle to switch the decision
as you think about it.  Therefore, R2R has a built in variable
layoutName, and the ability to run R2R commands only in a given layout.
It does not have a proper preprocessor (which is arguably what I should
do, except that the C/C++ preprocessor cannot be valid Stockholm
format).

{\tt set\_layout} \textit{layoutName}

sets the current layoutName.

{\tt if\_layout} \textit{layoutName}

The following commands (until the next {\tt if\_layout} command) are processed
only if the current layoutName is equal to \textit{layoutName}, or if
\textit{layoutName} is {\textquotedblleft}{\tt all}{\textquotedblright}.

Note: these commands are equivalent to others as follows:
\begin{itemize}
\item {\tt set\_layout quick} is equivalent to {\tt define layout quick}
\item {\tt if\_layout quick} is equivalent to {\tt ifdefeq layout quick}, except that no {\tt else} or {\tt endif} is required or allowed.
\item {\tt if\_layout any} is special, and not equivalent to other commands.  It cannot be mixed with {\tt define} commands.
\end{itemize}
}

\subsubsection{Other kinds of conditional commands}

If a command begins with any of the following words, the rest of the command
is ignored in certain circumstances:
\begin{itemize}
\item {\tt if\_skeleton} : the rest of the command is processed only
if skeleton drawing mode has been enabled (see Section~\ref{skeleton}).  Equivalent to {\tt ifdef skeleton}
\item {\tt if\_not\_skeleton} : reverse of above. Equivalent to {\tt ifndef skeleton}
\item {\tt if\_cfsqp} : the rest of the command is processed only if CFSQP
is available to the program.  Equivalent to {\tt ifdef cfsqp}
\item {\tt if\_not\_cfsqp} : reverse of above.  Equivalent to {\tt ifndef cfsqp}.
\end{itemize}

\subsection{Turning and positioning}

\subsubsection{Set\_dir}
\label{sec:setdir}

The {\tt set\_dir} command is used to set the orientation of the $5^\prime$-most nucleotide.  The orientation of other nucleotides are set either by R2R's default layout rules or by {\tt place\_explicit} commands.

\example{{\tt set\_dir pos0} {\it angle}}

Sets the orientation at the $5^\prime$-most nucleotide to \textit{angle}.

\example{{\tt set\_dir pos0} \textit{angle } {\tt f}}

Sets the orientation, and causes stems to be flipped (reflected along
the axis of their middle)  (This is caused by the extra
{\textquoteleft}{\tt f}{\textquoteright}.)

\subsubsection{Laying out arbitrary units like a bulge}
\label{sec:bulge}

\example{{\tt bulge}\textit{ label }}

Position the structural element starting at {\it label} as a bulge. (Note: implicitly, the bulge is placed
between the position immediately before the nucleotide at {\it label}, and the position immediately
after the drawing unit containing {\it label}.  In earlier versions of R2R, the user could set the before
\& after points explicitly, but I found this was never needed; the
added flexibility just allowed the user to draw things in the reverse
direction, and make errors.)

If N single-stranded nucleotides are laid out as a bulge from nucleotide
X to nucleotide Y, then the computer calculates a circle such that:

\begin{itemize}
\item X and Y are points on the circle
\item The distance between all consecutive nucleotides along the region
from X to Y (including the single-stranded region \textit{label}) is
the constant internucleotide length.
\end{itemize}

\example{{\tt bulge\_flip}\textit{ label }}

Same as {\tt bulge} (which was just discussed above) but draw the circle on
the other side of the line from nucleotide X to Y.

\subsubsection{Layout single-stranded loops along a straight line,
instead of circle}
\example{
{\tt layout\_straight} \textit{label}
}

This command forces drawing units that would normally be circular (bulges, internal loops and terminal loops) to be straight.  Usually the command is not necessary, because {\tt place\_explicit} commands that intersect circular drawing units will cause them to be straight.

\subsubsection[turn\_ss]{turn\_ss}
\example{
{\tt turn\_ss} \textit{label  turnAngle}
}

Turns a single-stranded region by the given \textit{turnAngle}.
The command is equivalent to the following {\tt place\_explicit} command:

\example{{\tt place\_explicit} {\it label} {\it label}{\tt --} {\it turnAngle}/2 {\tt 1 0 0 0} {\it turnAngle}}

\subsubsection{turn\_stem\_at\_internal}
\label{sec:turnsteminternal}
\example{
{\tt turn\_stem\_at\_internal}  \textit{label } \textit{dir} 
\textit{optimize-for}
}

Uses an internal loop to turn the stem 90 degrees.  \textit{label }must
identify either the $5^\prime$-most nucleotide in the $5^\prime$ part of the
internal loop or the $3^\prime$-most nucleotide in the $3^\prime$ part of the internal loop).  
\textit{dir }is either {\tt {}-1} or {\tt +1}, and this value is
multiplied by 90 and added to the current angle of the stem. 
\textit{optimize-for }is either {\tt L} or {\tt R}, which means optimize the sizes
of the circles for the bulges based on either the left or the right
sides of the internal loop.

Note: if you have problems with this, you can also try
the multistem junction commands. If it doesn{\textquoteright}t work,
try this: split the internal loop into bulges with
{\tt internal\_loop\_to\_bulges}, make the bulges as bulges-along-a-circle
with {\tt bulge}\textit{ label}, and finally use {\tt place\_explicit} to manually
set where the stem goes.

\subsubsection{disconnect\_from\_5prime}
\example{{\tt disconnect\_from\_5prime} \textit{label}}

R2R has default rules for positioning elements based on adjacent elements.
These rules are overriden by commands such as {\it place\_explicit} or
{\it bulge}.

The {\tt disconnect\_from\_5prime} command removes the implicit
positioning constraint between the structural element
at position \textit{label}, and the immediately-$5^\prime$ element.
Positioning will now happen by constraints on the other sides of the elements.

The position at \textit{label} must be the $5^\prime$-most position within the given strctural element, or the command will be ignored.

This command is useful in some very specific cases when drawing inline
style pseudoknots, where R2R's default layout is inconvenient, and requires
setting several {\tt place\_explicit} commands to force R2R to do its layout
based on the $3^\prime$ of an element (and subsequent elements).
The {\tt disconnect\_from\_5prime} command achieves this in one step.

\subsubsection{split\_ss}
\example{{\tt split\_ss} \textit{label}
}

Splits a continuous single-stranded region at internal position \textit{label}.  Without this command, the continuous single-stranded region will be treated as one unit in the structure.  The consequence of the split is that the two sides can be treated separately.  For example, one part could be positioned straight, and the other could be positioned with the {\it bulge} command.

\subsubsection{place\_explicit}
\label{sec:placeexplicit}
\example{{\tt place\_explicit} \textit{label  relativeLabel  placeAngle 
relativeVector.X relativeVector.Y  absoluteVector.X absoluteVector.Y 
finalAngle }[{\tt f}]
}

Positions a nucleotide relative to another.  \textit{label} specifies
the region that will be positioned.  Positioning is done relative to
\textit{relativeLabel}.  The position of \textit{relativeLabel} is
taken as an origin.  The following two positions are added to this
number, where all coordinates (.X and .Y) are specified in units of the
distance between consecutive nucleotides:

\begin{itemize}
\item The vector \textit{absoluteVector} is added directly
\item We form the unit direction vector in the direction
\textit{relativeLabel.angle }+ \textit{placeAngle}, where
\textit{relativeLabel.angle }is the angle of the backbone at the
nucleotide specified by \textit{relativeLabel}.  We then travel
\textit{relativeVector.X} units in this direction.  Then we rotate the
unit direction vector by +90 degrees, and travel
\textit{relativeVector.Y units}.
\end{itemize}

The angle of \textit{label }is computed by taking the angle of
\textit{relativeLabel} and adding \textit{finalAngle}.

It is possible to use {\tt place\_explicit} to place the right nucleotide of a
base pair.  In this case, remember that the backbone for nucleotides on the right side
($3^\prime$ side) of base pairs is considered to go in the opposite direction as (180
degrees from) the nucleotides on the left ($5^\prime$) side.
{\tt place\_explicit} can also be used to place the last nucleotide in a
structural unit (like the last nucleotide of a contiguous stem).

If the optional {\tt f} flag is given at the end of the line, the
left/rightness of \textit{label} is flipped relative to that of {\it relativeLabel}.

Note that {\tt place\_explicit} commands can be evaluated in both directions,
in a mathematically equivalent way.  In other words, if you position label A based on B,
there is an equivalent {\tt place\_explicit} command to position B based on A.
R2R will pick whichever layout order is convenient.

\comment{
\subsubsection{Nudging points over a bit}
{\tt offset\_pt} \textit{label  relativeVector.X relativeVector.Y}

\begin{itemize}
\item[] adds the given X and Y coordinates to what would otherwise be
the coordinates of the point \textit{label}.
\end{itemize}
}
\subsection{Layout of multi-stem junctions}
\subsubsection{Manual layout}
\label{sec:multistemjunctionbulgey}
\example{
{\tt multistem\_junction\_bulgey}  \textit{label}  \textit{[see below]}
}

This is a convenience function that allows you to render a multi-stem
junction by placing each stem in the junction relative to the previous
stem, and then having the computer render the single-stranded junctions
as bulges.  It is largely equivalent to a series of {\tt place\_explicit} and {\tt bulge}
commands.  The main benefit is that you only have to create one label
(i.e., \textit{label}).
Note that {\tt multistem\_junction\_bulgey} also provides some seldom-used functions that do not use {\tt bulge} layouts for single-stranded regions.

\textit{label }specifies the left nucleotide of the base pair that encloses the
multi-stem junction. For examples, 
see Section~\ref{sec:demomultistem}.
R2R will complain if you specify something that is not the
left nucleotide of an enclosing pair.

By default each junction within the multistem-junction is laid out as if a {\tt bulge} command
were used.
However, you can specify alternate layouts, as explained below.

After {\it label} is a series of commands.  Each command is one of the following:
\begin{itemize}
\item
{\tt J}\textit{i} \textit{anglePlace x1 y1 x2 y2 angleMove}

where {\tt J} is the literal letter and \textit{i }is a number specifying
which stem.  The enclosing stem is assumed to have a known position. 
Then the next stem is positioned relative to the left nucleotide of the
enclosing stem pair.  This next stem is positioned using {\tt J0}. Then the
next stem is positioned using {\tt J1} relative to this stem.  This continues
until we{\textquoteright}ve done all stems immediately enclosed by the
enclosing stem.  (Basically we{\textquoteright}re going around the
multi-stem junction from $5^\prime$ to $3^\prime$).  
The rest of the {\tt J}\textit{i} command is
the same as the {\tt place\_explicit} command.

For {\tt J0}, we position the left nucleotide of the basal pair of the next
stem, relative to the left nucleotide of the enclosing pair from the
enclosing stem.  For {\tt J1}, {\tt J2}, {\dots}, we position the left nucleotide
of the basal pair of a stem relative to the right nucleotide of the
basal pair of the previous stem.  This makes a bit of sense if you
think of going around clockwise, or equivalently it{\textquoteright}s
the closest column numbers in the alignment.
\item
{\tt J/base}{\it i} \textit{anglePlace x1 y1 x2 y2 angleMove}

This is the same as the {\tt J}{\it i} command just described, except that
the placement is always relative to the left nucleotide of the enclosing stem.
In other words, both {\tt J0} and {\tt J/base0} place relative to the left nucleotide of the enclosing stem,
but {\tt J/base1} also places relativive to this enclosing nucleotide, whereas
{\tt J1} places relative to the right nucleotide of the enclosing stem positioned by {\tt J0} or {\tt J/base0}.
This function is mainly used by R2R itself, so that R2R can give static layout commands that are equivalent
to computationally expensive solved layouts.
\item
{\tt b}\textit{i}
or\\
{\tt bf}\textit{i}
or\\
{\tt bs}\textit{i}
or\\
{\tt bpe}\textit{i } \textit{[place\_explicit stuff]}
or\\
{\tt bss}\textit{i}
or\\
{\tt btriangle}\textit{i}  \textit{corner}
or\\
{\tt btrianglef}\textit{i}  \textit{corner}

Here the {\textquoteleft}{\tt b}{\textquoteright} stands for
{\textquoteleft}bulge{\textquoteright}.
(It should arguably be ``j'' for junction, but that is taken for positioning
the stems surrounding the junctions.)
The bulges/junctions are numbered clockwise starting at
\textit{i}=0.
{\tt b}\textit{i}
lays out like a bulge (which is the default anyway). {\tt bf}\textit{i}
is a flipped bulge (i.e. the bulge goes the other way). {\tt bs}\textit{i}
lays the junction out along a straight line like the {\tt layout\_straight}
command (note: you{\textquoteright}re responsible for making sure the
length is good).  {\tt bpe}\textit{i} is an independent command that really only
makes sense with {\tt bs}\textit{i}, and it{\textquoteright}s a
{\tt place\_explicit} command that positions the bulge itself;
you{\textquoteright}ll usually have to use it with {\tt bs}\textit{i} since
otherwise the bulge will go 90 degrees.  A short-form for this is
{\tt bss}\textit{i}, which both says that the bulge should go straight, and
that it should be positioned straight from the base pair
(it{\textquoteright}s the same as {\tt bs}\textit{i} combined with
{\tt bpe}\textit{i} {\tt 0 1 0 0 0 0}). {\tt blinearstretch}\textit{i} says make the
bulge straight, and stretch it between the nucs it has to go. 
{\tt btriangle}\textit{i}  \textit{corner} says make the region into a
triangle (really, two linear segments), with the corner at the position
identified by the label \textit{corner }(which must be within the
bulge).  {\tt btrianglef}\textit{i} is the
flipped version, i.e., the corner point goes on the opposite side of
the direct line.
\item
{\tt bspecifyendpoints}\textit{i  relDir  beforeBulgePos.X
beforeBulgePos.Y  afterBulgePos.X afterBulgePos.Y}

Specify the positions of the endpoints of the bulge; the bulged
nucleotides will be positioned between them. The endpoint positions are
relative to the position of the left nucleotide of the enclosing base
pair in coordinates oriented at \textit{relDir}.  This is mostly useful for automated layout commands
generated by R2R's solver.
\item
{\tt bdrawcirc}\textit{i}

Says that the computer should draw the full circle related to bulge
\textit{i}, which is useful for debugging, or for doing touch-ups in
Illustrator.  If the bulge is defined such that it doesn{\textquoteright}t use a circular
bulge, R2R may fail.
\item
{\tt backbonelen} \textit{label  length}

Sets the length of the variable-length backbone identified by {\it label}  For example, these two commands would make sense in a given file:
\examplett{\#=GF R2R multistem\_junction\_bulgey J \dots backbonelen A 8
\dots\\
\#=GF R2R var\_backbone\_range A B
}
for columns labeled {\tt A}, {\tt B} and {\tt J}.
\end{itemize}

\subsubsection[Multi{}-stem junctions: automatic circular
layout]{Multi-stem junctions: automatic circular layout}
\label{sec:multistemjunctionsolver}
\example{
{\tt multistem\_junction\_circular}  \textit{label}  \textit{[see below]}

{\tt multistem\_junction\_circular\_solver}  \textit{label}  \textit{[see
below]}

{\tt multistem\_junction\_bulgecircley\_solver}  \textit{label}  \textit{[see
below]}

{\tt multistem\_junction\_bulgecircleynormals\_solver}  \textit{label} 
\textit{[see below]}

{\tt multistem\_junction\_bulgecircley\_stemnormals\_solver}  \textit{label} 
\textit{[see below]}
}

These commands are demonstrated and explained at a high level
in Section~\ref{sec:circlemultistem} or Section~\ref{sec:automultistem}.
The current section explains the details.
All of these commands attempt to find a circle on which to approximately position the nucleotides along a multistem junctions.

The parameter {\it label} must specify a column that is the left nucleotide of the base pair that encloses the multistem junction.

\textbf{Warning}: these commands often don't
work very well. 
It will fairly often produce wacky layouts with nucleotides crossing,
which happens if it cannot solve the non-linear program optimization
problem.  In rare cases, the program will crash.
(Note: the selection of a good starting point seems to be important for
the solver to solve highly non-linear problems.  However, it is difficult
to find such a starting point in an automated fashion.
The current starting value for the center of the circle seems to work well.)
With {\tt \_solver}, even slight changes in the initial fraction of stem
intersection (e.g  the number in the {\tt ai=}\textit{\#} option for stem
layout strategy) can be the difference between success and failure. 
I find the {\tt bulgecircley} commands to be more robust, but it's sometimes necessary
to try multiple versions of the {\tt \_solver} commands, and fiddle with parameters.

The commands differ as to their strategy for solution. 
{\tt multistem\_junction\_circular} uses a highly limited version of the
problem, and solves using a simple binary search over the problem
defined in one variable (the height of the circle along which the multistem junction is positioned).  With {\tt \_solver}, a
more general version of the problem can be posed in multiple variables,
and CFSQP is used. 
{\tt \_solver} will generally take
much longer to compute.

The problem the commands solve is also different.
{\tt multistem\_junction\_circular} and {\tt multistem\_junction\_circular\_solver}
 attempt to solve essentially the same problem: all nucleotides in
single-stranded junctions are
forced to be on a circle, and the computer attempts to fit the base-paired
nucleotides to
be close to the circle.  The best solution is the one with the
smoothest transition between stems and junctions.  (Mathematically, the
computer is minimizing the sum of squared deviations between the
distances of the stem nucleotides and the radius of the circle; perfect
smoothness means that the stem nucleotides will lie exactly on the circle.)

On the other hand, the commands {\tt multistem\_junction\_bulgecircley\_solver},
{\tt multistem\_junction\_bulgecircleynormals\_solver} and
{\tt multistem\_junction\_bulgecircley\_stemnormals\_solver} require that
junctions have to lie on some circle, but not necessarily all on the
same circle.  The computer then tries to find the junctions that fit a
common circle as closely as possible.  There are two specific ways to
do this, which actually seem to give similar results. With
{\tt multistem\_junction\_bulgecircley\_solver} the computer measures the
distance between points on the junction and the common circle.  The
points are evenly spaced, and so the computer is using a finite
approximation to the integral (which would be hard to compute). The
computer minimizes the mean square difference between the distances.
With {\tt multistem\_junction\_bulgecircleynormals\_solver}, the computer
attempts to minimize the differences between the normals of the common
circle curve to the normals in the bulge{\textquoteright}s circle, at
the given points.  (The normal is the vector perpendicular to the
circle at a given point.  The computer measures differences between
normals by the square of their dot products.)
{\tt multistem\_junction\_bulgecircley\_stemnormals\_solver} only checks for
the deviations of the normals at the nucleotides in a stem (i.e., at
the ends of a bulge).
Note that the {\tt multistem\_junction\_bulgecircley...} commands do not 
explicitly deal with circle intersection, and this parameter (e.g. {\tt ai}) is ignored.
However, it must still be supplied.

After the {\it label} is specified, additional commands are in a list, and each command element is as follows:
\begin{itemize}
\item {\tt s}\textit{i}  \textit{angle  strategy}

You must provide one of these options for each stem.  (The computer will
tell you if you don{\textquoteright}t.)

Tells the computer how to position stem \#\textit{i}.  The stems are
defined differently for the two implementations.  For {\tt \_solver} , stem
\#0 is the enclosing stem of the multistem junction, stem \#1 is the
first stem clockwise after the enclosing stem, and so on, going around
clockwise (or equivalently $5^\prime$-to-$3^\prime$).  For
{\tt multistem\_junction\_circular},  the enclosing stem has no number, and
always fits perfectly on the circle, and stem \#0 is the first stem
clockwise from the enclosing stem. 

\textit{angle }is the direction in which the stem points, relative to
the left nucleotide of the enclosing stem. For example,
\textit{angle}={\tt 90} means that the stem will be pointing clockwise by 90
degrees relative to the angle of the enclosing stem. In the usual
layout the enclosing stem is pointing up the page (from the outside of
the stem pointing into the multistem junction).  Thus, {\it angle}={\tt -90} is
pointing left, {\it angle}={\tt 0} is also pointing up the page, {\it angle}={\tt +90} is
pointing to the right. In the {\tt \_solver} commands, setting the direction
of stem \#0 (the enclosing stem) has the effect of rotating the other stems.

If you set an \textit{angle }that is not nicely compatible with the
tangent of the circle where the stem needs to go, things will look a
bit bad (or the computer might not be able to find a plausible
solution).  

\textit{strategy }says how to try to fit the base
(outermost pair) of the stem into the circle. \textit{strategy} is
ignored for the {\tt bulgecircley} options, but you must provide a legal
option.  Valid options for {\it strategy} are as follows:
\begin{itemize}
\item
\textit{strategy}={\tt l} (lowercase {\textquoteleft}{\tt L}{\textquoteright}) means
that the left nucleotide lies on the circle, and the right nucleotide
should go wherever it has to. 

\item\textit{strategy}={\tt r} is the reverse.  

\item
\textit{strategy}={\tt m} means that the midpoint of the outer basepair is
coincident with the circle.

\item
\textit{strategy=}{\tt aa} allows the computer to set the direction of the
stem to any angle, allowing the base of the stem to be flush with the
circle.  This makes the circle look best, but then the stems are not
likely to be horizontal or vertical.  Note: to ease the implementation you
still have to supply some number of \textit{angle}, even though this
number is ignored for \textit{strategy=}{\tt aa}.  Also, {\tt aa} is not
implemented for {\tt \_solver}, except for the enclosing stem (It could be
implemented for any stem, but there doesn{\textquoteright}t seem to be
a point.)

\item
\textit{strategy=\#}.  A number from 0-1 can be specified as the
strategy.  In this case, the circle will intersect at this point, which
is a linear interpolation from zero (left nuc) to one (right nuc). 
Thus  {\it strategy}={\tt 0} is equivalent to {\it strategy}={\tt l}, {\it strategy}={\tt 0.5} is same as
{\it strategy}={\tt m}, {\it strategy}={\tt 1} is same as {\it strategy}={\tt r}.  Note:
\textit{strategy=}{\tt aa} constrains the problem, and with {\tt \_solver} can
avoid some solutions when its applied to the enclosing stem (if you
apply \textit{strategy=}{\tt aa} to the enclosing stem, this constrains the
center of the circle to lie on the right bisector of the line between
the nucleotides of the enclosing pair, while otherwise the center of
the circle is not constrained).

\item
\textit{strategy=}{\tt ai}.  ({\textquotedblleft}ai{\textquotedblright} =
{\textquotedblleft}automatic intersect{\textquotedblright} in this
case).  Same as \textit{strategy=\#}, except the number becomes a free
variable in the optimization.  Only available for {\tt \_solver}.

\item
\textit{strategy=}{\tt ai=}\textit{\#}.  Similar to \textit{strategy=}{\tt ai}, but
in this case you specify the starting value before the optimization
(from 0 to 1).  This can be useful in cases where certain starting
values lead to infeasible parameters.

\item
\textit{strategy=}{\tt ar}.  Doesn{\textquoteright}t work well.  This is
similar to \textit{strategy=}{\tt ai}, but formulates the problem slightly
differently; the free variable is the radius of the left nucleotide to
the center of the circle, rather than the circle intersect.  I thought
it would lead to an easier function to optimize, since the anyangle
possibility should be easy to reach, where with {\it strategy}={\tt ai}, the
anyangle possibility requires going to the left or right. 
Unfortunately, {\it strategy}={\tt ar} seems to perform worse.
\end{itemize}

\item
{\tt all-stems-anyangle}

Equivalent to setting\textit{ strategy=}{\tt aa} for all stems.  Does not work with {\tt \_solver} commands, since there's no point.

\item {\tt fixed\_var\_backbone\_length} {\it label} {\it length}

Set a fixed length for a variable-length backbone, and do not allow the solver to modify its value.
{\it label} must specify the first label used in the {\tt var\_backbone\_range} command.
For example, given a command
\examplett{var\_backbone\_range X Y}
you might set
\examplett{multistem\_junction\_...\_solver ... fixed\_var\_backbone\_length X 5}
{\it length} is specified in units of {\tt internucleotideLen} (default: 0.105 inches),
and can be any real number.
Bad values (even negative values) are allowed, but will produce weird results.
\item
{\tt draw\_circ}

Tells the computer to actually draw the main circle that is the target for
layout of the multistem junction. This is useful for debugging (to see
what it{\textquoteright}s trying to do), and might be useful in
Illustrator, since it
might be helpful to have the circle that you can cut pieces out of.

\item
{\tt draw\_zero\_degrees}

Tells the computer to draw a line in the direction of 0 degrees within the multistem
junction {\_solver} command.  This is useful as a guide if your enclosing stem is directed in an
unusual direction, and you wish to align nucleotides.

\item
{\tt flipstem}  \textit{i}

Flips the \textit{i}\textsuperscript{th} stem, so that it goes inside
the multistem junction circle.  Useful for saving space.

\item
{\tt align\_stem\_horiz} \textit{ s t}

(only with {\tt \_solver}) The literals \textit{s} and \textit{t} are stem
numbers.  The solver will attempt to ensure that the midpoints of their base pairs
(that are on the multistem junction)
have the same X value (i.e. are horizontally aligned).

Warning: horizontal is defined relative to the standard coordinate system in which
the direction of the left base pair of the enclosing stem is zero degrees.  If you
are confused, add the {\tt draw\_zero\_degrees} directive.

Warning: adding alignment constraints can easily make the problem over-constrained,
and the computer might not be able to achieve all constraints, or might end up with
a completely messed-up layout.
\item
{\tt align\_stem\_vert} \textit{ s t}

Similar to {\tt align\_stem\_horiz}, but align vertically.
\item
{\tt align\_angle} \textit{angle s t}

Similar to {\tt align\_stem\_horiz}, but aligns at an arbitrary angle, specified by
{\it angle}.
The command {\tt align\_stem\_vert} \textit{ s t} is equivalent to
{\tt align\_angle 0} \textit{s t}, while 
{\tt align\_stem\_horiz} \textit{ s t} is equivalent to
{\tt align\_angle 90} \textit{s t}.

If {\it angle} is the letter {\tt h}, then alignment is horizontal.  If it's
{\tt v}, then alignment is vertical.

Mathematically, the computer is projecting the base pairs' midpoints onto the line
defined by {\it angle} using a scalar projection, and constrains the projected
positions to be the same.
\item
{\tt align\_nuc\_centers\_angle} {\it angle list1 list2}

A more general version of {\tt align\_angle}.
{\it list1} and {\it list2} each specify a set of one or more nucleotides.
The computer will calculate the midpoint (arithmetic mean) of each
of these nucleotide sets, and force the midpoints to be aligned.
The nucleotides within the sets can be any nucleotide within the multistem juction.

{\it list1} and {\it list2} are specified by a space-separated list of labels that are each terminated
by a dot (`{\tt .}').
The first thing in either list can be {\tt circle-center}, in which case the center of
the circle used in optimizing the multistem junction will be used.  {\tt circle-center}
must be first in the list if it is used.
(See {\tt demo/THF/THF.sto} for an example of its use.)

If {\it angle} is the letter {\tt h}, then alignment is horizontal.  If it's
{\tt v}, then alignment is vertical.
\item
{\tt try\_harder}

Try harder to solve the problem, by trying different initial values of certain
variables, and seeing which leads to the best solution.
The variables modified are {\tt initial\_radius} and {\tt initial\_first\_point\_angle} (see below).

For {\tt multistem\_junction\_circular\_solver}, I have encoded a few different values
that have worked in some circumstances.  The list is, of course, not complete.

For the {\tt ...\_bulgecircley\_...} commands, I have not come upon any cases
with a strong need for alternate
values.  Therefore, {\tt try\_harder} has no effect on these commands.

\item
{\tt initial\_radius} {\it number}

Specifies the circle radius (in inches) that the solver will use as an initial value in its optimization.
The starting values of variables can make a big difference in whether the optimizer
ends up with a good solution, an acceptable but not ideal solution or a complete mess.
Sometimes you can get good starting values (where necessary) from less-constrained problems
that the computer has solved.  The computer will output the optimal circle radius
after solving a problem.
The default is {\tt 0.459181}.

Note: I have considered trying to set initialization values like this using the simpler
{\tt multistem\_junction\_circular} function.  However, this function won't take into
account {\tt var\_backbone\_range} regions, whose lengths can be changed, so it won't
be a general solution.
\item
{\tt initial\_first\_point\_angle} {\it number}

(Only applicable to {\tt multistem\_junction\_circular\_solver}.)
Specifies an angle in degrees of a nucleotide immediately $5^\prime$ 
to the right nucleotide in the enclosing base pair.
The solver will use this angle as an initial value in the optimization.
The starting values of variables can make a big difference in whether the optimizer
ends up with a good solution, an acceptable but not ideal solution or a complete mess.

Sometimes you can get good starting values (where necessary) from less-constrained problems
that the computer has solved.  The computer will output the optimal first angle value
after solving a problem.
The default is {\tt 70}.
\item {\tt initial\_var\_backbone\_length} {\it label} {\it length}

Set an initial value for the variable-length backbone length, to help the optimizer reach a good
solution.  {\it label} must specify the first label used in the {\tt var\_backbone\_range}
command, as described under {\tt fixed\_var\_backbone\_length}, above.
{\it length} is specified in units of {\tt internucleotideLen} (default: 0.105 inches),
and can be any real number.
However, the solver will constrain it to be at least {\tt 1}.
The default initial length value is {\tt 2}.
\end{itemize}

All commands will optionally append to a file  a {\tt multistem\_junction\_bulgey} command that statically models the layout determined by the solver.  This will happen if the {\tt R2R\_DUMP\_FILE} shell environment variable is set, and this environmental variable specifies the path to the file.

With {\tt \_solver}, the computer will print some additional diagnostics.  It
will tell you the coordinates of the center of the circle and its
radius (mainCircle =).  Note: these coordinates are in inches, and
they{\textquoteright}re in a coordinate space where the left nucleotide
of the enclosing stem is at the origin (0,0), and the vector from this
left nucleotide to the right nucleotide is (1,0).  Up is negative in
the Y dimension.

It will also tell you the result of the objective function with the
optimal variables.  The objective function is broken into these
components:

\begin{itemize}
\item {\tt circleJoinedValue}: how close the total angles in the circle are to
360 degrees.  This should be very close to zero.
\item {\tt alignConstraintObjective}: irrelevant.  (Alignments are now specified as non-linear constraints.)
%how close to perfect any alignment
%constraints were achieved.  (Will be zero if there are no such
%constraints.)
\item {\tt changeInRadiusSum}: the main optional number.  My feeling is that
it{\textquoteright}s best if there{\textquoteright}s as few as possible
changes in the radius (distance from the circle center) that can arise
between stems and junctions.  Essentially this makes the circle as
smooth as possible.
\item {\tt rightNucFakeToActualDist}: essentially another way of looking at
circleJoinedValue
\item {\tt stem \# intersect fraction}: the final value (between zero and one)
with s\textit{\#} \textit{angle} ai option.
\item {\tt junc \#, text-col \# (raw \#) var backbone length = \#} : the
lengths chosen for var\_backbone\_range units in the junctions.  The
length units are angular, and 1 unit is equal to the rotation in angle
caused by the default internucleotide length with the given circle
radius.
\end{itemize}

The output will also show any constraints, and the value achieved for them.

\subsection{Changing layout of secondary structure}

\subsubsection{depair}
\example{{\tt depair} \textit{label [label{\dots}]}}

Assumes that \textit{label }is the left nucleotide of a base pair. 
Removes the pairing.  Useful for when one side of the pair should be a
gap, but not the other side.  \textit{label }can also specify a set of
left nucleotides (like with {\tt SS\_cons:}{\textless}). Multiple
\textit{label}s can be given, separated by spaces. For example, to
remove all the pairs in a pseudoknot corresponding to line {\tt SS\_cons\_1},
use:

\examplett{depair SS\_cons\_1:<}

\subsubsection{make\_pair}
\example{{\tt make\_pair} \textit{left-label} \textit{right-label} [\textit{ss-name}]}

Makes the nucleotide \textit{left-label} base pair with the nucleotide specified by \textit{right-label} by altering the secondary structure.  By default the alteration will happen in the primary secondary structure (i.e., {\tt SS\_cons}), but the optional \textit{ss-name} can specify an alternate secondary structure line to use.

In theory, you can specify multiple nucleotides in each of \textit{left-label} and \textit{right-label}, and the corresponding pairs (in reverse order for \textit{right-label}) will become base pairs.  Thus, again in theory, you can specify a stem at once.  However, in practice, I haven't tested this functionality.

\subsubsection[Internal\_loop\_to\_bulges]{Internal\_loop\_to\_bulges}

\example{{\tt internal\_loop\_to\_bulges} \textit{label}}

When \textit{label} is the beginning left nucleotide of an internal loop
within a hairpin, splits each side of the hairpin into bulges, so that
they can be manipulated independently.  This is also important if you
want to use the command {\tt bulge} \textit{label}.

\subsubsection{Ignore pseudoknots entirely}
\example{{\tt ignore\_ss} \textit{ssname}}

Entirely ignore the given secondary structure line.  For example,

\examplett{ignore\_ss \_1}

will ignore the pairings in the line {\tt \#=GC SS\_cons\_1}.

\textit{ssname}={\tt primary} is a special
name that corresponds to {\textquoteleft}{\tt SS\_cons}{\textquoteright}
(i.e. the primary secondary structure).  This is useful, because
Emacs/RALEE gets upset with lines that end in whitespace.  Thus,

\examplett{ignore\_ss primary}

will ignore the pairings in the line {\tt \#=GC SS\_cons}.

\subsubsection{Ignoring pseudoknots for the purposes of layout}
\label{sec:ignoressexceptforpairs}
\example{
{\tt ignore\_ss\_except\_for\_pairs}
 \textit{ssname  detail}
}

Do not position nucleotides at all based on the given secondary structure
consensus line.
The secondary structure line is specified in the same way
as with the {\tt ignore\_ss} command (see immediately above).

\textit{detail} can have the following values:
\begin{itemize}
\item {\tt outline} :  outline the nucleotides involved in
  base-pairing.  This is useful for ``callout'' layouts of pseudoknots.
\item {\tt outline-no-bpannot} : same as {\tt outline}, but ignore the {\it ssname} for the purposes of drawing covariation.

  With {\tt outline}, if {\it ssname} has covariation at a given nucleotide position, the nucleotide will be shaded green, even if the non-ignored {\tt SS\_cons} lines have no covariation.  With {\tt outline-no-bpannot}, the ignored {\it ssname} will not contribute to covariation shading.  Thus, with {\tt outline-no-bpannot}, the base pairs will be reflected only in the fact that they'll be outlined and there'll be an indication like 'pk\_1' next to them.
\item {\tt outline\_only\_bp}, this is similar to {\tt outline}, but
it only outlines the nucleotides that are really involved in a
basepair. By contrast, outline will attempt to also outline bulges.
\item {\tt ignore}, then don{\textquoteright}t do anything
with the base pairs.
\end{itemize}

\subsubsection{subst\_ss}
\example{{\tt subst\_ss}  \textit{replace-ss  into-ss}}

deletes the {\tt SS\_cons\_}\textit{into-ss }line, replacing it with whatever
is in {\tt SS\_cons}\textit{\_replace-ss}, and deleting
{\tt SS\_cons\_}\textit{replace-ss}. 
{\textquoteleft}{\tt primary}{\textquoteright} is a special name that
corresponds to {\textquoteleft}{\tt SS\_cons}{\textquoteright} (i.e. the
primary secondary structure).  This is useful, because Emacs/RALEE gets
upset with lines that end in whitespace.  Useful for drawing a
pseudoknot pairing, where the
pseudoknot pairing should be the main pairing.

In {\tt oneseq} mode, a line like {\tt \#=GR {\it hitId} SS\_cons\_{\it name}} can also specify a secondary structure line, when working with the hit {\it hitId}.  See {\tt merge\_ss} (next) for an example.

\subsubsection{merge\_ss}
\example{{\tt merge\_ss}  \textit{replace-ss  into-ss}}

Same as {\tt subst\_ss}, but merges the base pairs in, instead of clobbering. 
WARNING: the code does not check if the base pairs are compatible
(i.e., if they{\textquoteright}re pseudoknots relative to each other).

This command is useful if you want to keep a terminator as a separate
{\tt SS\_cons}, which means that when you use the {\tt \#=GC ACTUAL\_MOTIF} line,
you won{\textquoteright}t get only half of the terminator
(which would make the structure line invalid because it would have
unmatched brackets).  However, by
using {\tt merge\_ss}, you can put the terminator into the drawing with
R2R.

In {\tt oneseq} mode, a line like {\tt \#=GR {\it hitId} SS\_cons\_{\it name}} can also specify a secondary structure line, when working with the hit {\it hitId}.  For example, given the line:
\examplett{\#=GR NC\_018081.1/1633710-1634138 SS\_cons\_TT}
you can then have the command
\examplett{
\#=GF R2R ifdefeq oneseq NC\_018081.1/1633710-1634138\\
\#=GF R2R merge\_ss \_TT primary\\
\#=GF R2R endif
}
which will add a hairpin (TT=transcription terminator) to the primary secondary strcture.

\subsection{variable-length regions}
\label{sec:varlen}

\subsubsection{var\_hairpin}
\example{{\tt var\_hairpin}   \textit{leftpair  rightpair}}

Replace the terminal loop contained within \textit{leftpair } and
\textit{rightpair} with a variable-length hairpin symbol. 
\textit{leftpair} and \textit{rightpair} must be base-paired with each
other (so, I don{\textquoteright}t know why I require the user to
specify both{\dots})

\subsubsection{var\_term\_loop}
\example{{\tt var\_term\_loop}   \textit{leftpair  rightpair  message}}

\textit{leftpair } and \textit{rightpair }must be the matching base
pairs of the last base pair before the terminal loop.

Largely the same as {\tt var\_backbone\_range}, except that I think it draws a
nicer circle in the special case where you{\textquoteright}re replacing the whole terminal
loop.


\subsubsection{Variable-length backbone}
\example{{\tt var\_backbone\_range}  \textit{firstLabel  lastLabel [text]}}

Replace the nucleotides from \textit{firstLabel} to \textit{lastLabel}
 with a variable-length
backbone.  
The nucleotides within the {\it firstLabel} and {\it lastLabel}
columns are included within the variablen-length region.
All nucleotides within the variable-length region should be
single-stranded, or R2R will report an error that base pairs
are broken.
(However, you can use the {\tt ignore\_ss} command to remove pairing
that complicates the use of the {\tt var\_backbone\_range} command.
This strategy is demonstrated in the file {\tt demo/104/SAM-I-IV-variant.sto}.)

The variable-length region will be drawn with a thick black line.
The line will be labeled with
\textit{text}.  If \textit{text} contains the special string {\tt ntrange},
then the occurrence of {\tt ntrange} is replaced with the text {\textquotedblleft}x-y
{\tt nt}{\textquotedblright}, where x-y is the range of the number of nucleotides
within the variablen-length region.  If \textit{text} is not
given, then it simply gives the range ``x-y {\tt nt}''.  (This is the functionality
I normally use.)
You can include the text ``{\tt \textvisiblespace{}\textbackslash{}n\textvisiblespace{}}'' (where {\tt \textvisiblespace{}} is a single space), and this will put the remaining text on another line.

Note: if you{\textquoteright}re replacing an entire terminal loop, use
{\tt var\_term\_loop}, as it usually looks nicer.

Note: after applying a {\tt var\_backbone\_range} command, the {\it lastLabel}
will no longer work, but you can refer to the region using
the label {\it firstLabel}.

Note: in single-stranded and straight regions, the length of a
{\tt var\_backbone} is determined by the size of the text that
it{\textquoteright}s labeled with. But, in circular regions (bulges or other loops), the
{\tt var\_backbone} length is fixed (by default 3), regardless of the size of
the text.

\example{{\tt var\_backbone\_range\_if\_5prime\_nuc\_exists}  \textit{firstLabel 
lastLabel text}}

Same as {\tt var\_backbone\_range}, but only count sequences where there is at
least one nucleotide that is $5^\prime$ to the variable backbone.  This
is to account for sequences that are truncated (like environmental
sequences) that shouldn{\textquoteright}t be counted.

\example{{\tt var\_backbone\_range\_size\_fake\_nucs}  \textit{numNucs  firstLabel 
lastLabel text}}

Same as {\tt var\_backbone\_range} , but size the backbone as if it contained
\textit{numNucs} nucs.  This is silently ignored if the backbone is in
a straight region (i.e., not an internal loop, terminal loop or bulge).  {\it numNucs} can be any positive number, not just integers, but it will usually look bad if it's less than {\tt 1}.

\comment{
DEBUGGING R2R: there{\textquoteright}s an {\tt \#if 0} directive in the C++ code that
handles these; change it to a {\tt 1}.
}

\subsubsection{var\_stem}
\example{{\tt var\_stem}  \textit{leftOuterLabel  leftInnerLabel  rightInnerLabel 
rightOuterLabel}}

Replaces a stem region with a variablen-length stem.  The original stem
region is allowed to contain bulges/internal loops, but not terminal
loops.  The region is defined by the closed intervals
[\textit{leftOuterLabel},\textit{leftInnerLabel}] and
[\textit{rightInnerLabel},\textit{rightOuterLabel}], which are assumed
to pair to each other (\textit{left} pairs with \textit{right}).

It is required that \textit{leftOuterLabel }and \textit{rightOuterLabel
}must pair with each other, although this restriction could be relaxed
with improvements in the code.

\subsection{Annotation}
\subsubsection{Tick labels (like the nucleotide numbering in cleavage
diagrams)}
\label{sec:ticklabel}
\example{{\tt tick\_label} \textit{label text}}

Annotate the nucleotide identified by \textit{label} with \textit{text}.  The
\textit{text} is connected to the nucleotide with a tick mark (i.e., a short line).
Within {\it text}, you can include the text ``{\tt \textvisiblespace{}\textbackslash{}n\textvisiblespace{}}'' (where {\tt \textvisiblespace{}} is a single space), and this will put the remaining text on another line.
Thus, the command
\examplett{
tick\_label A one line \textbackslash{}n two line \textbackslash{}n red line \textbackslash{}n blue line
}
will take 4 lines.

\example{{\tt tick\_position\_label\_names}}

Useful for debugging and helping you work out where things are.  Labels
nucleotides with all valid labels (except for the {\tt SS\_cons} implicit
labels, like {\tt SS\_cons:{\textless}}).

\example{{\tt tick\_label\_regular\_numbering} \textit{start skip} [{\tt zero}] [{\tt firstNucNum} {\it num}]}

Uses the tick-style labeling to label nucleotides with their number. 
The nucleotide position \textit{start} is labeled, as is
\textit{start+skip}, \textit{start+2*skip}, {\dots}  When one sequence
is printed (like to show ligand-dependent changes in cleavage), there
is implicitly \textit{start}={\tt 0}, \textit{skip}={\tt 10}.  Note that the first,
$5^\prime$-most nucleotide is by default number 1, so in the {\tt oneseq} case, the
$5^\prime$-most nucleotide is not labeled, and the first labeled
nucleotide from the $5^\prime$ end is nucleotide number 10.  

You can change the number of the first nucleotide.  If you put {\tt zero} at the end, then the first nucleotide is numbered zero instead of one.  If you put {\tt firstNucNum} {\it num} at the end of the line, then the first nucleotide has the number {\it num}.  Putting \begin{quote}{\tt firstNucNum 0}\end{quote} at the end of the line is equivalent to using {\tt zero}.  And the line
\begin{quote}{\tt \#=GF R2R tick\_label\_regular\_numbering 0 10 firstNucNum 1}\end{quote}
is the default behavior.

Using {\tt zero} is mainly useful if you{\textquoteright}re using the
labels to aid in debugging the C++ code comprising R2R, since the R2R program uses zero-based numbers
internally.
Setting other numbers is useful if you're trying to match numbers from a previous figure.

The use of this command disables the default numbering of nucleotides in {\tt oneseq} mode.  However, you can always add this in explicitly with {\tt tick\_label\_regular\_numbering 0 10}

\example{{\tt tick\_label\_disable\_default\_numbering}}

Disables the numbering of nucleotides that is done by default when you run R2R in {\tt oneseq} mode.
\subsubsection{nobpannot}
\example{{\tt nobpannot}}

Do not shade any base pairs.  Normally, R2R will shade base pairs in consensus diagrams to annotate them as covarying, compatible, etc.

\subsubsection{Pseudo-bold fonts}
\example{
{\tt thick\_stroke\_font}  {\it label}  {\it stroke-width}
}

Draws nucleotides matching {\it label} with stroking, which has the effect of making them look bold.
(R2R cannot select bold fonts directly.)

Note: this feature has not been implemented with SVG output.  It only works with PDF.

\subsubsection{Changing nucleotide colors}
\label{sec:nuccolor}
\example{
{\tt nuc\_color} {\it label} {\it color}
}
Sets the font color of the nucleotide(s) specified by {\it label}.

\subsubsection{Circling nucleotides}
\example{{\tt circle\_nuc}  \textit{label  color }[{\tt width}\textit{ width-in-points}]}

Draws circles around all nucleotides with \textit{label}. The style of
the circling is like for cleavage diagrams.  The optional width 
directive specifies the thickness of the line forming the circle. (By
default it{\textquoteright}s the same as for the cleavage diagram.)

note1: To accommodate the circles, the size of nucleotides is reduced by a constant factor defined by
{\tt nucShrinkWithCircleNuc} (default: {\tt 0.8}), 
and the size of bond lines by a factor of 
{\tt pairBondScaleWithCircleNuc} (default: {\tt 1}, so no actual change).
To eliminate this behavior, or find alternate measurements that work, use
the {\tt SetDrawingParam} command; see Section~\ref{SetDrawingParam}.

\subsubsection{Boxing nucleotides}
\example{{\tt box\_nuc}  \textit{label  color }[{\tt width}\textit{ width-in-points}]}

Draws boxes (rectangles) around all nucleotides with \textit{label}. The
optional width  directive specifies the thickness of the line forming
the box. (By default it{\textquoteright}s the same as for the cleavage
diagram.)

\subsubsection{Outlining/inlining nucleotides}
\example{{\tt outline\_nuc}  \textit{label  color}}

Draws outline around the outside of all nucleotides with \textit{label}.
This is the same look as for drawing pseudoknots.

\textit{color} is currently ignored.

\example{{\tt inline\_nuc}  \textit{label  color}}

Same as {\tt outline\_nuc}, but draw the line on the other side.

Warning: R2R will attempt to automatically join the lines or arcs that comprise an outline.  However, sometimes its heuristics will fail, and it might be missing some lines.  In this case, try disabling the automatic joining behavior:
\examplett{SetDrawingParam outlineAutoJoin false}

\subsubsection{Boxing a set of nucleotides}
\example{{\tt box\_label}  \textit{label  X  Y  text}}

Draws a bounding box in gray around the nucleotides identified by
\textit{label}.  The box is labeled with \textit{text}, in the
direction specified by \textit{X,Y}.  {\it X} and {\it Y} are both either {\tt -1}, {\tt 0} or {\tt 1}. For example,
\examplett{box\_label L 1 0 this is a box}
would box the nucleotides identified by the label {\tt L}, and put the text ``{\tt this is a box}'' immediately to the right of the box.

\subsubsection{Shading nucleotides along the backbone}
\example{
{\tt shade\_along\_backbone}  \textit{label} [{\it label2} ...] {\it color}
}

Draws a thick line along the backbone defined by the nucleotides
belonging to any of the {\it labels} in the list.  The last parameter is always the color of the shading.
The line is thick enough to fully shade
the nucleotide letters, so its width is implicitly set by the nucleotide font size.

Because R2R is not programmed to handle all cases, there will be abrupt changes in the direction of the shading line when joining nucleotides in some directions.  These abrupt changes are usually not very noticeable.

Note: you can use the shading produced to highlight a set of nucleotides in
other ways.  The {\tt shade\_along\_backbone} command creates a complex path (series of lines 
and arcs) that can be manipulated in drawing programs.  Also, for example in Adobe Illustrator, select
the line and use the ``Outline stroke'' command (under the Object/path menu in Illustrator CS)
to convert the line to an outline shape, which you can then manipulate.


\subsubsection{Outlining a stretch of nucleotides around both ends}
\example{
{\tt outline\_along\_backbone}  \textit{label} [{\it label2} ...] {\it color}
}

This command has the same syntax and idea as {\tt shade\_along\_backbone}, except that it creates an outline around the nucleotides.

There are limitations of this command.  First, the method is somewhat of a hack: 
it simply uses the code for {\tt shade\_along\_backbone} to draw a larger shade in the outline color,
then uses the code again to draw a smaller white shade.  What's left is the outline, and there's no need to calculate the
actual outline.

The limitation is that you can't draw the outline on top of anything, because the white-out shading would obscure whatever's below.
Therefore, the outline is drawn below any covariation shading.  This often looks okay (and isn't a problem for unpaired nucleotides),
but sometimes does not look good.

One workaround is to use the {\tt shade\_along\_backbone} command, then use a drawing program to convert to an outline.  For example, 
in Adobe Illustrator the ``outline stroke'' command will do this conversion.

Note: users who are interested in implementing a more general solution, might wish to use the code in the {\tt 2geom} subdirectory of the Inkscape source code.  This code can create shapes that are unions of simpler shapes, and so it should be possible to create robust outlines of arbitrary regions.  I have not implemented this strategy in R2R, however.  A complication is that it will be necessary for the user to specify whether they want the inside of an internal/terminal loop to be a part of the shape, or not.  In general, I find it easier to use {\tt shade\_along\_backbone}, and then manipulate the shape in Adobe Illustrator.

\subsubsection{Shading the backbone in skeleton drawings}
\example{{\tt shade\_backbone\_skeleton}  \textit{label  color}}

This command is only relevant to {\tt skeleton} mode drawings.
It is similar to {\tt shade\_along\_backbone}, but shades the backbone itself.  Normally the backbone is drawn in black.

Note that the use of different colors for the backbone in skeleton drawings often does not look good because the isolated base pairs are small circles.  It can help to change the drawing style using this command:
\examplett{SetDrawingParam alongBackboneStyle 1}
(see Section~\ref{SetDrawingParam}).  The feature is used experimentally for the HEARO RNA \cite{CoolRNAs}, with output in {\tt demo/output-svg/HEARO.svg} or {\tt demo/output-pdf/HEARO.pdf}.

\subsubsection{Drawing circles associated with loops}
\example{{\tt draw\_circ} {\it label} {\it [label...]}}

Draws full circle for layout involving the nucleotide at the given {\it label}.
This command is useful for internal loops, terminal loops and bulges.
For multistem junctions, use the {\it draw\_circ} directive associated with the {\tt multistem\_junction\_circular\_...} commands.

These circles are sometimes useful in preparing other annotation in a general-purpose drawing program.

\subsubsection{Drawing direct lines between consecutive nucleotides.}
\example{{\tt straight\_line\_3prime} {\it label} {\it [label...]}}

Draws the straight line connecting the center of one nucleotide to the next nucleotide $3^\prime$ to it.

\subsection{Miscellaneous}
\subsubsection{Override default parameters for drawing}
\example{{\tt SetDrawingParam} ...}

The syntax for this command is the same as the {\tt SetDrawingParam} command in the {\tt .r2r\_meta} file, except that names and parameters are space-separated, as are all R2R command parameters within the Stockholm file.  For details on parameters, see Section~\ref{SetDrawingParam}.

Note: due to the way R2R processes files, changing a drawing parameter in one line of a file might
affect the use of that parameter in earlier lines of the file.

\subsubsection[No 5 prime label]{No $5^\prime$ label}
\example{{\tt no5}}

does not draw the $5^\prime$ label on the molecule.  Useful for drawing
pseudoknots and some modular structures.

\subsubsection{Adding G for transcription}
\example{{\tt g\_for\_transcription} \textit{number}}

Add G residues to the $5^\prime$ end of the sequence in the alignment. 
The number of Gs to add is given by  \textit{number}. Each G has an
asterisk added to it as if with the {\tt tick\_label} command. All of these
added Gs gets the special label
{\textquotedblleft}{\tt g\_with\_transcription}{\textquotedblright}, so you can refer to them by this label.
Non-genomic G nucleotides are often added to the $5^\prime$ end of RNAs because it increases the yield of {\it in vitro} transcription using T7 RNA polymerase.

\subsubsection{Keeping gap columns in the drawing}
\example{{\tt keep} \textit{label1  label2 {\dots}}}

Forces alignment positions corresponding to any of the listed labels to
be kept in the alignment, even if they would normally be considered
gaps and removed.

\subsubsection{Breaking pairs}

\example{{\tt depair} {\it label} {\it [other labels...]}}

Converts the columns matching any of the given {\it label}(s) to single-stranded.  It will be as if these nucleotides were never annotated as base paired.

For example, this command removes all base pairs in the {\tt \#=GC SS\_cons\_1} line:
\example{{\tt depair \_1:{\textless} \_1:{\textgreater}}}

\subsubsection{Explicitly setting/overriding the covariation shading}
\label{sec:setcovaryshade}

This command allows you to explicitly set the covariation shading of base pairs.  The setting will override whatever value is set in the input Stockholm-format file that was determined by a covariation analysis.

You can already modify covariation shades by modifying {\tt \#=GC} lines in the Stockholm file, but in some cases it's convenient to do this in the original .sto file.

\example{
{\tt set\_covary\_shade} {\it ssName} {\it label} {\it shade}
}

where
\begin{itemize}
\item {\it ssName} : identifies the {\tt \#=GC SS\_cons} line to modify.  The text ``{\tt primary}'' identifies the {\tt \#=GC SS\_cons} line.  To identify a line {\tt \#=GC SS\_cons}{\it something}, use ``{\it something}'' (without the quotes) as the value of {\it ssName}.
\item {\it label} : a valid label identifying positions to modify.  These positions must be the left ($5^\prime$) nucleotide of a base pair.  Modifying the right nucleotide has no effect.
\item {\it shade} :
  \begin{itemize}
  \item {\tt ?} : nothing.  This is typically not shaded.
\item {\tt 0} : no mutations.  This is typically shaded red.
\item {\tt 1} : compatible mutations.  Or R2R covariation, if R-scape is being used.  This is typically shaded blue.
\item {\tt 2} : covariation.  This is typically shaded green.
  \end{itemize}
\end{itemize}

\subsubsection{Deleting columns using an explicit command}
\example{
{\tt delcol} {\it label1} {\it label2} {\dots}
}

Deletes the columns specified.  One more more labels may be specified.
The effect of this command is equivalent to putting a dash character in the {\tt R2R\_LABEL} line, except that the 
{\tt delcol} command can be applied conditionally using {\tt ifdef} or {\tt ifdefeq} commands.

\section{Troubleshooting}

NOTE: R2R will parse the RNA structure into units, and commands only
apply to the first nucleotide in a unit.  
See Section~\ref{drawingunits}.

\begin{itemize}
\item Use the {\tt tick\_label} command to label nucleotides and make sure you
have the right one (or find out where it is), like
{\textquotedblleft}{\tt \#=GF R2R tick\_label\_regular\_numbering 1 1
zero}{\textquotedblright}
\item cyclic dependencies (very rare)

\begin{itemize}
\item try using {\tt ignore\_ss\_except\_for\_pairs} {\it SS\_cons-code} {\tt outline} for any
extra {\tt SS\_cons} lines, and see if the problem recurs.  For example, if you have {\tt \#=GC SS\_cons\_1}, use
{\tt ignore\_ss\_except\_for\_pairs \_1 outline}
\end{itemize}

\item
{\tt mark\_flip}

This R2R command marks nucleotides that are flipped (left vs. right) with ``f'', and un-flipped nucleotides with ``=''.
Only the $5^\prime$-most nucleotide in each structural unit (stem, bulge, etc.) are marked.

\item You can get R2R to show extra information about how it's using {\tt place\_explicit} commands.
See section~\ref{sec:inlinepseudoknot} for examples, and use the following command:
\examplett{\#=GF R2R SetDrawingParam showPlaceExplicit true}

\item You can get R2R to show you the direction of the backbone at each nucleotide using:
\examplett{\#=GF R2R SetDrawingParam showEachNucleotideDir true}

\end{itemize}

\section{Text output of r2r useful in debugging}

The {\tt r2r} program produces output that might help in debugging problems with your input (or indeed
the R2R code).  Sorry, there's no {\tt -{}-quiet} option.

\subsection{Start/end of file}

For each Stockholm input file, R2R will print
\example{
{\tt PROCESSING:} {\it file-name}
}
When it finishes parsing and layout of the RNA, it will print
\example{
{\tt DONE} {\it file-name}
}
Note that the actual drawing code is performed after parsing/layout of
all structures, and some error conditions are only detected in drawing.

\subsection{consensus lines and ``raw'' coordinates}

Next, R2R will draw the consensus sequence line, after columns are removed
because of gaps, explicit deletion or variable-length functions.
The codes for the consensus line (e.g. {\tt nnnRGGACGACC}) are explained
in the previous chapter for the {\tt \#=GC cons} line.
The consensus line defines the ``raw'' coordinates, which are numbered from
$5^\prime$ to $3^\prime$, starting at zero.

Next, all {\tt SS\_cons} lines are output (again after removal of columns).

\subsection{Preliminary secondary structure drawing units}

Next, R2R will dump its preliminary secondary structure units, called
``{\tt ssContextList}''.
This is immediately after parsing, and does not take into account any splitting
of drawing units that will be done later for {\tt place\_explicit} commands.

For each unit, it will print something like this:
\examplett{
[0,8:12,20) T,F,T Pair
}
In this example, the {\tt [0,8:12,20)} defines the nucleotide positions 
of the drawing unit, in raw coordinates.
{\tt [0,8:} is the left ($5^\prime$) side of the unit, and is a half-open interval.
In other words, it is the set \{0,1,2,3,4,5,6,7\}, thus including the first
number ({\tt 0}), but not including the second ({\tt 8}).
Similarly, {\tt :12,20)} defines the right side of the unit, and is also
a half-open interval.

The {\tt T,F,T} defines flags that are not important.

The {\tt Pair} defines the type of drawing unit.  The following are types:
\begin{itemize}
\item {\tt Pair} : stem
\item {\tt Outside} : single-stranded region outside a pair.  ({\tt Outside} is also
sometimes used for single-stranded regions within a hairpin that should be positioned
on a straight line, instead of on a circle.)
\item {\tt InternalLoop} : internal loop or bulge.  At this stage, also includes some outside
regions that are between (but not contained) in stems.  These will be fixed to type {\tt Outside}
later.
\item {\tt TerminalLoop}
\end{itemize}

\subsection{place\_explicit links}

The next section of the file shows how the drawing units are linked together by {\tt place\_explicit} commands, or
by implicit rules.

For example, here is one item of output:
\begin{verbatim}
ssContext
                [147,172;inf,=] {raw [0,23;109,109) }  Outside
        link
                posFrom: 22 , [147,172;inf,=] {raw [0,23;109,109) }
                posTo  : 23 , [173,177;278,282] {raw [23,28;82,87) }
                default rule
                -45 (1,0) -90
\end{verbatim}

The {\tt ssContext} line introduces a drawing unit whose links will be listed.
The {\tt raw} specification and the meaning of {\tt Outside} were explained above.

The text {\tt [147,172;inf,=]} defines the drawing unit in terms of text columns within your
input alignment file.  (As always, the left-most text column is numbered zero, to follow emacs.)
The equals sign ({\tt =}) says that the right part of the drawing unit is empty.
The range {\tt 147,172} is a doubly closed interval that includes text columns 147 and 172, and
the columns between them.

The {\tt link} introduces a {\tt place\_explicit}-style link, showing how the drawing units
can be positioned relative to one another.
{\tt posFrom: 22} specifies that raw position {\tt 22} will be the nucleotide position within
the first drawing unit.
{\tt posTo  : 23} species that position {\tt 23} is the nucleotide position within the second
drawing unit, whose nucleotide ranges are given.
{\tt default rule} means that this is the default placement, which can be overridden with a
{\tt place\_explicit} command.
{\tt -45 (1,0) -90} defines the {\tt place\_explicit}-style coordinates,
where {\tt -45} is the placement angle, {\tt (1,0)} is the relative units to travel,
and {\tt -90} is the relative angle of the next unit.
Thus, this is analogous to the following {\tt place\_explicit} command:
\example{
{\tt place\_explicit} {\it position-23} {\it position-22} {\tt -45 1 0 0 0 -90}
}

The text ``{\tt involvesCircularLayout==true}'' (within a {\tt link}) signifies that the
{\tt posFrom} or the {\tt posTo} drawing unit is an internal loop that will be laid out
along a circle.

Links relating to actual {\tt place\_explicit} commands look like this:
\begin{verbatim}
        link
                posFrom: 32 , [169,174;272,277] {raw [27,33;110,116) }
                posTo  : 36 , [181,185;198,202] {raw [36,41;51,56) }
                place_explicit demo/104/leu-phe-leader.cons.sto:67
                0 (0,0) -180
\end{verbatim}
The file name and line number of the {\tt place\_explicit} command are given.
The absolute shift coordinates are not shown.

\subsection{Selecting the place explicit commands}

Eventually, you will see lines like ``{\tt Calling PositionBackboneElement for this place\_explicit:}''.
These introduce the {\tt place\_explicit}-style links that R2R selected
to place the drawing units.
There is a special link called {\tt dummy rule to position first element}
that gives the position of the $5^\prime$-most drawing unit.

\section{Interacting with R2R using scripts}
\label{scriptedInteraction}

This section discusses feature that are useful if you want to generate R2R input by scripts, or if you want to extract layout information from R2R.

\subsection{Generating R2R input using scripts}

Scripts can explicitly select a layout for R2R using the {\tt place\_explicit}, {\tt set\_dir pos0}, {\tt bulge} and {\tt bulge\_flip} commands.  These commands are described elsewhere in this chapter and in the tutorial.  An example input file is in the directory {\tt demo/pedagogical/ScriptInputExample.sto}.

First, set the orientation of the backbone at the $5^\prime$-most position.
\example{
{\tt set\_dir pos0} {\it angle-in-degrees}
}
(see section~\ref{sec:setdir} for more on this command, and also how to flip the drawing so that stems are drawn in a mirror image.)

Then set the position of key nucleotides using the {\tt place\_explicit} command (see section~\ref{sec:placeexplicit}).  It will probably be most convenient the orient nucleotides relative to the $5^\prime$-most nucleotide and to refer to alignment positions by their numbers.  For example, to set alignment position 7 (note: the $5^\prime$-most column is numbered zero) to coordinates (4,5), oriented at an angle of 75 degrees relative to the angle of the $5^\prime$-most nucleotide, use this R2R command:
\examplett{
place\_explicit \#:7 pos0 0 0 0 4 5 75
}

Note:
\begin{itemize}
\item distance units are in terms of the (generally constant) distance between nucleotides, by default 0.105 inches.  See the {\tt internucleotideLen} drawing parameter on page~\pageref{internucleotideLen}.  Thus, by default, coordinates (4,5) in the above {\tt place\_explicit} command correspond to (0.42,0.525) inches.
\item The positive X direction is to the right, the positive Y direction is down.
\item When nucleotides X and Y are base paired, the position of Y depends on the orientation angle of nucleotide X, and the drawing parameter defining the distance between paired nucleotides (see {\tt pairLinkDist} in section~\ref{SetDrawingParam}).  Only one of the paired nucleotides should be positioned.  An exception is for pairs that are ignored (e.g., by the {\tt ignore\_ss} command); these nucleotides are not treated as being base paired by R2R for the purposes of determining the layout.
\end{itemize}

Use {\tt bulge} or {\tt bulge\_flip} (see section~\ref{sec:bulge}) to position nucleotides on a circle.  These commands will correctly set the orientation of each nucleotide in the bulge, and will consider the backbone to lie on an arc.  It is also possible to set the position of each nucleotide individually using {\tt place\_explicit} commands, but R2R will not realize that the nucleotides are on a circle.  This will affect the behavior of some commands, e.g., {\tt shade\_along\_backbone} and {\tt var\_backbone\_range}.  Currently, no command is implemented that would allow the input to specify a circle explicitly.

\subsection{Accessing layout information from R2R using scripts}
\label{sec:dumplayout}

R2R can dump information on the data it will draw, including which alignment positions were not drawn (because they're gappy or because of a user command) and the coordinates of the nucleotides.  The format is a tab-delimited file with multiple sections.

To direct R2R to output this information, set the {\tt dumpInfoFile} drawing parameter (see section~\ref{SetDrawingParam}) to the desired output file path.

The format of the file is as follows.  The file is tab-delimited, and any whitespace below is actually a tab character.  The file contains multiple sections.  Each section begins with a header line, giving a name of the section.  All headers except for the first ({\tt drawingName}) specify the number of lines contained in subsequent lines of the section.  So, to read this file, read a line to determine the section and number of lines of data, then read that number of lines, then read the next header line, etc.

An R2R output file (PDF or SVG) can contain multiple drawings in it.  For example, you might have a consensus diagram, a consensus of one or more pseudoknots and a single sequence ({\tt oneseq} mode).  All of these drawings will appear in the same dump file.  The order of sections in the file is always the same, except that the sections will be repeated for each drawing.

The sections and their order are as follows:
\begin{itemize}
\item Header line: {\tt drawingName} {\it name}

The name of the drawing whose data will come next.  The drawing name is given in at the top of each drawing in R2R's PDF or SVG output.

There are no additional data lines in this section, just the header line.
\item Header line: {\tt posToAlignCol} {\it num-data-lines}

R2R removes alignment columns from the drawing when they are gappy or as a result of some commands (e.g., {\tt var\_backbone\_range}).  This section defines the mapping from positions in the drawing to columns in the original input alignment.  The $5^\prime$-most position in the drawing is numbered zero, and the $5^\prime$-most column in the input alignment is also numbered zero.

Each data line in the section has two fields:
\begin{itemize}
\item A position index within the drawing.
\item The corresponding column in the original input alignment.
\end{itemize}
\item Header line: {\tt layout} {\it num-data-lines}

The fields in the data lines are as follows:
\begin{itemize}
\item Position index of nucleotide within the drawing.
\item The nucleotide letter (`A', `C', `G', ...).
\item X coordinate.  (Unit length is inches, which is R2R's internal unit.)
\item Y coordinate.
\item Is backbone flipped 180 degrees at this position (like mirror image, e.g., in commands like {\tt bulge\_flip})?  (1=Yes, 0=No.)
\item Does the backbone from this position $i$ to the next position $i+1$ go along a circle?  (1=Yes, along a circle, 0=No, along a straight line.)  Note: the radius of the circle is equal to the distance from the center of the circle to the position of the nucleotide.
\item If the backbone is linear (i.e., not a circle) at this position, then the orientation angle of backbone at this position, in degrees.  Otherwise undefined.
\item If backbone along a circle at this position: X coordinate of the circle's center.  Otherwise, undefined.
\item If backbone along a circle, Y coordinate of circle's center.
\item If backbone along a circle, does the next position ($i+1$) also intersect the circle?  (1=Yes, 0=No.)  In some cases, for example some multistem-junction solver results, the ends of a single-stranded junction are on a circle that does not intersect with the nucleotide at position $i+1$.  This issue only has implications if you want to trace along the backbone of the RNA.  (R2R has an ad hoc way of solving this.)
\item Is position actually a variable-length backbone symbol?  (1=Yes, 0=No.)
\item Is position actually a variable-length hairpin?
\item Is position actually a variable-length stem?
\item Is position actually a variable-length terminal loop?
\end{itemize}
\end{itemize}

\chapter{Reference: modular structures}
\label{chpt:modular}

\section{Sub-families}
You can display subsets of a family (with re-normalized conservation
stats) to:

\begin{itemize}
\item display the base pairs forming a pseudoknot separately
\item display distinct sub-types of a motif (like the GRRA/GYRA subsets
of GEMM \cite{Weinberg22Motifs,SudarsanGEMM})

Doing this requires running additional commandes; these commands could be automated using a Makefile.  Ignoring the issue
of automation, you use the script {\tt SelectSubFamilyFromStockholm.pl} to
extract the subset based on some predicate, then you calculate
conservation stats from it, then you run the generated {\tt .sto} file
through R2R.
\end{itemize}

Examples are given in Section~\ref{sec:modular}.

\section{Using
{\tt SelectSubFamilyFromStockholm.pl}}
Often pseudoknots and alternate stems cannot be drawn without breaking
the RNA backbone. In this case, you need to draw the pseudoknot as a
separate structure. The subset predicates allow you to do this.

They also allow you to draw structures that appear only in some RNA
sequences, like an optional stem.

\subsection{Command line}
To run  {\tt SelectSubFamilyFromStockholm.pl}, 

\example{{\tt perl SelectSubFamilyFromStockholm.pl} \textit{input.sto-file 
predicate-name}  {\tt {\textgreater}}  \textit{output.sto-file}}

Flags:
\begin{itemize}
\item {\tt -cutEmptyLines} : by default the script will replace lines containing rejected sequences with empty lines.  The advantage of this approach is that the line numbers in the file will remain the same.  Thus a line number reported in an error during downstream processing of {\it output.sto-file} will correspond to the same line number as in {\it input.sto-file}.  However, with this flag, blank lines will not be generated.
\end{itemize}

\subsection{How to define predicates}
\label{subfampredicates}
You define the predicate within the {\tt .sto} file.  To define a
column or columns in the alignment, use the following format:

\example{{\tt \#=GC SUBFAM\_LABEL\_}\textit{columns-name}}

Then annotate each column where an {\textquoteleft}{\tt X}{\textquoteright}
means that column is referenced by the predicate, and a dot
{\textquoteleft}{\tt .}{\textquoteright} means it is not.

There are two types of predicates:
{\textquoteleft}{\tt regex}{\textquoteright} or
{\textquoteleft}{\tt perl}{\textquoteright}.  

\subsubsection{Regex predicates}

Regex predicates match
all seqs whose selected columns match a given regular expression.  The
format is:

\example{{\tt \#=GF SUBFAM\_REGEX\_PRED}  \textit{predicate-name} [{\tt SUBFAM\_STRING} \textit{substString}] \textit{columns-name
 regex}}

where \textit{predicate-name} is the predicate you{\textquoteright}re
defining, \textit{columns-name} designates the columns referenced
(which you defined using {\tt \#=GC SUBFAM\_LABEL\_}\textit{columns-name})
and \textit{regex} is a valid regular expression in Perl.  

[{\tt SUBFAM\_STRING} \textit{substString}] : is optional, if you give it, it will use the alternate string for substituting in changing the file (See \ref{sec:predmod}).  If \textit{substString} is set to {\tt primary} , this will retain the primary modular structure, i.e., it will not modify the file


For example:

\example{{\tt \#=GC SUBFAM\_LABEL\_GNRA  ............XXXXXXX......... }}

(don{\textquoteright}t take this literally; see the {\tt GEMM.sto} file from
Sudarsan {\it et al.} for specifics)

\example{{\tt \#=GF SUBFAM\_REGEX\_PRED GYRA GNRA G[CU][AG]A[.][.][.]}}

In this example, we define a set of columns
{\textquoteleft}{\tt GNRA}{\textquoteright}, which is the 7 nucleotides
positions (columns) in the {\tt GNRA} tetraloop of GEMM RNAs.  (Sometimes
there are more nucleotides in the loop, so that{\textquoteright}s why
there are 7 positions.)  Then, we define a predicate
{\textquoteleft}{\tt GYRA}{\textquoteright} which sees if those 4 nucleotides
match GYRA.

\subsubsection{Perl predicates}

Perl predicates allow you to use arbitrary perl expressions, which are
convenient for Boolean logic.  The syntax is:

\example{{\tt \#=GF SUBFAM\_PERL\_PRED}  \textit{predicate-name}  [{\tt SUBFAM\_STRING} \textit{substString}]  \textit{perl-code}}

Currently there aren{\textquoteright}t that many things you can do with
the perl code (without implementing additional capabilities within
{\tt SelectSubFamilyFromStockholm.pl}).  
The {\it perl-code} used in defining a Perl predicate should end with a {\tt return} statement.
At the moment, the {\it perl-code} can reference the following variables:
\begin{itemize}
\item associative array {\tt \%predValue} (  {\tt \$predValue\{}\textit{predicate-name}{\tt \}} ) :
evaluates to true or false, where {\it predicate-name} is the name of a predicate.  {\it predicate-name} can refer to either a ``Regular expression'' or a ``Perl'' predicated, but the referenced predicted must have been defined earlier in the {\tt .sto} file than when it is used, otherwise its value will be undefined.
\item associative array {\tt \%labelToSeq} ( {\tt \$labelToSeq\{}\textit{label-name}{\tt \}} ) :
for the given label, evaluates to the current sequence's nucleotides in the columns associated with the label.  Thus, for example, the regex predicate (following the example in Figure~\ref{fig:modular})

[{\tt SUBFAM\_STRING} \textit{substString}] : is optional, and has the same effect as in the {\tt REGEX} case.

{\tt \#=GF SUBFAM\_REGEX\_PRED HAS\_GNRA TERM G[A-Z][AG]A}

is equivalent to this perl predicate

{\tt \#=GF SUBFAM\_PERL\_PRED HAS\_GNRA return \$labelToSeq\{TERM\} =\textasciitilde{} G[A-Z][AG]A;}
\item scalar {\tt \$hitId} : the identifier of the hit within the original Stockholm alignment file.
\item scalar {\tt \$seqId} : the identifier of the sequence, but only if the {\tt \$hitId} is in the Rfam format, seqId/start-end.  Otherwise, this variable will have the value {\tt undef}.
\item scalar {\tt \$start} : the nucleotide number of the $5^\prime$ end of the hit, but only if the {\tt \$hitId} is in the Rfam format, seqId/start-end.  Otherwise, this variable will have the value {\tt undef}.
\item scalar {\tt \$end} : the nucleotide number of the $3^\prime$ end of the hit, but only if the {\tt \$hitId} is in the Rfam format, seqId/start-end.  Otherwise, this variable will have the value {\tt undef}.
\end{itemize}

NOTE: predicate names cannot contain an underscore
({\textquoteleft}{\tt \_}{\textquoteright}).  (The problem is that it would
be ambiguous in the {\tt SUBFAM\_}\textit{pred}{\tt \_}\textit{label} substitution
code.)

For example,

\example{{\tt \#=GF SUBFAM\_PERL\_PRED other return !(\$predValue\{GYRA\}
{\textbar}{\textbar} \$predValue\{GRRA\});}}

Perl predicates also have access to any {\tt \#=GS} tags for the current hit,
in the map {\tt \%gsTag}.  This feature is used for GOLLD (see {\tt demo/exceptional/GOLLD.sto})
to eliminate sequences with regions
that could not be structurally aligned given the limits of other sequences available for comparative analysis.

The variable {\tt \%labelToSeq} is a hash mapping label lines (specified with {\tt \#=GC SUBFAM\_LABEL\_}{\it ...}) to the sequence of nucleotides or gaps for the current sequence.

The following functions are defined:
\begin{itemize}
\item {\tt IsWatsonCrickOrGU({\it \$X},{\it \$Y})}, where {\it \$X} and {\it \$Y} are strings containing a single character each (i.e., strings of length 1).  Returns true iff {\it \$X} and {\it \$Y} correspond to a Watson-Crick or a G-U base pair.  (If either character is a gap, then the function returns false.)
\item {\tt IsGap({\it \$X})}, where {\it \$x} is a single-character string (of length 1).  Returns true if {\it \$x} is a gap symbol, which is interpreted broadly.  {\it \$x} is a gap if it is not an upper- or lowercase letter.
\end{itemize}

\subsection{How the file is modified by applying a predicate}
\label{sec:predmod}
Obviously all sequences that match the predicate will be retained, but
all sequences that do not match the predicate will be removed.  Also
the {\tt \#=GR} and {\tt \#=GS} tags for discarded sequences will also
be discarded (caveat: if you use the {\tt SUBFAM\_KEEPALL} to force keeping a hit, things might get inconsistent due to a bug in the script).

Also, all lines that begin 
\example{{\tt \#=GC R2R}}
or 
\example{{\tt \#=GF R2R}}
are removed.

Lines of the form 
\example{{\tt \#=GC
SUBFAM\_}\textit{predicate-name}{\tt \_}\textit{other-stuff  text }}
have the
{\textquotedblleft}{\tt SUBFAM\_}\textit{predicate-name}{\tt \_}{\textquotedblright}
part removed, and same thing for 
\example{{\tt \#=GF
SUBFAM\_}\textit{predicate-name}{\tt \_}{\it other-stuff  text}}

The special code

\example{{\tt \#=GF SUBFAM\_KEEPALL 1}}

means that all subsequent lines will be kept no matter what.  This is
turned off with

\example{{\tt \#=GF SUBFAM\_KEEPALL 0}}

Also,

\example{{\tt \#=GF SUBFAM\_}\textit{predicate\_name}{\tt \_KEEPALL}}

keeps lines for the given predicate.

To show pseudoknot pairs as a special other diagram, just define a
predicate that matches everything, but use the modifications to define
a new {\tt \#=GC R2R\_LABEL} line, and probably other {\tt \#=GF R2R} commands.

You can also override the substitution string using

\example{{\tt \#=GF SUBFAM\_SUBST\_STRING}  \textit{predicate-name other-string}}

In this case, the script will look for \#=GC
{\tt SUBFAM\_}\textit{other-string}{\dots} instead of {\tt \#=GC
SUBFAM\_}\textit{predicate-name}{\dots} (and same for {\tt \#=GF}).  This is
useful for multiple predicates that all get drawn the same way.  

{\bf Note}:
the {\tt SUBFAM\_SUBST\_STRING} command must precede the references -- the
computer does not do multiple passes through the file.

\chapter{Meta-Makefile}
\label{sec:metamake}

This chapter explains the use of the {\tt MetamakeDemos.pl} script, which helps to automate the application
of the {\tt r2r} command.
This chapter assumes that you know what a Makefile is and basically what the {\tt make} command does.
It is not necessary to use the {\tt MetamakeDemos.pl} script, since {\tt r2r} commands can be given at
the UNIX command prompt directly.

The {\tt MetamakeDemos.pl} script searches for {\tt .sto} files and generates appropriate {\tt .r2r\_meta}
files, and a {\tt Makefile} to build PDF or SVG output.
In order to create {\tt .r2r\_meta} files, in some cases you must add additional markup to your {\tt .sto} files.
This directs the script to add {\tt oneseq} mode lines, or modular sub-structures, etc.

You can run the resulting Makefile as follows.  To create all PDF output files:
\examplett{
make all-pdf
}
To create all SVG output files:
\examplett{
make all-svg
}

\section{Running the script}

To run the script, you need to designate a directory where you will keep your {\tt .sto} files and {\tt cd} into that directory.

The script is normally run like this:
\examplett{
perl MetamakeDemos.pl
}
and must be run to initialize the directory and create a Makefile.

It searches the current directory and its subdirectories for {\tt .sto} files, using the standard
UNIX {\tt find} command.
However, it does not search the {\tt intermediate} subdirectory (because that is where intermediate files generated by the Makefile will go).

It will create a new file in the current directory that by default is called  {\tt Makefile}.
It will also create an appropriate {\tt .r2r\_meta} file corresponding to each {\tt .sto} file it sees.

\subsection{Script options}
\label{sec:Makefile:perloptions}

The script accepts the following command-line options, which are not normally used:
\begin{itemize}
\item {\tt -srcFileList} {\it filename} : directs the script not to search for {\tt .sto} files.  Instead {\it filename}
gives the list of input {\tt .sto} files.  Each line in {\it filename} specifies an input file, and is tab-delimited.
The first field is an actual source {\tt .sto} file.  For convenience of Windows/CygWin users, any backslashes in the file name
are converted to frontslashes.
The second field in the line is a name for the file.  The Makefile will copy the source {\tt .sto} file into a new
file called {\tt intermediate/}{\it name}{\tt .sto}, where {\it name} is the second field.
\item {\tt -makefile} {\it filename} : creates the new Makefile with the name {\it filename}.
(The default is {\tt -makefile Makefile}.).
\item {\tt -r2r} {\it executablefilename} : specifies the command to run for {\tt r2r}. (Default is {\tt -r2r ../src/r2r}.)
\item {\tt -subfam} {\it filename} : references the script {\tt SelectSubFamilyFromStockholm.pl}.  (Default is {\tt -subfam ../src/SelectSubFamilyFromStockholm.pl}.)
\item {\tt -noDumpFile} : do not get {\tt r2r} to dump solutions of multistem-junctions.
\item {\tt -noall} : do not create an {\tt all} target in the new Makefile.  This is useful if you want to {\tt include}
the Makefile generated by {\tt MetamakeDemos.pl} into some larger Makefile.
\item {\tt -doStoTex} : experimental feature.
\item {\tt -doStoOutput} : experimental feature.
\item {\tt -pubsto} {\it perl-file} : experimental feature.
\item {\tt -cleansto} {\it perl-file} : experimental feature.
\item {\tt -covaryWithRscape} : experimental feature.
\item {\tt -covaryUseSsConsWithRscape} : experimental feature, pass {\tt -s} to R-scape to get it to restrict the base pair analysis to the base pairs in {\tt \#=GC SS\_cons}.
\item {\tt -setDrawingParam "{\it params}" } : set drawing parameters (see Section~\ref{SetDrawingParam}) to different values for all files that the script is aware of.  This effectively changes the default values from those that are hardcoded in the R2R source code.  The parameter names and values should be separated by space characters.  You'll usually need to put this value in quotes because of the spaces.  For example:

  \examplett{-setDrawingParam "varHairpinNumFakePairs 2 varTerminalLoopRadius 0.14"}
\item {\tt -setDrawingParamFile} {\it file-name} : similar to {\tt -setDrawingParam}, but get the parameters from a file.  The parameter names and values should be separated by tab characters (this is different from the case with the {\tt -setDrawingParam} flag).  The file must consist of one line (but additional, blank lines are permitted).
\item {\tt -inkscapeHelveticaFontName} {\it fontName} : use this font name in output SVG files.  Inkscape is sometimes happier if this name matches a font that it has.  The default is ``{\tt Bitstream Vera Sans}''.  This flag arranges for the commands to run {\tt r2r} to add ``{\tt -{}-inkscape-helvetica-font-name} {\it fontName}''.
\end{itemize}

\subsection{Makefile targets}

The script generates a Makefile with the following targets:
\begin{itemize}
\item {\tt all} : equivalent to {\tt all-pdf} and {\tt all-svg}.  Not recommended in a multi-processor make session because of possible interactions between the solver.  Disabled by the {\tt -noall} flag.
\item {\tt all-pdf} : generate all PDF output files.
\item {\tt all-svg} : generate all SVG output files.
\item {\tt clean} : remove intermediate and output files (but keep the {\tt .r2r\_meta} files and {\tt Makefile} that were created by the script.).
\item {\tt clear-solver-cache} : remove cache files used by the CFSQP solver.
\item {\tt solver-pdf} : generate only PDF output files that use the CFSQP solver (and maybe a few that don't, but fool the script).  Used for debugging.
\item {\tt solver-svg} : analogous to {\tt solver-pdf}, but for SVG output.
\item {\tt concat.pdf} : generate {\tt all-pdf} and concatenate the output files together so you can
look at them quickly.  Mainly just useful for debugging.  Requires Ghostscript's {\tt gs} command, which is not distributed
with R2R.
\item {\tt solver.pdf} : same as {\tt concat.pdf}, but only concatenate the PDF output specified by the {\tt solver-pdf} target.
\end{itemize}

\section{Additional markup for .sto files}

The script creates an {\tt .r2r\_meta} file for each {\tt .sto} input file.
The {\tt .r2r\_meta} file will always include the {\tt .sto} file in R2R's default consensus mode,
but can include other types of runs.

\subsection{Implicit rules}

If you use a {\tt \#=GR ... DEL\_COLS} tag for any sequence in the alignment,
the script will draw that sequence in {\tt oneseq} mode (adding a line to the {\tt .r2r\_meta} file).

If you use a {\tt \#=GF SUBFAM\_PERL\_PRED} or {\tt \#=GF SUBFAM\_REGEX\_PRED} directive,
the script will instantiate the subfamily alignment using {\tt SelectSubFamilyFromStockholm.pl} (adding a command
to the Makefile),
and draw it (adding a line to the {\tt .r2r\_meta} file).
However, you can disable this on a per-predicate basis by adding the line
\example{
{\tt \#=GF Makefile disablepred} {\it predname}
}

\subsection{Explicit rules}

You can explicitly direct the creation of a new type of drawing within the {\tt .r2r\_meta} file as follows.
This also allows you to {\tt define} additional symbols for use with R2R's {\tt ifdef} and related commands.
Note that all fields are space-separated in this file, but the spaces will be converted to tab characters
for the {\tt .r2r\_meta} file.
The file {\tt demo/104/crcB.sto} contains a particularly overwrought array of example usages
of these directives.

\subsubsection{Oneseq}

\example{
{\tt \#=GF Makefile oneseq} ...
}
Directs the script to make a {\tt oneseq}-mode output.  For example:
\examplett{
\#=GF Makefile oneseq NC\_003888.3/314159-2653589 define alt-style 1
}

\subsubsection{Skeleton}

\example{
{\tt \#=GF Makefile skeleton}...
}
Directs the script to make a {\tt skeleton} or {\tt skeleton-with-pairbonds} drawing.
As above, additional {\tt define} commands can be given.

\subsubsection{Defines for consensus diagrams}

\example{
{\tt \#=GF Makefile define} ...
}
Directs the script to make a drawing of the consensus with additional {\tt define} commands.

\subsubsection{Using define directive with modular structure}
\example{
{\tt \#=GF Makefile pred} {\it predname} ...
}
Directs the script to draw a subfamily defined by predicate {\it predname}, with additional {\tt define} commands (after {\it predname}).
The predicate {\it predname} must be defined earlier in the file.

\subsubsection{Oneseq for a predicate defining a modular structure}

\example{
{\tt \#=GF Makefile oneseqpred} {\it hit} {\it predname} ...
}
Directs the script to draw a subfamily defined by the predicate {\it predname}
in {\tt oneseq} mode for the sequence named {\it hit}.  Optionally, {\tt define}
directives can appear after {\it predname}.
The predicate {\it predname} must be defined earlier in the file.

\subsubsection{Using a different sequence-weighting algorithm}
\label{makefileseqweighting}

{\bf WARNING:} In general, the GSC algorithm seems to be the best available sequence-weighting method.  Please read section \ref{whyweights}, which explains the importance of sequence weighting.

By default, R2R will weight sequences using the GSC algorithm.  However, with very large alignments, this algorithm can be slow.  (In computer science terms, it its time complexity is $O(n^3)$, for $n$ sequences in the alignment.)

To make this faster, especially if precise weighting is not important, you can change this algorithm, by causing the Makefile script to give R2R other flags.

If you add the following to the {\tt .sto} file:
\example{\tt
  \#=GF Makefile\_Weight\_Flag -{}-position-based-weights
}
then R2R will use a much faster weighting algorithm.

You can pass arbitrary flags (in place of {\tt -{}-position-based-weights}), which will be added between {\tt r2r} and {\tt -{}-GSC-weighted-consensus} on the command line.

\subsubsection{Weighting sequences using your own algorithms with MetamakeDemos}
\label{makefilemanualweighting}

{\bf WARNING:} please read section \ref{whyweights} on why you probably should not use this functionality.

To use manual weighting for a given Stockholm file, add a line like
\example{
{\tt \#=GF Makefile\_SeqWeighting} {\it command}
}
where {\it command} is a UNIX command to create a new Stockholm file with the weights.
Within {\it command}, the special text {\tt INPUTSTO} will be changed to the input Stockholm file, and similarly
{\tt OUTPUTSTO} will be changed to the output Stockholm file.
The result of the command should add the {\tt \#=GF USE\_THIS\_WEIGHT\_MAP} tag to the output Stockholm file,
as described in Section~\ref{manualweighting}.
If your Stockholm file already has a valid {\tt \#=GF USE\_THIS\_WEIGHT\_MAP} tag in it, you can use the following line:
\examplett{
\#=GF Makefile\_SeqWeighting cp INPUTSTO OUTPUTSTO
}
which just copies the input file (that already has the tag) to the output file.

\chapter{R2R source code}

\section{Summary of C++ and Perl source code files}

\begin{itemize}
\item {\tt AdobeGraphics.cpp} : Implements generic aspects of interface to draw in graphical files such as PDF-format files.
\item {\tt AdobeGraphics.h} : Interface to draw in graphical files.
\item {\tt AdobeGraphicsLayout.cpp} : Implements generic utility classes to assist in drawing and layout.
\item {\tt AdobeGraphicsLayout.h} : Header file for above.
\item {\tt AdobeGraphicsPdfLike.cpp} : Generic class to factor out similarities between drawing in PDF and SVG.
\item {\tt AdobeGraphicsPdfLike.h} : Header file for above.
\item {\tt Cm2HmmOptimize\_cfsqp.cpp} : Implements {\tt SolverWrapper} interface (see {\tt Optimize.h}) for CFSQP.  (File is originally from my earlier rigorous filter work.)
\item {\tt CommaSepFileReader.cpp} : Reading of files delimited by a single character (e.g., commas).  It is used to read tab-delimited {\tt .r2r\_meta} files and space-separated {\tt .sto} files.
\item {\tt CommaSepFileReader.h} : Header file for above.
\item {\tt GSCConsensus.cpp} : Implements the {\tt -{}-GSC-weighted-consensus} functionality.  The GSC algorithm itself is implemented by the code in {\tt NotByZasha/infernal-0.7/squid}.
\item {\tt LabelsAndProjection.cpp} : Functions related to parsing {\tt .sto} files related to projection of columns in an alignment or labeling of columns.
\item {\tt MiscExceptions.cpp} : Generic exception classes, for error handling.
\item {\tt MiscExceptions.h} : Header file for above.
\item {\tt NoUnderflowDouble.h} : Implements real numbers that have a greatly expanded range for their exponent.  Used by code related to CFSQP.
\item {\tt Optimize.cpp} : Generic functions for non-linear optimization.
Also implements most of the functionality needed for the solver cache.
\item {\tt Optimize\_cfsqp.cpp} : Implements a non-linear optimization solver for CFSQP.
\item {\tt Optimize\_nltop.cpp} : Implements a non-linear optimization solver for NLOPT.
\item {\tt Optimize.h} : Header file for above.
\item {\tt ParseOneStockholm.cpp} : Functions for interpreting R2R commands in a {\tt .sto} file.  Note: parsing happens in three passes.  Pass 1, in which the alignment and column-annotation data is read into data structures, is in the function {\tt OneStockholm\_try\_Pass1}.  The function {\tt OneStockholm\_try} implements passes 2 (process commands that remove nucleotide positions, e.g. {\tt var\_backbone\_range}) and 3 (all other commands).
\item {\tt ParseSs.cpp} : Functions related to parsing RNA secondary structures in bracket notation (i.e. the {\tt SS\_cons} lines).
\item {\tt PdfGraphics.cpp} : Drawing into PDF files.  Concrete implementation of {\tt AdobeGraphics} interface.
\item {\tt PdfGraphics.h} : Header file for above.
\item {\tt PositionBackbone.cpp} : Implements most functions for determining the position of nucleotides within a drawing.
\item {\tt PositionBackbone\_MultiStemCircular.cpp} : Implements various methods to draw multistem junctions, except solvers.
\item {\tt PositionBackbone\_MultiStemCircularSolver.cpp} : Implements solvers for multistem junctions.
\item {\tt R2R-Utils.cpp} : Utility functions related to parsing {\tt .sto} files and positioning nucleotides.
\item {\tt R2R.cpp} : Contains {\tt main} function, with top-level functions.  Sets up default values, parses {\tt .r2r\_meta} files, dispatches to other code.
\item {\tt R2R.h} : Declares data structures, classes and functions used by multiple files.
\item {\tt RnaDrawer.cpp} : Classes to actually draw RNA diagrams, based on already-determined nucleotide positions.
Includes code to find paths for the {\tt shade\_along\_backbone} command or {\tt skeleton} mode drawings.
\item {\tt SelectSubFamilyFromStockholm.pl} : Perl script to support extraction of a specified subset of sequences from a {\tt .sto} file, for modular structures (Chapter~\ref{chpt:modular}).
\item {\tt SvgGraphics.cpp} : Drawing into SVG files.  Concrete implementation of {\tt AdobeGraphics} interface.
\item {\tt SvgGraphics.h} : Header file for above.
\item {\tt SymbolicMath.cpp} : Implements symbolic expressions, used for CFSQP-based solver of multistem junctions.
\item {\tt SymbolicMath.h} : Header file for above.
\item {\tt UseDebugNew.h} : Header file used in debugging.
\item {\tt multiDimVector.h} : STL-style classes for multi-dimensional arrays.
\item {\tt stdafx.h} : Include standard system headers.
\item {\tt vectorPlus.h} : Extension of STL {\tt vector} class to detect array bounds errors.
\end{itemize}

\section{Other information}

\subsection{Overall layout of RNA}

To position the nucleotides in an RNA structure, the structure is first decomposed into structural elements: consecutive base pairs, internal loops or consecutive unpaired nucleotides. Default constraints define each element's orientation relative to another element. To apply user-defined constraints in an intuitive manner, a graph is defined whose nodes are structural elements and whose edges correspond to constraints. This graph's minimum spanning tree, with user-defined constraints prioritized over default constraints, yields the positioning constraints to apply. However, some rules direct nucleotides to be drawn on a circle connecting two already-positioned nucleotides, and these rules are applied after the nucleotides are positioned.

Constraints in the graph correspond to {\tt place\_explicit} commands.
This drawing order logic is applied after any {\tt multistem\_...} commands have been resolved into {\tt place\_explicit} commands, and is implemented in the file
{\tt PositionBackbone.cpp}.

\subsection{Inferring a path for the backbone}

In order to draw a {\tt skeleton}-style schematic drawing, or for shading consecutive nucleotides (e.g., using the {\tt shade\_along\_backbone} command), it is necessary to determine a path corresponding to the RNA's backbone. In R2R, this path is composed of straight lines or arcs. The path largely follows the lines and arcs used to position nucleotides.

Except when certain commands are used, all nucleotide positions in R2R are laid out either on a straight line, or along a circle, where each nucleotide is centered on the endpoint of the line/arc. For each position, the computer stores the line or circle defined at that point. However, it is not sufficient to use these path segments to connect the diagram, as they are only defined at the nucleotide itself and not for the connection to the next nucleotide.  Moreover, even when two nucleotides are connected, the connection might not be aesthetic.

Some nucleotide positions are not really points.  For example, the {\tt var\_backbone\_range} command results in a line or arc, even though it is represented as a nucleotide position.  Therefore, the first step is to expand nucleotides into separate points.  We call these anchor points, by analogy with the drawing of complex curves in programs such as Adobe Illustrator.

Thus, we are given a set of anchor points $a_1,a_2,\ldots,a_n$.  Each anchor point $a_i$ is associated with a specific position on the page, and is also associated with either a line or an arc that intersects that position.  However, the line or arc does not necessarily intersect $a_{i-1}$ or $a_{i+1}$.
We wish to find connectors from $a_i$ to $a_{i+1}$ for all $i$.

If $a_i$ and $a_{i+1}$ are both associated with lines and if the lines are coincident (i.e. the lines are an extension of one another), then we connect $a_i$ and $a_{i+1}$ with that line.  Similarly, if they are both associated with arcs that are coincident, we connect the anchor points with that arc.

If $a_i$ is associated with an arc, and if the arc of $a_i$ intersects the position of $a_{i+1}$, then we connect $a_i$ to $a_{i+1}$ using that arc.  In this case, it does not matter whether $a_{i+1}$ is associated with a line or an arc.  Analogous logic applies to the case where $a_{i+1}$ is associated with an arc, and $a_i$ is associated with a line, and the arc of $a_{i+1}$ intersects the position of $a_i$.

The arc of $a_i$ might not intersect the position of $a_{i+1}$.  For example, the {\tt multistem\_junction\_circular\_solver} leads to bulges that are drawn on a circle but do not (in general) intersect the adjacent paired nucleotides.  In cases like this, we declare a join error.  A join error denotes cases where R2R is not implemented with logic to join the anchor points nicely.  In the event of a join error, $a_i$ is connected to $a_{i+1}$ with a straight line.  In practice, these straight lines are not very problematic, because join errors are relatively rare, and because the deviation of the straight lines from an ideal connector is not usually very visible at the scale at which diagrams are drawn.

If anchor points $a_i$ and $a_{i+1}$ are associated with lines, we calculate their intersection.  If the intersection is equidistant from the positions of $a_i$ and $a_{i+1}$, then we join $a_i$ to $a_{i+1}$ using an arc that runs through $a_i$ and $a_{i+1}$.  The radius of the arc is the distance between $a_i$ (or equivalently $a_{i+1}$) and the intersection point.
This case frequently arises in R2R's default layout of stems (e.g., at the $5^\prime$ end of the RNA in Figure~\ref{fig:demo1iii}), or when the {\tt turn\_ss} command (or equivalent {\tt place\_explicit} command) is used.
If the intersection point is not equidistant from the two anchor points, we declare a join error.  In this case, it is likely possible to extend one of the lines associated with $a_i$ or $a_{i+1}$ in order to lead to an equidistant intersection, but R2R does not implement this logic.

This functionality is implemented in {\tt RnaDrawer.cpp}.

\index{\texttt{.sto}|see{Stockholm file format}}
\index{R2R commands!\texttt{\#=GC R2R\_LABEL}|see{Labels}}
%\chapter{Index}
%\printindex

\bibliographystyle{abbrv}
\bibliography{bio}
\end{document}
