\clearpage
\section{Options}
\label{section:options}
\setcounter{footnote}{0}

The whole list of options can be found using 

\user{R-scape -h}\\

Some important options are:
\subsection{Covariation statistic options}

\subsubsection{\prog{-E <x>}}  Target E-value is $x\geq 0$.

\subsubsection{\prog{--GT, --MI,  --MIr, --MIg, --CHI, --OMES, --RAF, --RAFS, }}
We favor the G-test covariation statistic, but a total of eight
covariation statistics are currently implemented in \rscape. For each
covariation statistic (GT, for instance), \rscape\ can also calculate
its average product correction (GTp) and its average sum corrections
(GTa). For each option above, appending ``p'' or ``a'' chooses one of the
corrections. For example, \prog{--GT} does the G-test statistic,
\prog{--GTp} does the APC-corrected G-test statistic, \prog{--GTa}
does the ASC-corrected G-test statistic.\\

The \rscape\ default is \prog{--GTp}.\\

Details of the definition and provenance of the different covariation
statistics can be found in the \rscape\ manuscript: Rivas, E. \& Eddy
S.~E., \textit{``A statistical test for conserved RNA structure shows
lack of evidence for structure in lncRNAs''}.

\noindent
In a nutshell, given two
alignment columns $i,j$,
%
\[
\begin{array}{lrcl}
  \mbox{G-test:\citep{Woolf1957}}                                       & \mathrm{GT}(i,j)   & = & 2 \, \sum_{a,b} \mathrm{Obs}^{ab}_{ij} \, \log \frac{ \mathrm{Obs}^{ab}_{ij} } { \mathrm{Exp}^{ab}_{ij} }, \\
  \mbox{Pearson's chi-square:}                                          & \mathrm{CHI}(i,j)  & = &      \sum_{a,b} \frac{ \left(\mathrm{Obs}^{ab}_{ij} - \mathrm{Exp}^{ab}_{ij}\right)^2 }{\mathrm{Exp}^{ab}_{ij}},\\ 
  \mbox{Mutual information:\citep{Shannon48,Gutell94b}}                 & \mathrm{MI}(i,j)   & = &      \sum_{a,b} P^{ab}_{ij} \, \log \frac{ P^{ab}_{ij} }{ p^{a}_{i} \, p^{b}_{j}},                       \\
  \mbox{MI normalized:\citep{Martin05}}                                 & \mathrm{MIr}(i,j)  & = & \frac{\mathrm{MI}(i,j)} {H(i,j)} = \frac{\mathrm{MI}(i,j)} { -\sum_{a,b} P^{ab}_{ij} \log  P^{ab}_{ij}},        \\
  \mbox{MI with gap penalty:\citep{LindgreenKrogh06}}                   & \mathrm{MIg}(i,j)  & = & \mathrm{MI}(i,j) - \frac{N^G_{ij}} {N},                                                                   \\
  \mbox{Obs-Minus-Exp-Squared:\citep{Fodor04}}                          & \mathrm{OMES}(i,j) & = &      \sum_{a,b} \frac{ \left(\mathrm{Obs}^{ab}_{ij} - \mathrm{Exp}^{ab}_{ij}\right)^2 }{N_{ij}},                \\
  \mbox{RNAalifold (RAF):\citep{Hofacker02}}                            & \mathrm{RAF}(i,j)  & = & \mathrm{B}_{i,j},                                      \\
  \mbox{RNAalifold Stacking (RAFS):\citep{LindgreenKrogh06}}            & \mathrm{RAFS}(i,j) & = & \frac{1}{4}\left(\mathrm{B}_{i-1,j+1}+2\,\mathrm{B}_{i,j}+\mathrm{B}_{i+1,j-1}\right).                                      \\
\end{array}
\]
%
\noindent
where $a,b$ are (non-gap) residues; $N$ is the total number of aligned
sequences; $\mathrm{Obs}^{ab}_{ij}$ is the observed count of $a:b$
pairs in columns $i,j$ (only counting when both a,b are residues);
$N_{ij}$ is the total number of residue pairs in columns $i,j$ (only
counting when both a,b are residues); $P^{ab}_{ij}$ is the observed
frequency of pair $a:b$ in columns $i,j$
($P^{ab}_{ij}=\frac{Obs^{ab}_{ij}}{N_{ij}}$); $\mathrm{Exp}^{ab}_{ij}=
N_{ij} p^a_ip^b_j$ is the expected frequency of pair $a:b$ assuming
$i,j$ are independent, where $p^a_i$ are the marginal frequencies of
$a$ residues in column $i$ (averaged to all other positions) ($p^a_i
= \frac{1}{L-1}\sum_{j\neq i} \sum_b P^{ab}_{ij}$); $N^G_{ij} = N -
N_{ij}$ is the number of pairs involving at least one gap symbol; the
definition of $\mathrm{B}_{i,j}$ used in the RAF and RAFS statistics
is involved, a concise definition can be found
elsewhere~\citep{LindgreenKrogh06}.

The background corrections~\citep{DunnGloor07} for a given
covariation statistic above $\mathrm{COV}(i,j)$ are,
%
\[
\begin{array}{lrcl}
  \mbox{\small Average product correction} & \mathrm{COVp}(i,j) & = &  \mathrm{COV}(i,j) - \frac { \mathrm{COV}(i) \mathrm{COV}(j) } { \mathrm{COV} }, \\
  \mbox{\small Average sum correction}     & \mathrm{COVa}(i,j) & = &  \mathrm{COV}(i,j) - \left( \mathrm{COV}(i) + \mathrm{COV}(j) - \mathrm{COV} \right). \\
\end{array}
\]



\subsubsection{\prog{--C2, --C16}}
For all the covariation statistics (except RAF and RAFS), one can do a
16-component (C16) or a two-component (C2) calculation, depending on
whether it uses the 16 possible pair combinations, or those are group
in two classes depending on whether they form a Watson-Crick pair (6
cases, including U:G and G:U), or whether they do not (10 cases).

\rscape's default is the 16 component covariation statistic, unless
the number of sequences in the alignment is $\leq$ 8 or the length of
the alignment is $\leq$ 50, in which case it uses the two-class
covariation statistic.

\subsection{Search options}

\subsubsection{\prog{--cyk}} An optimal secondary structure is computed that includes all significant base pairs. The files for this
maximum-covariation optimal structure all include the suffix
\prog{.cyk.}

When option \prog{--cyk} is used, a file with the original alignment  annotated with the R-scape structure in Stockholm format is produced.
This alignment has the suffix \prog{.cyk.sto}.

\subsubsection{\prog{--naive}} Reports the laundry list of all covariation scores, without any statistical significance (E-value)
associated to them. No null alignments are created.

\subsubsection{\prog{--tstart <n>}} Analyze starting from position $n >= 1$ in the alignment.

\subsubsection{\prog{--tend <n>}} Analyze ending at position $n <= L$ in the alignment.

\subsubsection{\prog{--window <n>}} \rscape\ can be run in a window scanning version for long alignments.
The window size is $n>0$.

\subsubsection{\prog{--slide <n>}} In scanning mode, this options sets the number of positions to move from window to window, $n >0$.



\subsection{Input alignment options}

\subsubsection{\prog{-I <x>}} Only sequences with less than $0<x\leq 1$
pairwise similarity are considered in the analysis.  Pairwise \%
identity is defined as the ratio of identical positions divided by the
minimum length of the two sequences. If this option is not used all
(weighted) sequences are used in the analysis.

\subsubsection{\prog{--gapthresh <x>}} Only columns with less than $0<x\leq 1$ fraction of gaps are considered in the analysis.

\subsubsection{\prog{--consensus}} If the alignment has a GC ``seq\_cons'' field, only consensus positions will be analyzed.

\subsubsection{\prog{--submsa <n>}} Analyzes a random subset of the input alignment.

\subsubsection{\prog{--treefile <f>}} A phylogenetic tree in Newick format can be given (by default a tree is created 
from the alignment using the program FastTree~\citep{Price10}).  \rscape\ checks that the  number of taxa and the names
of the taxa matches for all alignments analyzed.



\subsection{Options for importing a structure}

R-scape does not require to input a structure (either a RNA structure
or a protein contact map).\footnote{ For RNA alignments alone, if no
structure is provided, R-scape computes the best secondary structure
that includes the maximum number of significantly covarying
basepairs.}
\vspace{1mm}

\noindent
There are two ways to provide a contact map (or structure):

\begin{itemize}
\item By providing the alignment in Stockholm format with a ``ss\_cons'' field including the consensus
structure for the alignment. (For RNA alignments only.)
\item By analyzing a 3D structure provided in a PDB file. (For either RNA or peptide alignments.)
\end{itemize}

\noindent
These two options are incompatible with each other. If both a consensus structure is provided, and a PDB
file, the consensus structure will be ignored, and a structure will be calculated from the PDB file
using the program \prog{rnaview}~\citep{YangWesthof03}.

\noindent
If the sequence in the PDB file is found not to be a homolog of those
in the input alignment (using the program nhmmer), no contact map is
produced.


\noindent
The contacts are extracted using two methods:
\begin{itemize}

\item The distance between any two residues is calculated (as the  minimal Euclidean distance between any two atoms, except
for H atoms). Any two pairs at a distance not larger than a maximum
value (\texttt{contmaxD}) are called a ``contact''.

\item In addition if a nucleic acid alignment, R-scape runs the program \texttt{rnaview}~\citep{YangWesthof03}\\
(http://ndbserver.rutgers.edu/ndbmodule/services/download/rnaview.html)
in order to identify RNA base pairs within the contacts.
\end{itemize}

\vspace{2mm}
\noindent
The options that control the input of a structure or contact map are:

\subsubsection{\prog{--pdbfile <s>}} Reads a pdbfile associated to the alignment, and extracts the contacts
from it.

Option \prog{--pdbfile} is incompatible with \prog{--cyk}.

\subsubsection{\prog{--cntmaxD <x>}} Maximum distance (in Angstroms) allowed between two residues to define a ``contact'' is $\langle x\rangle$.

\subsubsection{\prog{--cntmind <n>}} Minimum distance (in residue positions) in the backbone between two residues required to define a ``contact'' is $\langle n\rangle$.
\vspace{5mm}

\noindent
Example of reading a structure from a PDB file for the FMN riboswitch:

\user{bin/R-scape --cntmaxD 4 --cntmind 3 --pdbfile tutorial/3f2q.pdb  tutorial/RF00050.sto}\\

\noindent
This command line extracts contacts from the pdb file that are at a
Eucledian distance $\leq 4 \AA$ in the PDB structure, and such that
they are at least 3 residues apart in the backbone.


\noindent
The output is

\begin{sreoutput}
# R-scape :: RNA Structural Covariation Above Phylogenetic Expectation
# R-scape 0.6.1 (July 2017)
# Copyright (C) 2016 Howard Hughes Medical Institute.
# Freely distributed under the GNU General Public License (GPLv3).
# - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
# 6 215 | bptype WWc
# 6 216 | bptype CONTACT
# 7 214 | bptype WWc
# 7 215 | bptype CONTACT
# 8 213 | bptype WWc
# 8 214 | bptype CONTACT
# 9 212 | bptype WWc
# 9 213 | bptype WWc
#
.
.
.
# 193 201 | bptype WWc
# 193 202 | bptype CONTACT
# 195 198 | bptype WHt
# 198 201 | bptype CONTACT
# contacts  153 (46 bpairs 32 wc bpairs)
# maxD      4.00
# mind      3
# L         139
# alen      221
# MSA RF00050_FMN.3f2q nseq 144 (144) alen 139 (221) avgid 69.18 (68.15) nbpairs 46 (0)
# Method Target_E-val [cov_min,conv_max] [FP | TP True Found | Sen PPV F] 
# GTp    0.05         [-9.58,216.11]     [3 | 14 46 17 | 30.43 82.35 44.44] 
#       left_pos       right_pos        score   E-value
#------------------------------------------------------------
*	     171	     183	216.11	4.41458e-11
*	     170	     184	211.69	8.02215e-11
*	     192	     202	168.72	1.91884e-08
*	       8	     213	149.72	2.19077e-07
*	     172	     182	138.67	8.92814e-07
*	     169	     185	137.23	1.10807e-06
**	      16	      30	133.45	1.70655e-06
*	      84	     186	125.61	4.75589e-06
*	      17	      29	112.05	2.66491e-05
*	       7	     214	111.13	2.96758e-05
*	       6	     215	96.44	0.000194387
*	      36	      87	96.33	0.000194387
~	       5	     216	131.03	0.000489499
*	      94	     163	78.82	0.00183574
 	       7	     213	107.69	0.0094673
 	       7	       8	98.52	0.0308494
*	      37	      86	53.10	0.0438743
\end{sreoutput}

\noindent
All coordinates are relative to the input alignment. The annotation of
all types of RNA basepairs (WWc, WWt, WHc,...) is produced by the
program \prog{rnaview}~\citep{YangWesthof03}.

\subsection{Options for type of pairs tested}

\subsubsection{\prog{--sampleall}}
The statistical test includes all possible pairs in the alignment.
This is the default option (both for nucleic acid or amino acid
alignments) when no list of contacts is identified. 

\subsubsection{\prog{--samplecontacts}}
The statistical test includes all the contacts identified in a PDB
or/and as a RNA secondary structure included with a input alignment in
Stockholm format.  This is the default option for amino acid
alignments if a PDB file is provided.  If no contacts are found, it
reverts to option \prog{--sampleall}.

\subsubsection{\prog{--samplebp}} For RNA alignments with only.
The statistical test includes basepairs of all 12 possible types.
This is the default option for RNA/DNA alignments if a PDB file is
provided.  If no consensus structure is found, it reverts to
option \prog{--sampleall}.

\subsubsection{\prog{--samplewc}} For RNA alignments only.
The statistical test includes only the canonical
(Watson-Crick/Watson-Crick type) basepairs (A:U, G:C, G:U).  This is
the default option for RNA/DNA alignments if a consensus secondary
structure is provided. If no canonical basepairs are found, it reverts
to option \prog{--sampleall}.

\subsection{Output options}

\subsubsection{\prog{--roc}}

Produces a tabular output that provides statistics for each score value.

File \emprog{tutorial/updated\_Arisong.roc} looks like:

\user{more tutorial/updated\_Arisong.roc}
\begin{sreoutput}
# MSA nseq 95 alen 65 avgid 66.352419 nbpairs 20 (20)
# Method: GTp
#cov_score  FP  TP Found  True  Negatives  Sen   PPV     F       E-value
121.79543   0   2  2      20    2060       10.00 100.00  18.18   4.07104e-05
121.44018   0   2  2      20    2060       10.00 100.00  18.18   4.29443e-05
121.08494   0   2  2      20    2060       10.00 100.00  18.18   4.53006e-05
120.72970   0   2  2      20    2060       10.00 100.00  18.18   4.53006e-05
...
\end{sreoutput}

This file produces a tabular output for each alignment as a function
of the covariation score, for plotting ROC curves. The values in the
file are described by the comment line. Notice that the number of
Trues (column 5) and Negatives (column 6) are fixed for a given
secondary structure and do not change.

\subsubsection{\prog{--outmsa <f>}} The actual alignment analyzed can be saved in Stockholm format to file $<$f$>$.


\subsection{Plotting options}

\subsubsection{\prog{--nofigures}} None of the graphical outputs are produced using this option.


\subsubsection{\prog{--r2rall}} Forces R2R to draw all positions in the alignment. By default only
those that are more than 50\% occupied or are base paired are
depicted.


\subsection{Other options}

\subsubsection{\prog{--seed <n>}} Sets the seed of the random number generator to $<$n$>$. Use n = 0 for a random seed.










